// zudoku/dist/vite/pagefind-dev-index.js
import path from "node:path";
import { createIndex } from "pagefind";
import { isRunnableDevEnvironment } from "vite";
import invariant from "../lib/util/invariant.js";
import { joinUrl } from "../lib/util/joinUrl.js";
import { getAppServerEntryPath } from "./config.js";
import { getDevHtml } from "./html.js";
import { InMemoryResponse } from "./prerender/InMemoryResponse.js";
import { routesToPaths } from "./prerender/utils.js";
export async function* buildPagefindDevIndex(vite, config) {
    const { index, errors } = await createIndex();
    invariant(index, `Failed to create pagefind index: ${errors.join(", ")}`);
    const pagefindIndex = index;
    const ssrEnvironment = vite.environments.ssr;
    if (!isRunnableDevEnvironment(ssrEnvironment)) {
        throw new Error("SSR environment is not runnable");
    }
    // Import the entry server module
    const serverModule = (await ssrEnvironment.runner.import(getAppServerEntryPath()));
    const routes = serverModule.getRoutesByConfig(config);
    const paths = routesToPaths(routes);
    const { basePath } = config;
    const template = getDevHtml({
        jsEntry: "/__z/entry.client.tsx",
        dir: config.site?.dir,
    });
    const transformedTemplate = await vite.transformIndexHtml("/", template);
    let indexed = 0;
    for (const urlPath of paths) {
        try {
            const pathname = joinUrl(basePath, urlPath);
            const url = joinUrl("http://localhost", pathname);
            const request = new Request(url);
            const response = new InMemoryResponse();
            await serverModule.render({
                template: transformedTemplate,
                request,
                response,
                routes,
                basePath,
                bypassProtection: true,
            });
            await response.isSent();
            if (response.statusCode < 400) {
                await pagefindIndex.addHTMLFile({
                    url: urlPath,
                    content: response.buffer,
                });
                indexed++;
            }
        }
        catch (error) {
            // biome-ignore lint/suspicious/noConsole: Log errors during indexing
            console.error(`Failed to index ${urlPath}:`, error);
        }
        yield {
            type: "progress",
            total: paths.length,
            current: indexed,
            path: urlPath,
        };
    }
    // Write the index to public/pagefind
    const outputPath = path.join(vite.config.publicDir, "pagefind");
    await pagefindIndex.writeFiles({ outputPath });
    yield { type: "complete", success: true, indexed };
}
