// zudoku/dist/lib/plugins/api-keys/settings/ApiKeyList.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useSuspenseQuery } from "@tanstack/react-query";
import { useState } from "react";
import { useZudoku } from "../../../components/context/ZudokuContext.js";
import { cn } from "../../../util/cn.js";
import { ZudokuError } from "../../../util/invariant.js";
import { CreateApiKeyDialog } from "../CreateApiKeyDialog.js";
import ApiKeyItem from "./ApiKeyItem.js";
export const ApiKeyList = ({ service }) => {
    const context = useZudoku();
    const { data } = useSuspenseQuery({
        queryFn: async () => {
            try {
                return await service.getConsumers(context);
            }
            catch (error) {
                if (error instanceof ZudokuError) {
                    throw error;
                }
                throw new ZudokuError("Cannot get API keys", {
                    cause: error,
                    title: "Error getting API keys",
                    developerHint: "Check the response of the API request for more information.",
                });
            }
        },
        queryKey: ["api-keys"],
        retry: false,
    });
    const [isCreateApiKeyOpen, setIsCreateApiKeyOpen] = useState(false);
    return (_jsx("div", { className: "mt-8", children: data.length === 0 ? (_jsxs("div", { className: "flex col-span-full flex-col justify-center gap-4 items-center p-8 border rounded-sm bg-muted/30 text-muted-foreground", children: [_jsxs("p", { className: "text-center", children: ["You have no API keys yet.", _jsx("br", {}), service.createKey && "Get started and create your first key."] }), service.createKey && (_jsx(CreateApiKeyDialog, { service: service, isOpen: isCreateApiKeyOpen, onOpenChange: setIsCreateApiKeyOpen }))] })) : (_jsx("ul", { className: cn("grid grid-cols-[1fr_min-content] col-span-6"), children: data.map((consumer) => (_jsx(ApiKeyItem, { consumer: consumer, onUpdate: service.updateConsumer, onRollKey: service.rollKey, onDeleteKey: service.deleteKey }, consumer.id))) })) }));
};
