// zudoku/lib/ui/Alert.js
import { j as r } from "../jsx-runtime-BzflLqGi.js";
import { c as n } from "../index-DI5SPFK9.js";
import { c as a } from "../cn-5-Gd1Dss.js";
const l = n(
  "grid gap-0.5 rounded-lg border text-left text-sm has-data-[slot=alert-action]:relative has-data-[slot=alert-action]:pr-18 has-[>svg]:grid-cols-[auto_1fr] has-[>svg]:gap-x-2 *:[svg]:row-span-2 *:[svg]:translate-y-0.5 *:[svg]:text-current *:[svg:not([class*='size-'])]:size-4 w-full relative group/alert",
  {
    variants: {
      variant: {
        default: "bg-card text-card-foreground",
        info: "bg-blue-50 text-blue-700 border-blue-200 dark:bg-blue-900/45 dark:text-blue-200 dark:border-blue-900/75",
        destructive: "text-destructive bg-card bg-destructive/5 border-destructive/20 *:data-[slot=alert-description]:text-destructive *:[svg]:text-current",
        warning: "text-warning-foreground bg-card bg-warning/5 border-warning/50 *:data-[slot=alert-description]:text-warning-foreground *:[svg]:text-current"
      },
      fit: {
        default: "px-2.5 py-2",
        loose: "p-4"
      }
    },
    defaultVariants: {
      variant: "default",
      fit: "default"
    }
  }
);
function c({
  className: t,
  fit: e,
  variant: s,
  ...o
}) {
  return /* @__PURE__ */ r.jsx(
    "div",
    {
      "data-slot": "alert",
      role: "alert",
      className: a(l({ variant: s, fit: e }), t),
      ...o
    }
  );
}
function g({ className: t, ...e }) {
  return /* @__PURE__ */ r.jsx(
    "div",
    {
      "data-slot": "alert-title",
      className: a(
        "font-medium group-has-[>svg]/alert:col-start-2 [&_a]:hover:text-foreground [&_a]:underline [&_a]:underline-offset-3",
        t
      ),
      ...e
    }
  );
}
function f({
  className: t,
  ...e
}) {
  return /* @__PURE__ */ r.jsx(
    "div",
    {
      "data-slot": "alert-description",
      className: a(
        "text-muted-foreground text-sm text-balance md:text-pretty [&_p:not(:last-child)]:mb-4 [&_a]:hover:text-foreground [&_a]:underline [&_a]:underline-offset-3",
        t
      ),
      ...e
    }
  );
}
function x({ className: t, ...e }) {
  return /* @__PURE__ */ r.jsx(
    "div",
    {
      "data-slot": "alert-action",
      className: a("absolute top-2 right-2", t),
      ...e
    }
  );
}
export {
  c as Alert,
  x as AlertAction,
  f as AlertDescription,
  g as AlertTitle
};
