// zudoku/lib/ui/AlertDialog.js
import { j as e } from "../jsx-runtime-BzflLqGi.js";
import * as o from "@radix-ui/react-alert-dialog";
import { c as l } from "../cn-5-Gd1Dss.js";
import { Button as d } from "./Button.js";
function m({
  ...t
}) {
  return /* @__PURE__ */ e.jsx(o.Root, { "data-slot": "alert-dialog", ...t });
}
function f({
  ...t
}) {
  return /* @__PURE__ */ e.jsx(o.Trigger, { "data-slot": "alert-dialog-trigger", ...t });
}
function s({
  ...t
}) {
  return /* @__PURE__ */ e.jsx(o.Portal, { "data-slot": "alert-dialog-portal", ...t });
}
function n({
  className: t,
  ...a
}) {
  return /* @__PURE__ */ e.jsx(
    o.Overlay,
    {
      "data-slot": "alert-dialog-overlay",
      className: l(
        "data-open:animate-in data-closed:animate-out data-closed:fade-out-0 data-open:fade-in-0 bg-black/10 duration-100 supports-backdrop-filter:backdrop-blur-xs fixed inset-0 z-50",
        t
      ),
      ...a
    }
  );
}
function x({
  className: t,
  size: a = "default",
  ...r
}) {
  return /* @__PURE__ */ e.jsxs(s, { children: [
    /* @__PURE__ */ e.jsx(n, {}),
    /* @__PURE__ */ e.jsx(
      o.Content,
      {
        "data-slot": "alert-dialog-content",
        "data-size": a,
        className: l(
          "data-open:animate-in data-closed:animate-out data-closed:fade-out-0 data-open:fade-in-0 data-closed:zoom-out-95 data-open:zoom-in-95 bg-background ring-foreground/10 gap-4 rounded-xl p-4 ring-1 duration-100 data-[size=default]:max-w-xs data-[size=sm]:max-w-xs data-[size=default]:sm:max-w-sm group/alert-dialog-content fixed top-1/2 left-1/2 z-50 grid w-full -translate-x-1/2 -translate-y-1/2 outline-none",
          t
        ),
        ...r
      }
    )
  ] });
}
function p({
  className: t,
  ...a
}) {
  return /* @__PURE__ */ e.jsx(
    "div",
    {
      "data-slot": "alert-dialog-header",
      className: l(
        "grid grid-rows-[auto_1fr] place-items-center gap-1.5 text-center has-data-[slot=alert-dialog-media]:grid-rows-[auto_auto_1fr] has-data-[slot=alert-dialog-media]:gap-x-4 sm:group-data-[size=default]/alert-dialog-content:place-items-start sm:group-data-[size=default]/alert-dialog-content:text-left sm:group-data-[size=default]/alert-dialog-content:has-data-[slot=alert-dialog-media]:grid-rows-[auto_1fr]",
        t
      ),
      ...a
    }
  );
}
function j({
  className: t,
  ...a
}) {
  return /* @__PURE__ */ e.jsx(
    "div",
    {
      "data-slot": "alert-dialog-footer",
      className: l(
        "bg-muted/50 -mx-4 -mb-4 rounded-b-xl border-t p-4 flex flex-col-reverse gap-2 group-data-[size=sm]/alert-dialog-content:grid group-data-[size=sm]/alert-dialog-content:grid-cols-2 sm:flex-row sm:justify-end",
        t
      ),
      ...a
    }
  );
}
function z({
  className: t,
  ...a
}) {
  return /* @__PURE__ */ e.jsx(
    "div",
    {
      "data-slot": "alert-dialog-media",
      className: l(
        "bg-muted mb-2 inline-flex size-10 items-center justify-center rounded-md sm:group-data-[size=default]/alert-dialog-content:row-span-2 *:[svg:not([class*='size-'])]:size-6",
        t
      ),
      ...a
    }
  );
}
function b({
  className: t,
  ...a
}) {
  return /* @__PURE__ */ e.jsx(
    o.Title,
    {
      "data-slot": "alert-dialog-title",
      className: l(
        "text-base font-medium sm:group-data-[size=default]/alert-dialog-content:group-has-data-[slot=alert-dialog-media]/alert-dialog-content:col-start-2",
        t
      ),
      ...a
    }
  );
}
function A({
  className: t,
  ...a
}) {
  return /* @__PURE__ */ e.jsx(
    o.Description,
    {
      "data-slot": "alert-dialog-description",
      className: l(
        "text-muted-foreground *:[a]:hover:text-foreground text-sm text-balance md:text-pretty *:[a]:underline *:[a]:underline-offset-3",
        t
      ),
      ...a
    }
  );
}
function D({
  className: t,
  variant: a = "default",
  size: r = "default",
  ...i
}) {
  return /* @__PURE__ */ e.jsx(d, { variant: a, size: r, asChild: !0, children: /* @__PURE__ */ e.jsx(
    o.Action,
    {
      "data-slot": "alert-dialog-action",
      className: l(t),
      ...i
    }
  ) });
}
function h({
  className: t,
  variant: a = "outline",
  size: r = "default",
  ...i
}) {
  return /* @__PURE__ */ e.jsx(d, { variant: a, size: r, asChild: !0, children: /* @__PURE__ */ e.jsx(
    o.Cancel,
    {
      "data-slot": "alert-dialog-cancel",
      className: l(t),
      ...i
    }
  ) });
}
export {
  m as AlertDialog,
  D as AlertDialogAction,
  h as AlertDialogCancel,
  x as AlertDialogContent,
  A as AlertDialogDescription,
  j as AlertDialogFooter,
  p as AlertDialogHeader,
  z as AlertDialogMedia,
  n as AlertDialogOverlay,
  s as AlertDialogPortal,
  b as AlertDialogTitle,
  f as AlertDialogTrigger
};
