// zudoku/lib/ui/Badge.js
import { j as n } from "../jsx-runtime-BzflLqGi.js";
import { Slot as o } from "@radix-ui/react-slot";
import { c as s } from "../index-DI5SPFK9.js";
import { c as d } from "../cn-5-Gd1Dss.js";
const g = s(
  "inline-flex items-center justify-center rounded-full border px-2 py-0.5 text-xs font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden",
  {
    variants: {
      variant: {
        default: "border-transparent bg-primary text-primary-foreground [a&]:hover:bg-primary/90",
        secondary: "border-transparent bg-secondary text-secondary-foreground [a&]:hover:bg-secondary/90",
        warning: "border-transparent bg-warning text-warning-foreground [a&]:hover:bg-warning/90",
        destructive: "border-transparent bg-destructive text-white [a&]:hover:bg-destructive/90 focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60",
        outline: "text-foreground [a&]:hover:bg-accent [a&]:hover:text-accent-foreground",
        muted: "bg-muted text-muted-foreground "
      }
    },
    defaultVariants: {
      variant: "default"
    }
  }
);
function f({
  className: r,
  variant: e,
  asChild: t = !1,
  ...a
}) {
  const i = t ? o : "span";
  return /* @__PURE__ */ n.jsx(
    i,
    {
      "data-slot": "badge",
      className: d(g({ variant: e }), r),
      ...a
    }
  );
}
export {
  f as Badge,
  g as badgeVariants
};
