// zudoku/lib/ui/ButtonGroup.js
import { j as e } from "../jsx-runtime-BzflLqGi.js";
import { Slot as s } from "@radix-ui/react-slot";
import { c as i } from "../index-DI5SPFK9.js";
import { c as a } from "../cn-5-Gd1Dss.js";
import { Separator as l } from "./Separator.js";
const d = i(
  "flex w-fit items-stretch [&>*]:focus-visible:z-10 [&>*]:focus-visible:relative [&>[data-slot=select-trigger]:not([class*='w-'])]:w-fit [&>input]:flex-1 has-[select[aria-hidden=true]:last-child]:[&>[data-slot=select-trigger]:last-of-type]:rounded-r-md has-[>[data-slot=button-group]]:gap-2",
  {
    variants: {
      orientation: {
        horizontal: "[&>*:not(:first-child)]:rounded-l-none [&>*:not(:first-child)]:border-l-0 [&>*:not(:last-child)]:rounded-r-none",
        vertical: "flex-col [&>*:not(:first-child)]:rounded-t-none [&>*:not(:first-child)]:border-t-0 [&>*:not(:last-child)]:rounded-b-none"
      }
    },
    defaultVariants: {
      orientation: "horizontal"
    }
  }
);
function h({
  className: o,
  orientation: t,
  ...r
}) {
  return /* @__PURE__ */ e.jsx(
    "div",
    {
      role: "group",
      "data-slot": "button-group",
      "data-orientation": t,
      className: a(d({ orientation: t }), o),
      ...r
    }
  );
}
function g({
  className: o,
  asChild: t = !1,
  ...r
}) {
  const n = t ? s : "div";
  return /* @__PURE__ */ e.jsx(
    n,
    {
      className: a(
        "bg-muted flex items-center gap-2 rounded-md border px-4 text-sm font-medium shadow-xs [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4",
        o
      ),
      ...r
    }
  );
}
function v({
  className: o,
  orientation: t = "vertical",
  ...r
}) {
  return /* @__PURE__ */ e.jsx(
    l,
    {
      "data-slot": "button-group-separator",
      orientation: t,
      className: a(
        "bg-input relative m-0! self-stretch data-[orientation=vertical]:h-auto",
        o
      ),
      ...r
    }
  );
}
export {
  h as ButtonGroup,
  v as ButtonGroupSeparator,
  g as ButtonGroupText,
  d as buttonGroupVariants
};
