// zudoku/lib/ui/CodeBlock.js
import { j as o } from "../jsx-runtime-BzflLqGi.js";
import { CheckIcon as d, CopyIcon as h } from "lucide-react";
import { lazy as t, Suspense as w, useRef as x } from "react";
import { c as s } from "../cn-5-Gd1Dss.js";
import { u } from "../useCopyToClipboard-B_085nfO.js";
const $ = {
  typescript: {
    pattern: /^(ts|typescript)$/,
    icon: t(() => import("../typescript-C26xdBDC.js")),
    width: 15
  },
  javascript: {
    pattern: /^(js|javascript)$/,
    icon: t(() => import("../javascript-CcmIpL4G.js"))
  },
  react: {
    pattern: /^(react|tsx|jsx)$/,
    icon: t(() => import("../react-DHpVpxRv.js")),
    width: 22
  },
  markdown: {
    pattern: /^(md|markdown)$/,
    icon: t(() => import("../markdown-DYGWCmGQ.js")),
    width: 22
  },
  mdx: {
    pattern: /^mdx$/,
    icon: t(() => import("../mdx-CHwYvXd4.js")),
    width: 28
  },
  java: {
    pattern: /^(java)$/,
    icon: t(() => import("../java-CVLzHfb1.js")),
    width: 22
  },
  json: {
    pattern: /^jsonc?$/,
    icon: t(() => import("../json-4AyP4uiY.js"))
  },
  yaml: {
    pattern: /^yaml$/,
    icon: t(() => import("../yaml-BGsJItKv.js"))
  },
  toml: {
    pattern: /^toml$/,
    icon: t(() => import("../toml-Co9mpdct.js"))
  },
  shell: {
    pattern: /^(shell|bash|sh|zsh|term|terminal|ansi)$/,
    icon: t(() => import("../shell-HUv9oVtp.js"))
  },
  python: {
    pattern: /^(py|python)$/,
    icon: t(() => import("../python-QIQAE5Ei.js"))
  },
  csharp: {
    pattern: /^(cs|csharp|vb)$/,
    icon: t(() => import("../csharp-D8MIL50B.js")),
    width: 20
  },
  rust: {
    pattern: /^(rs|rust)$/,
    icon: t(() => import("../rust-DYnLHAi2.js")),
    width: 20
  },
  ruby: {
    pattern: /^(rb|ruby)$/,
    icon: t(() => import("../ruby-B2dU8Ny5.js"))
  },
  php: {
    pattern: /^php$/,
    icon: t(() => import("../php-rQXzo7K_.js")),
    width: 24
  },
  html: {
    pattern: /^html?$/,
    icon: t(() => import("../html-MGnI2uzP.js")),
    width: 15
  },
  css: {
    pattern: /^css$/,
    icon: t(() => import("../css-Bt6hr1td.js"))
  },
  objectivec: {
    pattern: /^(objc|objectivec)$/,
    icon: t(() => import("../objectivec-HZY8shkd.js")),
    width: 16
  },
  swift: {
    pattern: /^swift$/,
    icon: t(() => import("../swift-B4z6ig1Z.js"))
  },
  go: {
    pattern: /^go$/,
    icon: t(() => import("../go-D2VsmIOS.js")),
    width: 28
  },
  xml: {
    pattern: /^xml$/,
    icon: t(() => import("../xml-BQOOC04j.js"))
  },
  kotlin: {
    pattern: /^(kt|kotlin)$/,
    icon: t(() => import("../kotlin-v2plddBQ.js")),
    width: 14
  },
  graphql: {
    pattern: /^(gql|graphql)$/,
    icon: t(() => import("../graphql-BtA6M4m5.js"))
  },
  zig: {
    pattern: /^zig$/,
    icon: t(() => import("../zig-CUV2sTct.js")),
    width: 28
  },
  scala: {
    pattern: /^scala$/,
    icon: t(() => import("../scala-CeKInBR8.js")),
    width: 18
  },
  dart: {
    pattern: /^dart$/,
    icon: t(() => import("../dart-B0vy1jWB.js")),
    width: 20
  },
  ocaml: {
    pattern: /^ocaml$/,
    icon: t(() => import("../ocaml-DqsdDdwb.js")),
    width: 20
  },
  c: {
    pattern: /^c$/,
    icon: t(() => import("../c-B-NUhs61.js")),
    width: 20
  },
  cpp: {
    pattern: /^(cpp|c\+\+)$/,
    icon: t(() => import("../cpp-79Paht7T.js")),
    width: 20
  },
  commonlisp: {
    pattern: /^(clisp|common-lisp|lisp|cl|commonlisp)$/,
    icon: t(() => import("../commonlisp-De080z23.js")),
    width: 18
  },
  elixir: {
    pattern: /^elixir$/,
    icon: t(() => import("../elixir-Ds8r0sF8.js")),
    width: 18
  },
  powershell: {
    pattern: /^(pshell|powershell|ps1)$/,
    icon: t(() => import("../powershell-CQje9pm1.js")),
    width: 18
  }
}, b = ({ language: i }) => {
  if (!i) return null;
  const r = Object.values($).find((n) => n.pattern.test(i));
  if (!r) return null;
  const e = r.width ?? 18;
  return /* @__PURE__ */ o.jsx(w, { fallback: /* @__PURE__ */ o.jsx("div", { style: { width: e, height: "1em" } }), children: /* @__PURE__ */ o.jsx(r.icon, { className: "h-auto", style: { width: e } }) });
}, k = ({
  children: i,
  title: r = "Code",
  language: e,
  showCopy: n = "hover",
  showLanguageIndicator: j,
  showLineNumbers: a,
  ...l
}) => {
  const [p, m] = u(), c = x(null);
  return i ? /* @__PURE__ */ o.jsxs(
    "div",
    {
      className: s(
        "border code-block-wrapper relative group rounded-xl overflow-hidden",
        a && "line-numbers",
        l.className
      ),
      children: [
        /* @__PURE__ */ o.jsxs("div", { className: "border-b flex items-center py-1 font-sans bg-black/2", children: [
          /* @__PURE__ */ o.jsxs("div", { className: "flex items-center gap-1.5 flex-1 text-sm w-full px-3", children: [
            /* @__PURE__ */ o.jsx(b, { language: e }),
            r
          ] }),
          " ",
          n !== "never" && /* @__PURE__ */ o.jsx(
            "button",
            {
              type: "button",
              "aria-label": "Copy code",
              title: "Copy code",
              className: s(
                "transition p-1.5 mx-1 rounded-lg",
                !p && "hover:bg-accent hover:brightness-95"
              ),
              disabled: p,
              onClick: () => {
                c.current?.textContent && m(c.current.textContent);
              },
              children: p ? /* @__PURE__ */ o.jsx(d, { className: "text-emerald-600", size: 14 }) : /* @__PURE__ */ o.jsx(h, { size: 14 })
            }
          )
        ] }),
        /* @__PURE__ */ o.jsx(
          "div",
          {
            className: "code-block text-sm not-prose scrollbar [&_code]:px-3 [&_code]:py-2",
            ref: c,
            children: i
          }
        )
      ]
    }
  ) : null;
};
export {
  k as CodeBlock
};
