// zudoku/lib/ui/Command.js
import { j as o } from "../jsx-runtime-BzflLqGi.js";
import { _ as t } from "../index-CrcNWbel.js";
import { SearchIcon as i } from "lucide-react";
import * as s from "react";
import { D as p, a as l } from "../Dialog-hlvmmQ_c.js";
import { c as r } from "../cn-5-Gd1Dss.js";
const m = s.forwardRef(({ className: a, ...e }, d) => /* @__PURE__ */ o.jsx(
  t,
  {
    ref: d,
    className: r(
      "flex h-full w-full flex-col overflow-hidden rounded-md bg-popover text-popover-foreground",
      a
    ),
    ...e
  }
));
m.displayName = t.displayName;
const j = ({
  children: a,
  command: e,
  content: d,
  ...n
}) => /* @__PURE__ */ o.jsx(p, { ...n, children: /* @__PURE__ */ o.jsx(
  l,
  {
    ...d,
    className: r(
      "overflow-hidden p-0 shadow-lg top-[15vh] translate-y-[0%]",
      d?.className
    ),
    "aria-describedby": void 0,
    children: /* @__PURE__ */ o.jsx(
      m,
      {
        ...e,
        className: r(
          "[&_[cmdk-group-heading]]:px-2 [&_[cmdk-group-heading]]:font-medium [&_[cmdk-group-heading]]:text-muted-foreground [&_[cmdk-group]:not([hidden])_~[cmdk-group]]:pt-0 [&_[cmdk-group]]:px-2 [&_[cmdk-input-wrapper]_svg]:h-5 [&_[cmdk-input-wrapper]_svg]:w-5 [&_[cmdk-input]]:h-12 [&_[cmdk-item]]:px-2 [&_[cmdk-item]]:py-3 [&_[cmdk-item]_svg]:h-5 [&_[cmdk-item]_svg]:w-5",
          e?.className
        ),
        children: a
      }
    )
  }
) }), c = s.forwardRef(({ className: a, ...e }, d) => /* @__PURE__ */ o.jsxs("div", { className: "flex items-center gap-2 border-b px-3", "cmdk-input-wrapper": "", children: [
  /* @__PURE__ */ o.jsx(i, { className: "size-4! shrink-0 opacity-50" }),
  /* @__PURE__ */ o.jsx(
    t.Input,
    {
      ref: d,
      className: r(
        "flex h-11 w-full rounded-md bg-transparent py-3 text-sm outline-hidden placeholder:text-muted-foreground disabled:cursor-not-allowed disabled:opacity-50",
        a
      ),
      ...e
    }
  )
] }));
c.displayName = t.Input.displayName;
const u = s.forwardRef(({ className: a, ...e }, d) => /* @__PURE__ */ o.jsx("div", { className: "flex items-center", "cmdk-input-wrapper": "", children: /* @__PURE__ */ o.jsx(
  t.Input,
  {
    ref: d,
    className: r(
      "flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-xs transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-hidden disabled:cursor-not-allowed disabled:opacity-50",
      "focus-visible:outline-hidden focus-visible:ring-1 focus-visible:ring-ring",
      a
    ),
    ...e
  }
) }));
u.displayName = t.Input.displayName;
const f = s.forwardRef(({ className: a, ...e }, d) => /* @__PURE__ */ o.jsx(
  t.List,
  {
    ref: d,
    className: r("max-h-[300px] overflow-y-auto overflow-x-hidden", a),
    ...e
  }
));
f.displayName = t.List.displayName;
const x = s.forwardRef((a, e) => /* @__PURE__ */ o.jsx(
  t.Empty,
  {
    ref: e,
    className: "py-6 text-center text-sm",
    ...a
  }
));
x.displayName = t.Empty.displayName;
const g = s.forwardRef(({ className: a, ...e }, d) => /* @__PURE__ */ o.jsx(
  t.Group,
  {
    ref: d,
    className: r(
      "overflow-hidden p-1 text-foreground [&_[cmdk-group-heading]]:px-2 [&_[cmdk-group-heading]]:py-1.5 [&_[cmdk-group-heading]]:text-xs [&_[cmdk-group-heading]]:font-medium [&_[cmdk-group-heading]]:text-muted-foreground",
      a
    ),
    ...e
  }
));
g.displayName = t.Group.displayName;
const h = s.forwardRef(({ className: a, ...e }, d) => /* @__PURE__ */ o.jsx(
  t.Separator,
  {
    ref: d,
    className: r("-mx-1 h-px bg-border", a),
    ...e
  }
));
h.displayName = t.Separator.displayName;
const y = s.forwardRef(({ className: a, ...e }, d) => /* @__PURE__ */ o.jsx(
  t.Item,
  {
    ref: d,
    className: r(
      "relative flex cursor-default gap-2 select-none items-center rounded-xs px-2 py-1.5 text-sm outline-hidden data-[disabled=true]:pointer-events-none data-[selected='true']:bg-accent data-[selected=true]:text-accent-foreground data-[disabled=true]:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0",
      a
    ),
    ...e
  }
));
y.displayName = t.Item.displayName;
const N = ({
  className: a,
  ...e
}) => /* @__PURE__ */ o.jsx(
  "span",
  {
    className: r(
      "ml-auto text-xs tracking-widest text-muted-foreground",
      a
    ),
    ...e
  }
);
N.displayName = "CommandShortcut";
export {
  m as Command,
  j as CommandDialog,
  x as CommandEmpty,
  g as CommandGroup,
  u as CommandInlineInput,
  c as CommandInput,
  y as CommandItem,
  f as CommandList,
  h as CommandSeparator,
  N as CommandShortcut
};
