// zudoku/lib/ui/Dialog.js
import { j as a } from "../jsx-runtime-BzflLqGi.js";
import * as e from "@radix-ui/react-dialog";
import { CircleIcon as n, XIcon as i } from "lucide-react";
import { c as r } from "../cn-5-Gd1Dss.js";
function x({
  ...t
}) {
  return /* @__PURE__ */ a.jsx(e.Root, { "data-slot": "dialog", ...t });
}
function p({
  ...t
}) {
  return /* @__PURE__ */ a.jsx(e.Trigger, { "data-slot": "dialog-trigger", ...t });
}
function d({
  ...t
}) {
  return /* @__PURE__ */ a.jsx(e.Portal, { "data-slot": "dialog-portal", ...t });
}
function m({
  ...t
}) {
  return /* @__PURE__ */ a.jsx(e.Close, { "data-slot": "dialog-close", ...t });
}
function c({
  className: t,
  ...o
}) {
  return /* @__PURE__ */ a.jsx(
    e.Overlay,
    {
      "data-slot": "dialog-overlay",
      className: r(
        "backdrop-blur-xs transition-all",
        "data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 fixed inset-0 z-50 bg-black/32",
        t
      ),
      ...o
    }
  );
}
function j({
  className: t,
  children: o,
  showCloseButton: s = !0,
  ...l
}) {
  return /* @__PURE__ */ a.jsxs(d, { "data-slot": "dialog-portal", children: [
    /* @__PURE__ */ a.jsx(c, {}),
    /* @__PURE__ */ a.jsxs(
      e.Content,
      {
        "data-slot": "dialog-content",
        className: r(
          "bg-background data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 fixed top-[50%] left-[50%] z-50 grid w-full max-w-[calc(100%-2rem)] translate-x-[-50%] translate-y-[-50%] gap-4 rounded-lg border p-6 shadow-lg duration-200 sm:max-w-xl",
          t
        ),
        ...l,
        children: [
          o,
          s && /* @__PURE__ */ a.jsxs(e.Close, { className: "absolute p-3 group -left-1.5 -top-1.5 rounded-xs opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-hidden focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-accent data-[state=open]:text-muted-foreground", children: [
            /* @__PURE__ */ a.jsx(
              n,
              {
                className: "h-2 w-2 transition-all border border-transparent  group-hover:border-white/20 rounded-full group-hover:scale-180 group-hover:text-red-400",
                strokeWidth: 2,
                fill: "currentColor"
              }
            ),
            /* @__PURE__ */ a.jsx(
              i,
              {
                className: "h-2 w-2 absolute top-3 left-3 text-transparent group-hover:text-red-800 transition-colors",
                strokeWidth: 2,
                fill: "currentColor"
              }
            ),
            /* @__PURE__ */ a.jsx("span", { className: "sr-only", children: "Close" })
          ] })
        ]
      }
    )
  ] });
}
function h({ className: t, ...o }) {
  return /* @__PURE__ */ a.jsx(
    "div",
    {
      "data-slot": "dialog-header",
      className: r("flex flex-col gap-2 text-center sm:text-left", t),
      ...o
    }
  );
}
function b({ className: t, ...o }) {
  return /* @__PURE__ */ a.jsx(
    "div",
    {
      "data-slot": "dialog-footer",
      className: r(
        "flex flex-col-reverse gap-2 sm:flex-row sm:justify-end",
        t
      ),
      ...o
    }
  );
}
function v({
  className: t,
  ...o
}) {
  return /* @__PURE__ */ a.jsx(
    e.Title,
    {
      "data-slot": "dialog-title",
      className: r("text-lg leading-none font-semibold", t),
      ...o
    }
  );
}
function D({
  className: t,
  ...o
}) {
  return /* @__PURE__ */ a.jsx(
    e.Description,
    {
      "data-slot": "dialog-description",
      className: r("text-muted-foreground text-sm", t),
      ...o
    }
  );
}
export {
  x as Dialog,
  m as DialogClose,
  j as DialogContent,
  D as DialogDescription,
  b as DialogFooter,
  h as DialogHeader,
  c as DialogOverlay,
  d as DialogPortal,
  v as DialogTitle,
  p as DialogTrigger
};
