// zudoku/lib/ui/DismissibleAlert.js
import { j as s } from "../jsx-runtime-BzflLqGi.js";
import { Slot as m } from "@radix-ui/react-slot";
import { XIcon as c } from "lucide-react";
import { createContext as a, useState as u, useCallback as x, use as p } from "react";
import { u as A } from "../useLatest-hmRS46UF.js";
import { Alert as b, AlertAction as f } from "./Alert.js";
import { Button as C } from "./Button.js";
const n = a({
  dismiss: () => {
  }
}), D = () => p(n), I = ({ onDismiss: t, ...e }) => {
  const [i, r] = u(!1), o = A(t), l = x(() => {
    r(!0), o.current?.();
  }, [o]);
  return i ? null : /* @__PURE__ */ s.jsx(n, { value: { dismiss: l }, children: /* @__PURE__ */ s.jsx(b, { ...e }) });
}, d = ({
  asChild: t,
  ...e
}) => {
  const { dismiss: i } = D(), r = t ? m : C;
  return /* @__PURE__ */ s.jsx(r, { onClick: i, ...e });
}, N = ({ children: t }) => /* @__PURE__ */ s.jsx(f, { children: /* @__PURE__ */ s.jsx(
  d,
  {
    variant: "ghost",
    size: "icon-xxs",
    className: "hover:text-current hover:bg-[color-mix(in_srgb,currentColor_10%,transparent)]!",
    children: t ?? /* @__PURE__ */ s.jsx(c, { className: "size-3.5" })
  }
) });
export {
  I as DismissibleAlert,
  N as DismissibleAlertAction,
  d as DismissibleAlertClose
};
