// zudoku/lib/ui/EmbeddedCodeBlock.js
import { j as e } from "../jsx-runtime-BzflLqGi.js";
import { CheckIcon as p, CopyIcon as x } from "lucide-react";
import { useState as f, useRef as v } from "react";
import { c as t } from "../cn-5-Gd1Dss.js";
import { Button as b } from "./Button.js";
const C = ({
  children: s,
  fullHeight: n,
  language: c,
  showCopy: o = "hover",
  showCopyText: a,
  showLanguageIndicator: d = !0,
  showLineNumbers: u,
  ...m
}) => {
  const [i, l] = f(!1), r = v(null);
  return s ? /* @__PURE__ */ e.jsxs(
    "div",
    {
      className: t(
        "code-block-wrapper relative group bg-muted/50",
        u && "line-numbers",
        n && "h-full"
      ),
      children: [
        /* @__PURE__ */ e.jsx("div", { className: "relative overflow-auto", children: /* @__PURE__ */ e.jsx(
          "div",
          {
            className: t(
              "code-block text-sm not-prose scrollbar [&>pre]:overflow-x-auto [&_code]:p-2",
              n && "h-full [&>pre]:h-full",
              m.className
            ),
            ref: r,
            children: s
          }
        ) }),
        d && /* @__PURE__ */ e.jsx(
          "span",
          {
            className: t(
              "absolute top-1.5 end-3 text-[11px]! font-mono text-muted-foreground transition group-hover:opacity-0 pointer-events-none",
              o === "always" && "hidden"
            ),
            children: c
          }
        ),
        o !== "never" && /* @__PURE__ */ e.jsxs(
          b,
          {
            type: "button",
            variant: "outline",
            size: "icon-xs",
            className: t(
              "absolute top-2 end-2 p-2",
              o === "hover" && "opacity-0 group-hover:opacity-100",
              a && "flex gap-2 items-center font-medium"
            ),
            disabled: i,
            onClick: () => {
              r.current?.textContent && (l(!0), navigator.clipboard.writeText(r.current.textContent), setTimeout(() => l(!1), 2e3));
            },
            children: [
              i ? /* @__PURE__ */ e.jsx(
                p,
                {
                  className: "shrink-0 text-emerald-600 dark:text-emerald-300",
                  size: 13,
                  strokeWidth: 2.5,
                  absoluteStrokeWidth: !0
                }
              ) : /* @__PURE__ */ e.jsx(x, { className: "shrink-0", size: 13 }),
              a && "Copy"
            ]
          }
        )
      ]
    }
  ) : null;
};
export {
  C as EmbeddedCodeBlock
};
