// zudoku/lib/ui/Frame.js
import { j as a } from "../jsx-runtime-BzflLqGi.js";
import { c as t } from "../cn-5-Gd1Dss.js";
function s({ className: e, ...r }) {
  return /* @__PURE__ */ a.jsx(
    "div",
    {
      "data-slot": "frame",
      className: t(
        "relative flex flex-col rounded-2xl bg-muted p-1",
        e
      ),
      ...r
    }
  );
}
function n({ className: e, ...r }) {
  return /* @__PURE__ */ a.jsx(
    "div",
    {
      "data-slot": "frame-panel",
      className: t(
        "relative bg-clip-padding rounded-xl border bg-card p-5 shadow-xs",
        "before:pointer-events-none before:absolute before:inset-0 before:rounded-[calc(var(--radius-xl)-1px)] before:shadow-[0_1px_--theme(--color-black/4%)] dark:bg-clip-border dark:before:shadow-[0_-1px_--theme(--color-white/8%)]",
        e
      ),
      ...r
    }
  );
}
function d({ className: e, ...r }) {
  return /* @__PURE__ */ a.jsx(
    "header",
    {
      "data-slot": "frame-panel-header",
      className: t("flex flex-col p-4", e),
      ...r
    }
  );
}
function f({ className: e, ...r }) {
  return /* @__PURE__ */ a.jsx(
    "div",
    {
      "data-slot": "frame-panel-title",
      className: t("text-sm font-semibold", e),
      ...r
    }
  );
}
function m({
  className: e,
  ...r
}) {
  return /* @__PURE__ */ a.jsx(
    "div",
    {
      "data-slot": "frame-panel-description",
      className: t("text-sm text-muted-foreground", e),
      ...r
    }
  );
}
function i({ className: e, ...r }) {
  return /* @__PURE__ */ a.jsx(
    "footer",
    {
      "data-slot": "frame-panel-footer",
      className: t("flex flex-col gap-1 px-5 py-4", e),
      ...r
    }
  );
}
export {
  s as Frame,
  m as FrameDescription,
  i as FrameFooter,
  d as FrameHeader,
  n as FramePanel,
  f as FrameTitle
};
