// zudoku/lib/ui/NativeSelect.js
import { j as t } from "../jsx-runtime-BzflLqGi.js";
import { ChevronDownIcon as o } from "lucide-react";
import { c as i } from "../cn-5-Gd1Dss.js";
function l({ className: e, ...r }) {
  return /* @__PURE__ */ t.jsxs(
    "div",
    {
      className: "group/native-select relative w-fit has-[select:disabled]:opacity-50",
      "data-slot": "native-select-wrapper",
      children: [
        /* @__PURE__ */ t.jsx(
          "select",
          {
            "data-slot": "native-select",
            className: i(
              "border-input placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 dark:hover:bg-input/50 h-9 w-full min-w-0 appearance-none rounded-md border bg-transparent px-3 py-2 pr-9 text-sm shadow-xs transition-[color,box-shadow] outline-none disabled:pointer-events-none disabled:cursor-not-allowed",
              "focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px]",
              "aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive",
              "scheme-light dark:scheme-dark",
              e
            ),
            ...r
          }
        ),
        /* @__PURE__ */ t.jsx(
          o,
          {
            className: "text-muted-foreground pointer-events-none absolute top-1/2 right-3.5 size-4 -translate-y-1/2 opacity-50 select-none",
            "aria-hidden": "true",
            "data-slot": "native-select-icon"
          }
        )
      ]
    }
  );
}
function d({ ...e }) {
  return /* @__PURE__ */ t.jsx("option", { "data-slot": "native-select-option", ...e });
}
function c({
  className: e,
  ...r
}) {
  return /* @__PURE__ */ t.jsx(
    "optgroup",
    {
      "data-slot": "native-select-optgroup",
      className: i(e),
      ...r
    }
  );
}
export {
  l as NativeSelect,
  c as NativeSelectOptGroup,
  d as NativeSelectOption
};
