// zudoku/lib/ui/Secret.js
import { j as e } from "../jsx-runtime-BzflLqGi.js";
import { EyeOffIcon as g, EyeIcon as b, CheckIcon as u, CopyIcon as N } from "lucide-react";
import { useState as y } from "react";
import { c as l } from "../cn-5-Gd1Dss.js";
import { u as w } from "../useCopyToClipboard-B_085nfO.js";
import { Button as t } from "./Button.js";
const k = ({
  secret: o,
  revealed: a = !1,
  previewChars: i = 5,
  className: s
}) => {
  if (a)
    return /* @__PURE__ */ e.jsx("span", { className: s, children: o });
  const c = "•••• ".repeat(o.slice(0, -i).length / 5) + "•".repeat(o.slice(0, -i).length % 5), n = i > 0 ? o.slice(-i) : "";
  return /* @__PURE__ */ e.jsx(
    "span",
    {
      className: l(
        "block overflow-hidden text-ellipsis whitespace-nowrap",
        s
      ),
      dir: "rtl",
      children: /* @__PURE__ */ e.jsxs("span", { dir: "ltr", className: "inline", children: [
        /* @__PURE__ */ e.jsx("span", { className: "opacity-50", children: c }),
        n && /* @__PURE__ */ e.jsx("span", { children: n })
      ] })
    }
  );
}, T = ({
  secret: o,
  preview: a = 5,
  className: i,
  status: s,
  onCopy: c,
  revealed: n,
  onReveal: d,
  noReveal: x
}) => {
  const m = Math.abs(a), [p, h] = y(!1), [f, j] = w(), r = n ?? p;
  return /* @__PURE__ */ e.jsxs(
    "div",
    {
      className: l(
        "flex gap-2 items-center text-sm border rounded-md px-1",
        i
      ),
      children: [
        /* @__PURE__ */ e.jsxs(
          "div",
          {
            className: l(
              "font-mono w-full h-9 items-center flex px-2 text-xs gap-2 min-w-0",
              r ? "overflow-x-auto" : "overflow-hidden"
            ),
            children: [
              s && /* @__PURE__ */ e.jsx(
                "div",
                {
                  className: l(
                    "rounded-full shrink-0 w-2 h-2 mr-2",
                    s === "active" && "bg-emerald-400",
                    s === "expired" && "bg-neutral-200",
                    s === "expiring" && "bg-yellow-400",
                    s === "revoked" && "bg-red-400",
                    s === "none" && "opacity-0"
                  )
                }
              ),
              /* @__PURE__ */ e.jsx(
                k,
                {
                  secret: o,
                  revealed: r,
                  previewChars: m
                }
              )
            ]
          }
        ),
        x !== !0 && /* @__PURE__ */ e.jsx(
          t,
          {
            variant: "ghost",
            onClick: () => {
              h((v) => !v), d?.(!r);
            },
            size: "icon-xxs",
            children: r ? /* @__PURE__ */ e.jsx(g, { className: "size-3.5" }) : /* @__PURE__ */ e.jsx(b, { className: "size-3.5" })
          }
        ),
        /* @__PURE__ */ e.jsx(
          t,
          {
            variant: "ghost",
            onClick: () => {
              j(o), c?.(o);
            },
            size: "icon-xxs",
            children: f ? /* @__PURE__ */ e.jsx(u, { className: "size-3.5" }) : /* @__PURE__ */ e.jsx(N, { className: "size-3.5" })
          }
        )
      ]
    }
  );
};
export {
  T as Secret,
  k as SecretText
};
