// zudoku/lib/ui/Toggle.js
import { j as a } from "../jsx-runtime-BzflLqGi.js";
import * as e from "@radix-ui/react-toggle";
import { c as s } from "../index-DI5SPFK9.js";
import * as d from "react";
import { c as f } from "../cn-5-Gd1Dss.js";
const m = s(
  "inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors hover:bg-muted hover:text-muted-foreground focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 data-[state=on]:bg-accent data-[state=on]:text-accent-foreground [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 gap-2",
  {
    variants: {
      variant: {
        default: "bg-transparent",
        outline: "border border-input bg-transparent hover:bg-accent hover:text-accent-foreground"
      },
      size: {
        default: "h-10 px-3 min-w-10",
        sm: "h-9 px-2.5 min-w-9",
        lg: "h-11 px-5 min-w-11"
      }
    },
    defaultVariants: {
      variant: "default",
      size: "default"
    }
  }
), c = d.forwardRef(({ className: t, variant: o, size: i, ...n }, r) => /* @__PURE__ */ a.jsx(
  e.Root,
  {
    ref: r,
    className: f(m({ variant: o, size: i, className: t })),
    ...n
  }
));
c.displayName = e.Root.displayName;
export {
  c as Toggle,
  m as toggleVariants
};
