// zudoku/lib/ui/Tooltip.js
import { j as o } from "../jsx-runtime-BzflLqGi.js";
import * as r from "@radix-ui/react-tooltip";
import { TooltipArrow as u } from "@radix-ui/react-tooltip";
import { cn as n } from "./util.js";
function d({
  delayDuration: t = 0,
  ...i
}) {
  return /* @__PURE__ */ o.jsx(
    r.Provider,
    {
      "data-slot": "tooltip-provider",
      delayDuration: t,
      ...i
    }
  );
}
function p({
  ...t
}) {
  return /* @__PURE__ */ o.jsx(d, { children: /* @__PURE__ */ o.jsx(r.Root, { "data-slot": "tooltip", ...t }) });
}
function m({
  ...t
}) {
  return /* @__PURE__ */ o.jsx(r.Trigger, { "data-slot": "tooltip-trigger", ...t });
}
function f({
  className: t,
  sideOffset: i = 0,
  children: e,
  ...a
}) {
  return /* @__PURE__ */ o.jsx(r.Portal, { children: /* @__PURE__ */ o.jsxs(
    r.Content,
    {
      "data-slot": "tooltip-content",
      sideOffset: i,
      className: n(
        "bg-foreground text-background animate-in fade-in-0 zoom-in-95 data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=closed]:zoom-out-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 z-50 w-fit origin-(--radix-tooltip-content-transform-origin) rounded-md px-3 py-1.5 text-xs text-balance",
        t
      ),
      ...a,
      children: [
        e,
        /* @__PURE__ */ o.jsx(r.Arrow, { className: "bg-foreground fill-foreground z-50 size-2.5 translate-y-[calc(-50%-2px)] rotate-45 rounded-[2px]" })
      ]
    }
  ) });
}
export {
  p as Tooltip,
  u as TooltipArrow,
  f as TooltipContent,
  d as TooltipProvider,
  m as TooltipTrigger
};
