// zudoku/lib/ui/Value.js
import { j as o } from "../jsx-runtime-BzflLqGi.js";
import { CheckIcon as n, CopyIcon as c } from "lucide-react";
import { c as m } from "../cn-5-Gd1Dss.js";
import { u as l } from "../useCopyToClipboard-B_085nfO.js";
import { Button as a } from "./Button.js";
const h = ({
  value: s,
  className: e,
  onCopy: i
}) => {
  const [r, t] = l();
  return /* @__PURE__ */ o.jsxs(
    "div",
    {
      className: m(
        "flex gap-2 items-center text-sm border rounded-md px-1",
        e
      ),
      children: [
        /* @__PURE__ */ o.jsx("div", { className: "font-mono w-full h-9 items-center flex px-2 text-xs gap-2", children: /* @__PURE__ */ o.jsx("span", { className: "w-full truncate", children: /* @__PURE__ */ o.jsx("div", { className: "w-40 inline-block md:w-fit", children: s }) }) }),
        /* @__PURE__ */ o.jsx(
          a,
          {
            variant: "ghost",
            onClick: () => {
              t(s), i?.(s);
            },
            size: "icon",
            children: r ? /* @__PURE__ */ o.jsx(n, { size: 16 }) : /* @__PURE__ */ o.jsx(c, { size: 16 })
          }
        )
      ]
    }
  );
};
export {
  h as Value
};
