// zudoku/dist/lib/plugins/search-pagefind/PagefindSearch.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { VisuallyHidden } from "@radix-ui/react-visually-hidden";
import { keepPreviousData, useQuery } from "@tanstack/react-query";
import { ListPlusIcon } from "lucide-react";
import { lazy, useEffect, useRef, useState } from "react";
import { Button } from "zudoku/ui/Button.js";
import { CommandDialog, CommandEmpty, CommandInput, } from "zudoku/ui/Command.js";
import { DialogTitle } from "zudoku/ui/Dialog.js";
import { Kbd, KbdGroup } from "zudoku/ui/Kbd.js";
import { useAuthState } from "../../authentication/state.js";
import { useZudoku } from "../../components/context/ZudokuContext.js";
import { SEARCH_PROTECTED_SECTION } from "../../core/RouteGuard.js";
import { joinUrl } from "../../util/joinUrl.js";
import { getResults } from "./get-results.js";
import { ResultList } from "./ResultList.js";
const IndexingDialog = lazy(() => import("./IndexingDialog.js"));
const DEFAULT_RANKING = {
    // Slightly lower than default because API docs tend to have repetitive terms (parameter names, HTTP methods, etc.)
    termFrequency: 0.8,
    // Lower than default because API documentation pages tend to be longer due to comprehensive endpoint documentation
    pageLength: 0.6,
    // Slightly higher than default because in technical documentation, exact matches should be prioritized
    termSimilarity: 1.2,
    // Slightly lower than default because API docs might have legitimate repetition of terms
    termSaturation: 1.2,
};
const importPagefind = (basePath) => import.meta.env.DEV
    ? // @ts-expect-error TypeScript can't resolve the import
        import(/* @vite-ignore */ "/pagefind/pagefind.js")
    : import(/* @vite-ignore */ joinUrl(basePath, "/pagefind/pagefind.js"));
const usePagefind = (options) => {
    const { options: { basePath }, } = useZudoku();
    const { data: pagefind, ...result } = useQuery({
        queryKey: ["pagefind", options.ranking],
        retry: false,
        queryFn: async () => {
            const pagefind = await importPagefind(basePath);
            await pagefind.init();
            await pagefind.options({
                ranking: {
                    termFrequency: options.ranking?.termFrequency ?? DEFAULT_RANKING.termFrequency,
                    pageLength: options.ranking?.pageLength ?? DEFAULT_RANKING.pageLength,
                    termSimilarity: options.ranking?.termSimilarity ?? DEFAULT_RANKING.termSimilarity,
                    termSaturation: options.ranking?.termSaturation ?? DEFAULT_RANKING.termSaturation,
                },
            });
            return pagefind;
        },
        enabled: typeof window !== "undefined",
    });
    if (result.isError && result.error.message !== "NOT_BUILT_YET") {
        // biome-ignore lint/suspicious/noConsole: Logging allowed here
        console.error(result.error);
    }
    return { ...result, pagefind };
};
export const PagefindSearch = ({ isOpen, onClose, options, }) => {
    const { pagefind, error, isError } = usePagefind(options);
    const [searchTerm, setSearchTerm] = useState("");
    const [selectedValue, setSelectedValue] = useState("");
    const auth = useAuthState();
    const context = useZudoku();
    const inputRef = useRef(null);
    const { data: searchResults } = useQuery({
        queryKey: ["pagefind-search", searchTerm, auth.isAuthenticated],
        queryFn: async () => {
            const filters = auth.isAuthenticated
                ? undefined
                : { not: { section: SEARCH_PROTECTED_SECTION } };
            const search = await pagefind?.search(searchTerm, { filters });
            if (!search)
                return [];
            return getResults({ search, options, auth, context });
        },
        placeholderData: keepPreviousData,
        enabled: !!pagefind && !!searchTerm,
    });
    useEffect(() => {
        if (!searchResults?.length)
            return;
        const firstResult = searchResults.at(0);
        if (!firstResult)
            return;
        const firstValue = `${firstResult.meta.title}-${firstResult.url}`;
        setSelectedValue(firstValue);
    }, [searchResults]);
    return (_jsxs(CommandDialog, { command: {
            shouldFilter: false,
            loop: true,
            value: selectedValue,
            onValueChange: setSelectedValue,
        }, content: { className: "max-w-[750px]" }, open: isOpen, onOpenChange: onClose, children: [_jsx(VisuallyHidden, { children: _jsx(DialogTitle, { children: "Search" }) }), _jsx(CommandInput, { ref: inputRef, placeholder: "Search...", value: searchTerm, onValueChange: setSearchTerm, disabled: isError }), _jsx(CommandEmpty, { children: searchTerm ? (_jsxs("div", { className: "flex flex-col items-center", children: ["No results found.", _jsx(Button, { variant: "link", onClick: () => {
                                setSearchTerm("");
                                inputRef.current?.focus();
                            }, children: "Clear search" })] })) : ("Start typing to search") }), isError && error.message !== "NOT_BUILT_YET" ? (_jsx("div", { className: "p-4 text-sm", children: "An error occurred while loading search." })) : (_jsxs(_Fragment, { children: [_jsx(ResultList, { basePath: context.options.basePath, searchResults: searchResults ?? [], searchTerm: searchTerm, onClose: onClose, maxSubResults: options.maxSubResults }), _jsxs("div", { className: "flex justify-between p-2 items-center", children: [_jsxs("div", { className: "flex items-center text-xs text-muted-foreground", children: [_jsxs(KbdGroup, { className: "ms-0 me-1", children: [_jsx(Kbd, { children: "\u2191" }), _jsx(Kbd, { children: "\u2193" })] }), "Navigate", _jsx(KbdGroup, { className: "ms-4 me-1", children: _jsx(Kbd, { children: "\u21B5" }) }), "Select", _jsx(KbdGroup, { className: "ms-4 me-1", children: _jsx(Kbd, { children: "Esc" }) }), "Close dialog"] }), import.meta.env.DEV && (_jsx(IndexingDialog, { children: _jsxs(Button, { variant: "outline", className: "h-7 text-xs text-muted-foreground", children: [_jsx(ListPlusIcon, {}), "Build Search Index"] }) }))] })] }))] }));
};
