// zudoku/dist/lib/plugins/api-keys/CreateApiKey.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useMutation, useQueryClient } from "@tanstack/react-query";
import { useForm } from "react-hook-form";
import { useNavigate } from "react-router";
import { ActionButton } from "zudoku/ui/ActionButton.js";
import { Alert, AlertDescription, AlertTitle } from "zudoku/ui/Alert.js";
import { DialogClose, DialogFooter } from "zudoku/ui/Dialog.js";
import { Select, SelectContent, SelectGroup, SelectItem, SelectTrigger, SelectValue, } from "zudoku/ui/Select.js";
import { useZudoku } from "../../components/context/ZudokuContext.js";
import { useAuth } from "../../hooks/index.js";
import { Button } from "../../ui/Button.js";
import { Input } from "../../ui/Input.js";
export const CreateApiKey = ({ service, onOpenChange, }) => {
    const context = useZudoku();
    const queryClient = useQueryClient();
    const navigate = useNavigate();
    const form = useForm({
        defaultValues: {
            expiresOn: "30",
        },
    });
    const auth = useAuth();
    const createKeyMutation = useMutation({
        mutationFn: ({ description, expiresOn }) => {
            if (!service.createKey) {
                throw new Error("createKey not implemented");
            }
            const expiresOnDate = expiresOn !== "never" ? addDaysToDate(Number(expiresOn)) : undefined;
            return service.createKey({
                apiKey: {
                    description: description || "Secret Key",
                    expiresOn: expiresOnDate,
                },
                context,
                auth,
            });
        },
        onSuccess: async () => {
            await queryClient.invalidateQueries({ queryKey: ["api-keys"] });
            await navigate("/settings/api-keys/");
        },
    });
    if (!service.createKey) {
        return null;
    }
    return (_jsxs("form", { onSubmit: form.handleSubmit((data) => createKeyMutation.mutate({ ...data }, {
            onSuccess: () => onOpenChange(false),
        })), children: [createKeyMutation.error && (_jsxs(Alert, { variant: "destructive", className: "mb-4", children: [_jsx(AlertTitle, { children: "Error" }), _jsx(AlertDescription, { children: createKeyMutation.error.message })] })), _jsxs("div", { className: "flex gap-2 flex-col text-sm font-medium", children: ["Name", _jsx(Input, { ...form.register("description") }), "Expiration", _jsxs(Select, { onValueChange: (value) => form.setValue("expiresOn", value), defaultValue: form.getValues("expiresOn"), children: [_jsx(SelectTrigger, { children: _jsx(SelectValue, {}) }), _jsx(SelectContent, { children: _jsxs(SelectGroup, { children: [[7, 30, 60, 90].map((option) => (_jsxs(SelectItem, { value: String(option), children: [option, " days"] }, option))), _jsx(SelectItem, { value: "never", children: "Never" })] }) })] }), _jsxs(DialogFooter, { children: [_jsx(DialogClose, { asChild: true, children: _jsx(Button, { variant: "outline", children: "Cancel" }) }), _jsx(ActionButton, { isPending: createKeyMutation.isPending, children: "Generate Key" })] })] })] }));
};
const addDaysToDate = (days) => {
    const date = new Date();
    date.setDate(date.getDate() + days);
    return date.toISOString();
};
