// zudoku/dist/lib/components/Banner.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { CircleXIcon } from "lucide-react";
import { useState } from "react";
import { cn } from "../util/cn.js";
import { useZudoku } from "./context/ZudokuContext.js";
const COLOR_MAP = {
    info: "bg-blue-500",
    note: "bg-gray-500",
    tip: "bg-green-600",
    caution: "bg-orange-500",
    danger: "bg-rose-500",
};
export const Banner = () => {
    const { options } = useZudoku();
    const { site } = options;
    const [isBannerOpen, setIsBannerOpen] = useState(true);
    if (!site?.banner || !isBannerOpen) {
        return _jsx("style", { children: `:root { --banner-height: 0px; }` });
    }
    const mappedColor = site.banner.color && site.banner.color in COLOR_MAP
        ? COLOR_MAP[site.banner.color]
        : !site.banner.color
            ? "bg-primary"
            : undefined;
    const style = !mappedColor ? { backgroundColor: site.banner.color } : {};
    return (_jsxs("div", { className: cn("relative text-primary-foreground text-sm font-medium px-4 py-2 flex gap-2 items-center lg:h-(--banner-height)", mappedColor), style: style, children: [_jsx("div", { className: "w-full", children: site.banner.message }), site.banner.dismissible && (_jsx("button", { type: "button", className: "md:absolute md:end-4 -m-1.5 p-1.5 hover:bg-accent-foreground/10 rounded-md", onClick: () => setIsBannerOpen(false), children: _jsx(CircleXIcon, { size: 16 }) }))] }));
};
