// zudoku/dist/lib/components/Framed.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { cva } from "class-variance-authority";
import { cn } from "../util/cn.js";
const frameVariants = cva("", {
    variants: {
        size: {
            sm: "max-w-sm",
            md: "max-w-md",
            lg: "max-w-lg",
            xl: "max-w-xl",
            "2xl": "max-w-2xl",
            full: "max-w-full",
        },
        align: {
            center: "mx-auto",
            start: "me-auto",
            end: "ms-auto",
        },
    },
    defaultVariants: {
        size: "xl",
        align: "center",
    },
});
export const Framed = ({ caption, className, size = "xl", align = "center", children, }) => (_jsxs("figure", { className: cn(frameVariants({ size, align }), className), children: [_jsx("div", { className: "relative overflow-hidden rounded-lg border border-border bg-muted/50 p-1 shadow-sm", children: _jsx("div", { className: "not-prose rounded-md [&_img]:rounded-md! relative overflow-hidden bg-background", children: children }) }), caption && (_jsx("figcaption", { className: "mt-3 text-center text-sm text-muted-foreground", children: caption }))] }));
