// zudoku/dist/lib/components/Mermaid.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useQuery } from "@tanstack/react-query";
import { useId } from "react";
import { Alert, AlertDescription, AlertTitle } from "zudoku/ui/Alert.js";
import { useTheme } from "../hooks/index.js";
import { Spinner } from "./Spinner.js";
let mermaidPromise = null;
const loadMermaid = () => {
    if (!mermaidPromise) {
        mermaidPromise = import("mermaid")
            .then((mod) => mod.default)
            .catch((error) => {
            throw new Error("Mermaid is not installed. Please install it with: npm install mermaid", { cause: error });
        });
    }
    return mermaidPromise;
};
export const Mermaid = ({ chart, config, ...props }) => {
    const id = useId();
    const theme = useTheme();
    const { data: svg, error, isPending, } = useQuery({
        queryKey: ["mermaid", chart, config, theme.resolvedTheme],
        queryFn: async () => {
            const mermaid = await loadMermaid();
            mermaid.initialize({
                theme: theme.resolvedTheme === "dark" ? "dark" : "base",
                ...config,
            });
            const { svg } = await mermaid.render(id, chart);
            return svg;
        },
        enabled: typeof window !== "undefined",
        retry: false,
    });
    if (error)
        return (_jsxs(Alert, { className: "flex flex-col gap-2", variant: "destructive", children: [_jsx(AlertTitle, { children: "Mermaid Error" }), _jsx(AlertDescription, { className: "overflow-auto wrap-break-word whitespace-pre-wrap font-mono text-xs", children: error.message })] }));
    if (isPending)
        return _jsx(Spinner, {});
    // biome-ignore lint/security/noDangerouslySetInnerHtml: Mermaid generates valid SVG
    return _jsx("div", { ...props, dangerouslySetInnerHTML: { __html: svg } });
};
