// zudoku/dist/lib/authentication/state.js
import { create } from "zustand";
import { createJSONStorage, persist } from "zustand/middleware";
import { syncZustandState } from "../util/syncZustandState.js";
export const useAuthState = create()(persist((set) => ({
    isAuthenticated: false,
    isPending: true,
    profile: null,
    providerData: null,
    setAuthenticationPending: () => set(() => ({
        isAuthenticated: false,
        isPending: false,
        profile: null,
        providerData: null,
    })),
    setLoggedOut: () => set(() => ({
        isAuthenticated: false,
        isPending: false,
        profile: null,
        providerData: null,
    })),
    setLoggedIn: ({ profile, providerData, }) => set(() => ({
        isAuthenticated: true,
        isPending: false,
        profile,
        providerData,
    })),
}), {
    merge: (persistedState, currentState) => {
        return {
            ...currentState,
            isPending: false,
            ...(typeof persistedState === "object" ? persistedState : {}),
        };
    },
    name: "auth-state",
    storage: createJSONStorage(() => localStorage),
}));
syncZustandState(useAuthState);
