// zudoku/dist/lib/ui/Frame.js
import { jsx as _jsx } from "react/jsx-runtime";
import { cn } from "../util/cn.js";
function Frame({ className, ...props }) {
    return (_jsx("div", { "data-slot": "frame", className: cn("relative flex flex-col rounded-2xl bg-muted p-1", className), ...props }));
}
function FramePanel({ className, ...props }) {
    return (_jsx("div", { "data-slot": "frame-panel", className: cn("relative bg-clip-padding rounded-xl border bg-card p-5 shadow-xs", "before:pointer-events-none before:absolute before:inset-0 before:rounded-[calc(var(--radius-xl)-1px)] before:shadow-[0_1px_--theme(--color-black/4%)] dark:bg-clip-border dark:before:shadow-[0_-1px_--theme(--color-white/8%)]", className), ...props }));
}
function FrameHeader({ className, ...props }) {
    return (_jsx("header", { "data-slot": "frame-panel-header", className: cn("flex flex-col p-4", className), ...props }));
}
function FrameTitle({ className, ...props }) {
    return (_jsx("div", { "data-slot": "frame-panel-title", className: cn("text-sm font-semibold", className), ...props }));
}
function FrameDescription({ className, ...props }) {
    return (_jsx("div", { "data-slot": "frame-panel-description", className: cn("text-sm text-muted-foreground", className), ...props }));
}
function FrameFooter({ className, ...props }) {
    return (_jsx("footer", { "data-slot": "frame-panel-footer", className: cn("flex flex-col gap-1 px-5 py-4", className), ...props }));
}
export { Frame, FramePanel, FrameHeader, FrameTitle, FrameDescription, FrameFooter, };
