// zudoku/lib/Button-qpSb3K7a.js
import { j as a } from "./jsx-runtime-BzflLqGi.js";
import { Slot as d } from "@radix-ui/react-slot";
import { c } from "./index-DI5SPFK9.js";
import * as u from "react";
import { c as g } from "./cn-5-Gd1Dss.js";
const v = c(
  "not-prose inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive",
  {
    variants: {
      variant: {
        default: "bg-primary text-primary-foreground hover:bg-primary/90",
        destructive: "bg-destructive text-white hover:bg-destructive/90 focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60",
        outline: "border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50",
        secondary: "bg-secondary text-secondary-foreground hover:bg-secondary/80",
        ghost: "hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50",
        link: "text-primary underline-offset-4 hover:underline",
        expand: "flex gap-1.5 border bg-transparent rounded-xl text-muted-foreground hover:text-foreground"
      },
      size: {
        default: "h-9 px-4 py-2",
        sm: "h-8 rounded-md px-3 text-xs",
        lg: "h-10 rounded-md px-8",
        xl: "h-14 rounded-lg px-10 text-lg",
        icon: "size-9",
        "icon-xxs": "size-6",
        "icon-xs": "size-7",
        "icon-sm": "size-8",
        "icon-lg": "size-10"
      }
    },
    defaultVariants: {
      variant: "default",
      size: "default"
    }
  }
), l = u.forwardRef(
  ({ className: r, variant: t, size: i, asChild: e = !1, ...o }, n) => {
    const s = e ? d : "button";
    return /* @__PURE__ */ a.jsx(
      s,
      {
        type: e ? void 0 : "button",
        className: g(v({ variant: t, size: i, className: r })),
        ref: n,
        ...o
      }
    );
  }
);
l.displayName = "Button";
export {
  l as B
};
