// zudoku/lib/DropdownMenu-DN0jNrjj.js
import { j as e } from "./jsx-runtime-BzflLqGi.js";
import * as o from "@radix-ui/react-dropdown-menu";
import "lucide-react";
import { c as r } from "./cn-5-Gd1Dss.js";
function m({
  ...t
}) {
  return /* @__PURE__ */ e.jsx(o.Root, { "data-slot": "dropdown-menu", ...t });
}
function p({
  ...t
}) {
  return /* @__PURE__ */ e.jsx(
    o.Trigger,
    {
      "data-slot": "dropdown-menu-trigger",
      ...t
    }
  );
}
function c({
  className: t,
  sideOffset: a = 4,
  ...n
}) {
  return /* @__PURE__ */ e.jsx(o.Portal, { children: /* @__PURE__ */ e.jsx(
    o.Content,
    {
      "data-slot": "dropdown-menu-content",
      sideOffset: a,
      className: r(
        "bg-popover text-popover-foreground data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 z-50 max-h-(--radix-dropdown-menu-content-available-height) min-w-[8rem] origin-(--radix-dropdown-menu-content-transform-origin) overflow-x-hidden overflow-y-auto rounded-md border p-1 shadow-md",
        t
      ),
      ...n
    }
  ) });
}
function l({
  ...t
}) {
  return /* @__PURE__ */ e.jsx(o.Group, { "data-slot": "dropdown-menu-group", ...t });
}
function f({
  className: t,
  inset: a,
  variant: n = "default",
  ...d
}) {
  return /* @__PURE__ */ e.jsx(
    o.Item,
    {
      "data-slot": "dropdown-menu-item",
      "data-inset": a,
      "data-variant": n,
      className: r(
        "focus:bg-accent focus:text-accent-foreground data-[variant=destructive]:text-destructive data-[variant=destructive]:focus:bg-destructive/10 dark:data-[variant=destructive]:focus:bg-destructive/20 data-[variant=destructive]:focus:text-destructive data-[variant=destructive]:*:[svg]:!text-destructive [&_svg:not([class*='text-'])]:text-muted-foreground relative flex cursor-default items-center gap-2 rounded-sm px-2 py-1.5 text-sm outline-hidden select-none data-[disabled]:pointer-events-none data-[disabled]:opacity-50 data-[inset]:pl-8 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4",
        t
      ),
      ...d
    }
  );
}
function v({
  className: t,
  inset: a,
  ...n
}) {
  return /* @__PURE__ */ e.jsx(
    o.Label,
    {
      "data-slot": "dropdown-menu-label",
      "data-inset": a,
      className: r(
        "px-2 py-1.5 text-sm font-medium data-[inset]:pl-8",
        t
      ),
      ...n
    }
  );
}
function x({
  className: t,
  ...a
}) {
  return /* @__PURE__ */ e.jsx(
    o.Separator,
    {
      "data-slot": "dropdown-menu-separator",
      className: r("bg-border -mx-1 my-1 h-px", t),
      ...a
    }
  );
}
export {
  m as D,
  p as a,
  c as b,
  f as c,
  v as d,
  x as e,
  l as f
};
