// zudoku/lib/Frame-DKlOmSkU.js
import { j as a } from "./jsx-runtime-BzflLqGi.js";
import { Slot as l } from "@radix-ui/react-slot";
import { c as m } from "./index-DI5SPFK9.js";
import { c as r } from "./cn-5-Gd1Dss.js";
import { Separator as d } from "./ui/Separator.js";
function g({ className: t, ...e }) {
  return /* @__PURE__ */ a.jsx(
    "div",
    {
      role: "list",
      "data-slot": "item-group",
      className: r("group/item-group flex flex-col", t),
      ...e
    }
  );
}
function v({
  className: t,
  ...e
}) {
  return /* @__PURE__ */ a.jsx(
    d,
    {
      "data-slot": "item-separator",
      orientation: "horizontal",
      className: r("my-0", t),
      ...e
    }
  );
}
const f = m(
  "group/item flex items-center border border-transparent text-sm rounded-md transition-colors [a]:hover:bg-accent/50 [a]:transition-colors duration-100 flex-wrap outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px]",
  {
    variants: {
      variant: {
        default: "bg-transparent",
        outline: "border-border",
        muted: "bg-muted/50"
      },
      size: {
        default: "p-4 gap-4 ",
        sm: "py-3 px-4 gap-2.5"
      }
    },
    defaultVariants: {
      variant: "default",
      size: "default"
    }
  }
);
function j({
  className: t,
  variant: e = "default",
  size: o = "default",
  asChild: s = !1,
  ...n
}) {
  const i = s ? l : "div";
  return /* @__PURE__ */ a.jsx(
    i,
    {
      "data-slot": "item",
      "data-variant": e,
      "data-size": o,
      className: r(f({ variant: e, size: o, className: t })),
      ...n
    }
  );
}
function N({ className: t, ...e }) {
  return /* @__PURE__ */ a.jsx(
    "div",
    {
      "data-slot": "item-content",
      className: r(
        "flex flex-1 flex-col gap-1 [&+[data-slot=item-content]]:flex-none",
        t
      ),
      ...e
    }
  );
}
function h({ className: t, ...e }) {
  return /* @__PURE__ */ a.jsx(
    "div",
    {
      "data-slot": "item-title",
      className: r(
        "flex w-fit items-center gap-2 text-sm leading-snug font-medium",
        t
      ),
      ...e
    }
  );
}
function F({ className: t, ...e }) {
  return /* @__PURE__ */ a.jsx(
    "p",
    {
      "data-slot": "item-description",
      className: r(
        "text-muted-foreground line-clamp-2 text-sm leading-normal font-normal text-balance",
        "[&>a:hover]:text-primary [&>a]:underline [&>a]:underline-offset-4",
        t
      ),
      ...e
    }
  );
}
function I({ className: t, ...e }) {
  return /* @__PURE__ */ a.jsx(
    "div",
    {
      "data-slot": "item-actions",
      className: r("flex items-center gap-2", t),
      ...e
    }
  );
}
function w({ className: t, ...e }) {
  return /* @__PURE__ */ a.jsx(
    "div",
    {
      "data-slot": "frame",
      className: r(
        "relative flex flex-col rounded-2xl bg-muted p-1",
        t
      ),
      ...e
    }
  );
}
function k({ className: t, ...e }) {
  return /* @__PURE__ */ a.jsx(
    "div",
    {
      "data-slot": "frame-panel",
      className: r(
        "relative bg-clip-padding rounded-xl border bg-card p-5 shadow-xs",
        "before:pointer-events-none before:absolute before:inset-0 before:rounded-[calc(var(--radius-xl)-1px)] before:shadow-[0_1px_--theme(--color-black/4%)] dark:bg-clip-border dark:before:shadow-[0_-1px_--theme(--color-white/8%)]",
        t
      ),
      ...e
    }
  );
}
function y({ className: t, ...e }) {
  return /* @__PURE__ */ a.jsx(
    "header",
    {
      "data-slot": "frame-panel-header",
      className: r("flex flex-col p-4", t),
      ...e
    }
  );
}
function _({ className: t, ...e }) {
  return /* @__PURE__ */ a.jsx(
    "div",
    {
      "data-slot": "frame-panel-title",
      className: r("text-sm font-semibold", t),
      ...e
    }
  );
}
function z({
  className: t,
  ...e
}) {
  return /* @__PURE__ */ a.jsx(
    "div",
    {
      "data-slot": "frame-panel-description",
      className: r("text-sm text-muted-foreground", t),
      ...e
    }
  );
}
function S({ className: t, ...e }) {
  return /* @__PURE__ */ a.jsx(
    "footer",
    {
      "data-slot": "frame-panel-footer",
      className: r("flex flex-col gap-1 px-5 py-4", t),
      ...e
    }
  );
}
export {
  w as F,
  j as I,
  N as a,
  h as b,
  I as c,
  k as d,
  v as e,
  g as f,
  y as g,
  z as h,
  S as i,
  _ as j,
  F as k
};
