// zudoku/lib/hook-BqOwTUgE.js
import { u as f } from "./useQuery-DSKGiCQr.js";
import { u as s } from "./chunk-EPOLDU6W-CLHrVxOG.js";
import { a as n } from "./ZudokuContext-Mfno-z8f.js";
import { u as c } from "./ZudokuReactContext-DGJAP1sN.js";
const u = ({
  refetchOnWindowFocus: e
} = {}) => {
  const { authentication: t } = n();
  return f({
    refetchOnWindowFocus: e,
    queryKey: ["refresh-user-profile"],
    enabled: typeof t < "u" && typeof t?.refreshUserProfile == "function",
    queryFn: () => t?.refreshUserProfile?.()
  });
}, w = () => {
  const { authentication: e } = n(), t = c(), r = s(), o = typeof e < "u", { refetch: i } = u({
    refetchOnWindowFocus: "always"
  });
  return {
    email: t.profile?.email,
    isVerified: t.profile?.emailVerified,
    supportsEmailVerification: typeof e?.requestEmailVerification == "function",
    refresh: () => void i(),
    requestEmailVerification: async (a) => {
      if (process.env.NODE_ENV === "development" && console.warn(
        "requestEmailVerification is deprecated. Use useVerifiedEmail."
      ), !o)
        throw new Error("Authentication is not enabled.");
      await e.requestEmailVerification?.(
        { navigate: r },
        {
          ...a,
          redirectTo: a?.redirectTo ?? window.location.href
        }
      );
    }
  };
}, E = () => {
  const { authentication: e } = n(), t = c(), r = typeof e < "u", o = s();
  return u(), {
    isAuthEnabled: r,
    ...t,
    login: async (i) => {
      if (!r)
        throw new Error("Authentication is not enabled.");
      await e.signIn(
        { navigate: o },
        {
          ...i,
          redirectTo: i?.redirectTo ?? window.location.href
        }
      );
    },
    logout: async () => {
      if (!r)
        throw new Error("Authentication is not enabled.");
      await e.signOut({ navigate: o }), o("/", { replace: !0 });
    },
    signup: async (i) => {
      if (!r)
        throw new Error("Authentication is not enabled.");
      await e.signUp(
        { navigate: o },
        {
          ...i,
          redirectTo: i?.redirectTo ?? window.location.href
        }
      );
    },
    requestEmailVerification: async (i) => {
      if (process.env.NODE_ENV === "development" && console.warn(
        "requestEmailVerification is deprecated on useAuth. Use the requestEmailVerification method from useVerifiedEmail hook instead."
      ), !r)
        throw new Error("Authentication is not enabled.");
      await e.requestEmailVerification?.(
        { navigate: o },
        {
          ...i,
          redirectTo: i?.redirectTo ?? window.location.href
        }
      );
    }
  };
};
export {
  u as a,
  w as b,
  E as u
};
