// zudoku/lib/index-Ckdk300F.js
import { j as t } from "./jsx-runtime-BzflLqGi.js";
import { ChevronsDownUpIcon as kn, ChevronsUpDownIcon as En, PlusCircleIcon as Je, XIcon as It, CheckIcon as Dt, PaperclipIcon as me, TrashIcon as Rn, FileInput as In, ScanTextIcon as yt, Grid2x2PlusIcon as vt, ChevronDownIcon as Dn, TableOfContentsIcon as Fn, LockIcon as zn, CircleAlertIcon as Vn, Circle as Bn, Unlink2Icon as _n, DownloadIcon as Ft, CornerDownRightIcon as qn, MinusCircleIcon as bt, CornerDownLeftIcon as Un, SquareCodeIcon as Mn, EyeOffIcon as Gn, EyeIcon as Hn, UnplugIcon as Qn, SendIcon as Jn, CopyIcon as Wn, IdCardLanyardIcon as Xn, ShapesIcon as Kn, PlayIcon as Zn, CirclePlayIcon as Yn } from "lucide-react";
import { r as es, e as ts, a as ns, N as ss, aM as rs, m as jt } from "./chunk-EPOLDU6W-CLHrVxOG.js";
import { Button as K } from "./ui/Button.js";
import { e as as, j as V } from "./ZudokuContext-Mfno-z8f.js";
import { Z as os } from "./invariant-B_t_F2s_.js";
import * as xe from "react";
import { createContext as zt, use as is, useRef as E, useEffect as U, useMemo as Vt, useCallback as D, startTransition as ls, useState as k, Fragment as le, useLayoutEffect as cs, useTransition as ds, useContext as us } from "react";
import { VisuallyHidden as ms } from "@radix-ui/react-visually-hidden";
import { z as ps } from "./index-DAWHN3cH.js";
import { u as hs } from "./hook-BqOwTUgE.js";
import { c as Ze, p as Ye, a as Bt, s as fs } from "./ZudokuReactContext-DGJAP1sN.js";
import { B } from "./Button-qpSb3K7a.js";
import { D as et, a as tt, c as nt, d as _t, e as qt, g as gs } from "./Dialog-hlvmmQ_c.js";
import { j as Ct, k as Nt, l as wt, m as xs, g as ys } from "./Zudoku-BPvTuMy8.js";
import { u as vs } from "./useMutation-CgOLW9Kh.js";
import { u as oe, b as Ut, C as St, a as bs, F as js } from "./index.esm-BYObtETB.js";
import * as st from "@radix-ui/react-collapsible";
import { a as Mt, b as Gt, c as Ht, d as Qt, f as Ae, S as Cs, h as Ns } from "./Secret-9xKh4cip.js";
import { c as S } from "./cn-5-Gd1Dss.js";
import { u as ws } from "./useCopyToClipboard-B_085nfO.js";
import { u as Ss } from "./useLatest-hmRS46UF.js";
import { D as Jt, a as Wt, b as Xt, d as Ts, e as As, f as Os, c as Oe } from "./DropdownMenu-DN0jNrjj.js";
import { I as Ps } from "./Input-Cx-GeKoF.js";
import { c as ye } from "./createVariantComponent-Dc0vtOvr.js";
import * as Tt from "@radix-ui/react-checkbox";
import * as Pe from "@radix-ui/react-popover";
import { PopoverAnchor as $s } from "@radix-ui/react-popover";
import { P as Ls } from "./index-CrcNWbel.js";
import { f as ks, g as Es, C as Rs, b as Is } from "./Command-N6VujV30.js";
import { C as Ds } from "./Card-DCdq37aA.js";
import * as Kt from "@radix-ui/react-label";
import { c as Fs } from "./index-DI5SPFK9.js";
import * as he from "@radix-ui/react-radio-group";
import { S as zs } from "./Spinner-CI6bRyZw.js";
import { A as Vs, a as Bs, b as _s } from "./Mermaid-CJQJeZ9Q.js";
import { u as qs } from "./useQuery-DSKGiCQr.js";
import { u as Us } from "./useSuspenseQuery-DbPM3Wv6.js";
let _e;
const Ms = (e) => {
  if (e.errors?.[0])
    throw new os(e.errors[0].message, {
      developerHint: "Check your configuration value `apis.type` and `apis.input` in the Zudoku config."
    });
};
class Gs {
  constructor(n) {
    this.config = n;
  }
  #e = async () => (_e || (_e = import("./createServer-D7FuO_aJ.js").then(
    (n) => n.createServer(this.config)
  )), _e);
  #t = async (n) => this.config.server ? fetch(this.config.server, n) : (await this.#e()).fetch("http://localhost/graphql", n);
  fetch = async (n, s) => {
    const a = n.match(/query (\w+)/)?.[1], r = await this.#t({
      method: "POST",
      body: JSON.stringify({ query: n, variables: s, operationName: a }),
      headers: { "Content-Type": "application/json" }
    });
    if (!r.ok)
      throw new Error("Network response was not ok");
    const o = await r.json();
    return Ms(o), o.data;
  };
}
function qe(e, n) {
  if (!!!e)
    throw new Error(n);
}
function Hs(e) {
  return typeof e == "object" && e !== null;
}
function Qs(e, n) {
  if (!!!e)
    throw new Error(
      n ?? "Unexpected invariant triggered."
    );
}
const Js = /\r\n|[\n\r]/g;
function We(e, n) {
  let s = 0, a = 1;
  for (const r of e.body.matchAll(Js)) {
    if (typeof r.index == "number" || Qs(!1), r.index >= n)
      break;
    s = r.index + r[0].length, a += 1;
  }
  return {
    line: a,
    column: n + 1 - s
  };
}
function Ws(e) {
  return Zt(
    e.source,
    We(e.source, e.start)
  );
}
function Zt(e, n) {
  const s = e.locationOffset.column - 1, a = "".padStart(s) + e.body, r = n.line - 1, o = e.locationOffset.line - 1, i = n.line + o, c = n.line === 1 ? s : 0, l = n.column + c, m = `${e.name}:${i}:${l}
`, d = a.split(/\r\n|[\n\r]/g), f = d[r];
  if (f.length > 120) {
    const x = Math.floor(l / 80), p = l % 80, v = [];
    for (let u = 0; u < f.length; u += 80)
      v.push(f.slice(u, u + 80));
    return m + At([
      [`${i} |`, v[0]],
      ...v.slice(1, x + 1).map((u) => ["|", u]),
      ["|", "^".padStart(p)],
      ["|", v[x + 1]]
    ]);
  }
  return m + At([
    // Lines specified like this: ["prefix", "string"],
    [`${i - 1} |`, d[r - 1]],
    [`${i} |`, f],
    ["|", "^".padStart(l)],
    [`${i + 1} |`, d[r + 1]]
  ]);
}
function At(e) {
  const n = e.filter(([a, r]) => r !== void 0), s = Math.max(...n.map(([a]) => a.length));
  return n.map(([a, r]) => a.padStart(s) + (r ? " " + r : "")).join(`
`);
}
function Xs(e) {
  const n = e[0];
  return n == null || "kind" in n || "length" in n ? {
    nodes: n,
    source: e[1],
    positions: e[2],
    path: e[3],
    originalError: e[4],
    extensions: e[5]
  } : n;
}
class rt extends Error {
  /**
   * An array of `{ line, column }` locations within the source GraphQL document
   * which correspond to this error.
   *
   * Errors during validation often contain multiple locations, for example to
   * point out two things with the same name. Errors during execution include a
   * single location, the field which produced the error.
   *
   * Enumerable, and appears in the result of JSON.stringify().
   */
  /**
   * An array describing the JSON-path into the execution response which
   * corresponds to this error. Only included for errors during execution.
   *
   * Enumerable, and appears in the result of JSON.stringify().
   */
  /**
   * An array of GraphQL AST Nodes corresponding to this error.
   */
  /**
   * The source GraphQL document for the first location of this error.
   *
   * Note that if this Error represents more than one node, the source may not
   * represent nodes after the first node.
   */
  /**
   * An array of character offsets within the source GraphQL document
   * which correspond to this error.
   */
  /**
   * The original error thrown from a field resolver during execution.
   */
  /**
   * Extension fields to add to the formatted error.
   */
  /**
   * @deprecated Please use the `GraphQLErrorOptions` constructor overload instead.
   */
  constructor(n, ...s) {
    var a, r, o;
    const { nodes: i, source: c, positions: l, path: m, originalError: d, extensions: f } = Xs(s);
    super(n), this.name = "GraphQLError", this.path = m ?? void 0, this.originalError = d ?? void 0, this.nodes = Ot(
      Array.isArray(i) ? i : i ? [i] : void 0
    );
    const x = Ot(
      (a = this.nodes) === null || a === void 0 ? void 0 : a.map((v) => v.loc).filter((v) => v != null)
    );
    this.source = c ?? (x == null || (r = x[0]) === null || r === void 0 ? void 0 : r.source), this.positions = l ?? x?.map((v) => v.start), this.locations = l && c ? l.map((v) => We(c, v)) : x?.map((v) => We(v.source, v.start));
    const p = Hs(
      d?.extensions
    ) ? d?.extensions : void 0;
    this.extensions = (o = f ?? p) !== null && o !== void 0 ? o : /* @__PURE__ */ Object.create(null), Object.defineProperties(this, {
      message: {
        writable: !0,
        enumerable: !0
      },
      name: {
        enumerable: !1
      },
      nodes: {
        enumerable: !1
      },
      source: {
        enumerable: !1
      },
      positions: {
        enumerable: !1
      },
      originalError: {
        enumerable: !1
      }
    }), d != null && d.stack ? Object.defineProperty(this, "stack", {
      value: d.stack,
      writable: !0,
      configurable: !0
    }) : Error.captureStackTrace ? Error.captureStackTrace(this, rt) : Object.defineProperty(this, "stack", {
      value: Error().stack,
      writable: !0,
      configurable: !0
    });
  }
  get [Symbol.toStringTag]() {
    return "GraphQLError";
  }
  toString() {
    let n = this.message;
    if (this.nodes)
      for (const s of this.nodes)
        s.loc && (n += `

` + Ws(s.loc));
    else if (this.source && this.locations)
      for (const s of this.locations)
        n += `

` + Zt(this.source, s);
    return n;
  }
  toJSON() {
    const n = {
      message: this.message
    };
    return this.locations != null && (n.locations = this.locations), this.path != null && (n.path = this.path), this.extensions != null && Object.keys(this.extensions).length > 0 && (n.extensions = this.extensions), n;
  }
}
function Ot(e) {
  return e === void 0 || e.length === 0 ? void 0 : e;
}
function F(e, n, s) {
  return new rt(`Syntax Error: ${s}`, {
    source: e,
    positions: [n]
  });
}
class so {
  /**
   * The character offset at which this Node begins.
   */
  /**
   * The character offset at which this Node ends.
   */
  /**
   * The Token at which this Node begins.
   */
  /**
   * The Token at which this Node ends.
   */
  /**
   * The Source document the AST represents.
   */
  constructor(n, s, a) {
    this.start = n.start, this.end = s.end, this.startToken = n, this.endToken = s, this.source = a;
  }
  get [Symbol.toStringTag]() {
    return "Location";
  }
  toJSON() {
    return {
      start: this.start,
      end: this.end
    };
  }
}
class Yt {
  /**
   * The kind of Token.
   */
  /**
   * The character offset at which this Node begins.
   */
  /**
   * The character offset at which this Node ends.
   */
  /**
   * The 1-indexed line number on which this Token appears.
   */
  /**
   * The 1-indexed column number at which this Token begins.
   */
  /**
   * For non-punctuation tokens, represents the interpreted value of the token.
   *
   * Note: is undefined for punctuation tokens, but typed as string for
   * convenience in the parser.
   */
  /**
   * Tokens exist as nodes in a double-linked-list amongst all tokens
   * including ignored tokens. <SOF> is always the first node and <EOF>
   * the last.
   */
  constructor(n, s, a, r, o, i) {
    this.kind = n, this.start = s, this.end = a, this.line = r, this.column = o, this.value = i, this.prev = null, this.next = null;
  }
  get [Symbol.toStringTag]() {
    return "Token";
  }
  toJSON() {
    return {
      kind: this.kind,
      value: this.value,
      line: this.line,
      column: this.column
    };
  }
}
const Ks = {
  Name: [],
  Document: ["definitions"],
  OperationDefinition: [
    "description",
    "name",
    "variableDefinitions",
    "directives",
    "selectionSet"
  ],
  VariableDefinition: [
    "description",
    "variable",
    "type",
    "defaultValue",
    "directives"
  ],
  Variable: ["name"],
  SelectionSet: ["selections"],
  Field: ["alias", "name", "arguments", "directives", "selectionSet"],
  Argument: ["name", "value"],
  FragmentSpread: ["name", "directives"],
  InlineFragment: ["typeCondition", "directives", "selectionSet"],
  FragmentDefinition: [
    "description",
    "name",
    // Note: fragment variable definitions are deprecated and will removed in v17.0.0
    "variableDefinitions",
    "typeCondition",
    "directives",
    "selectionSet"
  ],
  IntValue: [],
  FloatValue: [],
  StringValue: [],
  BooleanValue: [],
  NullValue: [],
  EnumValue: [],
  ListValue: ["values"],
  ObjectValue: ["fields"],
  ObjectField: ["name", "value"],
  Directive: ["name", "arguments"],
  NamedType: ["name"],
  ListType: ["type"],
  NonNullType: ["type"],
  SchemaDefinition: ["description", "directives", "operationTypes"],
  OperationTypeDefinition: ["type"],
  ScalarTypeDefinition: ["description", "name", "directives"],
  ObjectTypeDefinition: [
    "description",
    "name",
    "interfaces",
    "directives",
    "fields"
  ],
  FieldDefinition: ["description", "name", "arguments", "type", "directives"],
  InputValueDefinition: [
    "description",
    "name",
    "type",
    "defaultValue",
    "directives"
  ],
  InterfaceTypeDefinition: [
    "description",
    "name",
    "interfaces",
    "directives",
    "fields"
  ],
  UnionTypeDefinition: ["description", "name", "directives", "types"],
  EnumTypeDefinition: ["description", "name", "directives", "values"],
  EnumValueDefinition: ["description", "name", "directives"],
  InputObjectTypeDefinition: ["description", "name", "directives", "fields"],
  DirectiveDefinition: ["description", "name", "arguments", "locations"],
  SchemaExtension: ["directives", "operationTypes"],
  ScalarTypeExtension: ["name", "directives"],
  ObjectTypeExtension: ["name", "interfaces", "directives", "fields"],
  InterfaceTypeExtension: ["name", "interfaces", "directives", "fields"],
  UnionTypeExtension: ["name", "directives", "types"],
  EnumTypeExtension: ["name", "directives", "values"],
  InputObjectTypeExtension: ["name", "directives", "fields"],
  TypeCoordinate: ["name"],
  MemberCoordinate: ["name", "memberName"],
  ArgumentCoordinate: ["name", "fieldName", "argumentName"],
  DirectiveCoordinate: ["name"],
  DirectiveArgumentCoordinate: ["name", "argumentName"]
}, Zs = new Set(Object.keys(Ks));
function ro(e) {
  const n = e?.kind;
  return typeof n == "string" && Zs.has(n);
}
var Pt;
(function(e) {
  e.QUERY = "query", e.MUTATION = "mutation", e.SUBSCRIPTION = "subscription";
})(Pt || (Pt = {}));
function Xe(e) {
  return e === 9 || e === 32;
}
function fe(e) {
  return e >= 48 && e <= 57;
}
function en(e) {
  return e >= 97 && e <= 122 || // A-Z
  e >= 65 && e <= 90;
}
function tn(e) {
  return en(e) || e === 95;
}
function Ys(e) {
  return en(e) || fe(e) || e === 95;
}
function er(e) {
  var n;
  let s = Number.MAX_SAFE_INTEGER, a = null, r = -1;
  for (let i = 0; i < e.length; ++i) {
    var o;
    const c = e[i], l = tr(c);
    l !== c.length && (a = (o = a) !== null && o !== void 0 ? o : i, r = i, i !== 0 && l < s && (s = l));
  }
  return e.map((i, c) => c === 0 ? i : i.slice(s)).slice(
    (n = a) !== null && n !== void 0 ? n : 0,
    r + 1
  );
}
function tr(e) {
  let n = 0;
  for (; n < e.length && Xe(e.charCodeAt(n)); )
    ++n;
  return n;
}
function nr(e, n) {
  const s = e.replace(/"""/g, '\\"""'), a = s.split(/\r\n|[\n\r]/g), r = a.length === 1, o = a.length > 1 && a.slice(1).every((p) => p.length === 0 || Xe(p.charCodeAt(0))), i = s.endsWith('\\"""'), c = e.endsWith('"') && !i, l = e.endsWith("\\"), m = c || l, d = !(n != null && n.minimize) && // add leading and trailing new lines only if it improves readability
  (!r || e.length > 70 || m || o || i);
  let f = "";
  const x = r && Xe(e.charCodeAt(0));
  return (d && !x || o) && (f += `
`), f += s, (d || m) && (f += `
`), '"""' + f + '"""';
}
var b;
(function(e) {
  e.SOF = "<SOF>", e.EOF = "<EOF>", e.BANG = "!", e.DOLLAR = "$", e.AMP = "&", e.PAREN_L = "(", e.PAREN_R = ")", e.DOT = ".", e.SPREAD = "...", e.COLON = ":", e.EQUALS = "=", e.AT = "@", e.BRACKET_L = "[", e.BRACKET_R = "]", e.BRACE_L = "{", e.PIPE = "|", e.BRACE_R = "}", e.NAME = "Name", e.INT = "Int", e.FLOAT = "Float", e.STRING = "String", e.BLOCK_STRING = "BlockString", e.COMMENT = "Comment";
})(b || (b = {}));
class sr {
  /**
   * The previously focused non-ignored token.
   */
  /**
   * The currently focused non-ignored token.
   */
  /**
   * The (1-indexed) line containing the current token.
   */
  /**
   * The character offset at which the current line begins.
   */
  constructor(n) {
    const s = new Yt(b.SOF, 0, 0, 0, 0);
    this.source = n, this.lastToken = s, this.token = s, this.line = 1, this.lineStart = 0;
  }
  get [Symbol.toStringTag]() {
    return "Lexer";
  }
  /**
   * Advances the token stream to the next non-ignored token.
   */
  advance() {
    return this.lastToken = this.token, this.token = this.lookahead();
  }
  /**
   * Looks ahead and returns the next non-ignored token, but does not change
   * the state of Lexer.
   */
  lookahead() {
    let n = this.token;
    if (n.kind !== b.EOF)
      do
        if (n.next)
          n = n.next;
        else {
          const s = ar(this, n.end);
          n.next = s, s.prev = n, n = s;
        }
      while (n.kind === b.COMMENT);
    return n;
  }
}
function rr(e) {
  return e === b.BANG || e === b.DOLLAR || e === b.AMP || e === b.PAREN_L || e === b.PAREN_R || e === b.DOT || e === b.SPREAD || e === b.COLON || e === b.EQUALS || e === b.AT || e === b.BRACKET_L || e === b.BRACKET_R || e === b.BRACE_L || e === b.PIPE || e === b.BRACE_R;
}
function ie(e) {
  return e >= 0 && e <= 55295 || e >= 57344 && e <= 1114111;
}
function Le(e, n) {
  return nn(e.charCodeAt(n)) && sn(e.charCodeAt(n + 1));
}
function nn(e) {
  return e >= 55296 && e <= 56319;
}
function sn(e) {
  return e >= 56320 && e <= 57343;
}
function Z(e, n) {
  const s = e.source.body.codePointAt(n);
  if (s === void 0)
    return b.EOF;
  if (s >= 32 && s <= 126) {
    const a = String.fromCodePoint(s);
    return a === '"' ? `'"'` : `"${a}"`;
  }
  return "U+" + s.toString(16).toUpperCase().padStart(4, "0");
}
function P(e, n, s, a, r) {
  const o = e.line, i = 1 + s - e.lineStart;
  return new Yt(n, s, a, o, i, r);
}
function ar(e, n) {
  const s = e.source.body, a = s.length;
  let r = n;
  for (; r < a; ) {
    const o = s.charCodeAt(r);
    switch (o) {
      // Ignored ::
      //   - UnicodeBOM
      //   - WhiteSpace
      //   - LineTerminator
      //   - Comment
      //   - Comma
      //
      // UnicodeBOM :: "Byte Order Mark (U+FEFF)"
      //
      // WhiteSpace ::
      //   - "Horizontal Tab (U+0009)"
      //   - "Space (U+0020)"
      //
      // Comma :: ,
      case 65279:
      // <BOM>
      case 9:
      // \t
      case 32:
      // <space>
      case 44:
        ++r;
        continue;
      // LineTerminator ::
      //   - "New Line (U+000A)"
      //   - "Carriage Return (U+000D)" [lookahead != "New Line (U+000A)"]
      //   - "Carriage Return (U+000D)" "New Line (U+000A)"
      case 10:
        ++r, ++e.line, e.lineStart = r;
        continue;
      case 13:
        s.charCodeAt(r + 1) === 10 ? r += 2 : ++r, ++e.line, e.lineStart = r;
        continue;
      // Comment
      case 35:
        return or(e, r);
      // Token ::
      //   - Punctuator
      //   - Name
      //   - IntValue
      //   - FloatValue
      //   - StringValue
      //
      // Punctuator :: one of ! $ & ( ) ... : = @ [ ] { | }
      case 33:
        return P(e, b.BANG, r, r + 1);
      case 36:
        return P(e, b.DOLLAR, r, r + 1);
      case 38:
        return P(e, b.AMP, r, r + 1);
      case 40:
        return P(e, b.PAREN_L, r, r + 1);
      case 41:
        return P(e, b.PAREN_R, r, r + 1);
      case 46:
        if (s.charCodeAt(r + 1) === 46 && s.charCodeAt(r + 2) === 46)
          return P(e, b.SPREAD, r, r + 3);
        break;
      case 58:
        return P(e, b.COLON, r, r + 1);
      case 61:
        return P(e, b.EQUALS, r, r + 1);
      case 64:
        return P(e, b.AT, r, r + 1);
      case 91:
        return P(e, b.BRACKET_L, r, r + 1);
      case 93:
        return P(e, b.BRACKET_R, r, r + 1);
      case 123:
        return P(e, b.BRACE_L, r, r + 1);
      case 124:
        return P(e, b.PIPE, r, r + 1);
      case 125:
        return P(e, b.BRACE_R, r, r + 1);
      // StringValue
      case 34:
        return s.charCodeAt(r + 1) === 34 && s.charCodeAt(r + 2) === 34 ? mr(e, r) : lr(e, r);
    }
    if (fe(o) || o === 45)
      return ir(e, r, o);
    if (tn(o))
      return pr(e, r);
    throw F(
      e.source,
      r,
      o === 39 ? `Unexpected single quote character ('), did you mean to use a double quote (")?` : ie(o) || Le(s, r) ? `Unexpected character: ${Z(e, r)}.` : `Invalid character: ${Z(e, r)}.`
    );
  }
  return P(e, b.EOF, a, a);
}
function or(e, n) {
  const s = e.source.body, a = s.length;
  let r = n + 1;
  for (; r < a; ) {
    const o = s.charCodeAt(r);
    if (o === 10 || o === 13)
      break;
    if (ie(o))
      ++r;
    else if (Le(s, r))
      r += 2;
    else
      break;
  }
  return P(
    e,
    b.COMMENT,
    n,
    r,
    s.slice(n + 1, r)
  );
}
function ir(e, n, s) {
  const a = e.source.body;
  let r = n, o = s, i = !1;
  if (o === 45 && (o = a.charCodeAt(++r)), o === 48) {
    if (o = a.charCodeAt(++r), fe(o))
      throw F(
        e.source,
        r,
        `Invalid number, unexpected digit after 0: ${Z(
          e,
          r
        )}.`
      );
  } else
    r = Ue(e, r, o), o = a.charCodeAt(r);
  if (o === 46 && (i = !0, o = a.charCodeAt(++r), r = Ue(e, r, o), o = a.charCodeAt(r)), (o === 69 || o === 101) && (i = !0, o = a.charCodeAt(++r), (o === 43 || o === 45) && (o = a.charCodeAt(++r)), r = Ue(e, r, o), o = a.charCodeAt(r)), o === 46 || tn(o))
    throw F(
      e.source,
      r,
      `Invalid number, expected digit but got: ${Z(
        e,
        r
      )}.`
    );
  return P(
    e,
    i ? b.FLOAT : b.INT,
    n,
    r,
    a.slice(n, r)
  );
}
function Ue(e, n, s) {
  if (!fe(s))
    throw F(
      e.source,
      n,
      `Invalid number, expected digit but got: ${Z(
        e,
        n
      )}.`
    );
  const a = e.source.body;
  let r = n + 1;
  for (; fe(a.charCodeAt(r)); )
    ++r;
  return r;
}
function lr(e, n) {
  const s = e.source.body, a = s.length;
  let r = n + 1, o = r, i = "";
  for (; r < a; ) {
    const c = s.charCodeAt(r);
    if (c === 34)
      return i += s.slice(o, r), P(e, b.STRING, n, r + 1, i);
    if (c === 92) {
      i += s.slice(o, r);
      const l = s.charCodeAt(r + 1) === 117 ? s.charCodeAt(r + 2) === 123 ? cr(e, r) : dr(e, r) : ur(e, r);
      i += l.value, r += l.size, o = r;
      continue;
    }
    if (c === 10 || c === 13)
      break;
    if (ie(c))
      ++r;
    else if (Le(s, r))
      r += 2;
    else
      throw F(
        e.source,
        r,
        `Invalid character within String: ${Z(
          e,
          r
        )}.`
      );
  }
  throw F(e.source, r, "Unterminated string.");
}
function cr(e, n) {
  const s = e.source.body;
  let a = 0, r = 3;
  for (; r < 12; ) {
    const o = s.charCodeAt(n + r++);
    if (o === 125) {
      if (r < 5 || !ie(a))
        break;
      return {
        value: String.fromCodePoint(a),
        size: r
      };
    }
    if (a = a << 4 | ce(o), a < 0)
      break;
  }
  throw F(
    e.source,
    n,
    `Invalid Unicode escape sequence: "${s.slice(
      n,
      n + r
    )}".`
  );
}
function dr(e, n) {
  const s = e.source.body, a = $t(s, n + 2);
  if (ie(a))
    return {
      value: String.fromCodePoint(a),
      size: 6
    };
  if (nn(a) && s.charCodeAt(n + 6) === 92 && s.charCodeAt(n + 7) === 117) {
    const r = $t(s, n + 8);
    if (sn(r))
      return {
        value: String.fromCodePoint(a, r),
        size: 12
      };
  }
  throw F(
    e.source,
    n,
    `Invalid Unicode escape sequence: "${s.slice(n, n + 6)}".`
  );
}
function $t(e, n) {
  return ce(e.charCodeAt(n)) << 12 | ce(e.charCodeAt(n + 1)) << 8 | ce(e.charCodeAt(n + 2)) << 4 | ce(e.charCodeAt(n + 3));
}
function ce(e) {
  return e >= 48 && e <= 57 ? e - 48 : e >= 65 && e <= 70 ? e - 55 : e >= 97 && e <= 102 ? e - 87 : -1;
}
function ur(e, n) {
  const s = e.source.body;
  switch (s.charCodeAt(n + 1)) {
    case 34:
      return {
        value: '"',
        size: 2
      };
    case 92:
      return {
        value: "\\",
        size: 2
      };
    case 47:
      return {
        value: "/",
        size: 2
      };
    case 98:
      return {
        value: "\b",
        size: 2
      };
    case 102:
      return {
        value: "\f",
        size: 2
      };
    case 110:
      return {
        value: `
`,
        size: 2
      };
    case 114:
      return {
        value: "\r",
        size: 2
      };
    case 116:
      return {
        value: "	",
        size: 2
      };
  }
  throw F(
    e.source,
    n,
    `Invalid character escape sequence: "${s.slice(
      n,
      n + 2
    )}".`
  );
}
function mr(e, n) {
  const s = e.source.body, a = s.length;
  let r = e.lineStart, o = n + 3, i = o, c = "";
  const l = [];
  for (; o < a; ) {
    const m = s.charCodeAt(o);
    if (m === 34 && s.charCodeAt(o + 1) === 34 && s.charCodeAt(o + 2) === 34) {
      c += s.slice(i, o), l.push(c);
      const d = P(
        e,
        b.BLOCK_STRING,
        n,
        o + 3,
        // Return a string of the lines joined with U+000A.
        er(l).join(`
`)
      );
      return e.line += l.length - 1, e.lineStart = r, d;
    }
    if (m === 92 && s.charCodeAt(o + 1) === 34 && s.charCodeAt(o + 2) === 34 && s.charCodeAt(o + 3) === 34) {
      c += s.slice(i, o), i = o + 1, o += 4;
      continue;
    }
    if (m === 10 || m === 13) {
      c += s.slice(i, o), l.push(c), m === 13 && s.charCodeAt(o + 1) === 10 ? o += 2 : ++o, c = "", i = o, r = o;
      continue;
    }
    if (ie(m))
      ++o;
    else if (Le(s, o))
      o += 2;
    else
      throw F(
        e.source,
        o,
        `Invalid character within String: ${Z(
          e,
          o
        )}.`
      );
  }
  throw F(e.source, o, "Unterminated string.");
}
function pr(e, n) {
  const s = e.source.body, a = s.length;
  let r = n + 1;
  for (; r < a; ) {
    const o = s.charCodeAt(r);
    if (Ys(o))
      ++r;
    else
      break;
  }
  return P(
    e,
    b.NAME,
    n,
    r,
    s.slice(n, r)
  );
}
const hr = 10, rn = 2;
function an(e) {
  return ke(e, []);
}
function ke(e, n) {
  switch (typeof e) {
    case "string":
      return JSON.stringify(e);
    case "function":
      return e.name ? `[function ${e.name}]` : "[function]";
    case "object":
      return fr(e, n);
    default:
      return String(e);
  }
}
function fr(e, n) {
  if (e === null)
    return "null";
  if (n.includes(e))
    return "[Circular]";
  const s = [...n, e];
  if (gr(e)) {
    const a = e.toJSON();
    if (a !== e)
      return typeof a == "string" ? a : ke(a, s);
  } else if (Array.isArray(e))
    return yr(e, s);
  return xr(e, s);
}
function gr(e) {
  return typeof e.toJSON == "function";
}
function xr(e, n) {
  const s = Object.entries(e);
  return s.length === 0 ? "{}" : n.length > rn ? "[" + vr(e) + "]" : "{ " + s.map(
    ([r, o]) => r + ": " + ke(o, n)
  ).join(", ") + " }";
}
function yr(e, n) {
  if (e.length === 0)
    return "[]";
  if (n.length > rn)
    return "[Array]";
  const s = Math.min(hr, e.length), a = e.length - s, r = [];
  for (let o = 0; o < s; ++o)
    r.push(ke(e[o], n));
  return a === 1 ? r.push("... 1 more item") : a > 1 && r.push(`... ${a} more items`), "[" + r.join(", ") + "]";
}
function vr(e) {
  const n = Object.prototype.toString.call(e).replace(/^\[object /, "").replace(/]$/, "");
  if (n === "Object" && typeof e.constructor == "function") {
    const s = e.constructor.name;
    if (typeof s == "string" && s !== "")
      return s;
  }
  return n;
}
const br = globalThis.process && // eslint-disable-next-line no-undef
process.env.NODE_ENV === "production", jr = (
  /* c8 ignore next 6 */
  // FIXME: https://github.com/graphql/graphql-js/issues/2317
  br ? function(n, s) {
    return n instanceof s;
  } : function(n, s) {
    if (n instanceof s)
      return !0;
    if (typeof n == "object" && n !== null) {
      var a;
      const r = s.prototype[Symbol.toStringTag], o = (
        // We still need to support constructor's name to detect conflicts with older versions of this library.
        Symbol.toStringTag in n ? n[Symbol.toStringTag] : (a = n.constructor) === null || a === void 0 ? void 0 : a.name
      );
      if (r === o) {
        const i = an(n);
        throw new Error(`Cannot use ${r} "${i}" from another module or realm.

Ensure that there is only one instance of "graphql" in the node_modules
directory. If different versions of "graphql" are the dependencies of other
relied on modules, use "resolutions" to ensure only one version is installed.

https://yarnpkg.com/en/docs/selective-version-resolutions

Duplicate "graphql" modules cannot be used at the same time since different
versions may have different capabilities and behavior. The data from one
version used in the function from another could produce confusing and
spurious results.`);
      }
    }
    return !1;
  }
);
class on {
  constructor(n, s = "GraphQL request", a = {
    line: 1,
    column: 1
  }) {
    typeof n == "string" || qe(!1, `Body must be a string. Received: ${an(n)}.`), this.body = n, this.name = s, this.locationOffset = a, this.locationOffset.line > 0 || qe(
      !1,
      "line in locationOffset is 1-indexed and must be positive."
    ), this.locationOffset.column > 0 || qe(
      !1,
      "column in locationOffset is 1-indexed and must be positive."
    );
  }
  get [Symbol.toStringTag]() {
    return "Source";
  }
}
function Cr(e) {
  return jr(e, on);
}
function Nr(e) {
  const n = Cr(e) ? e : new on(e), s = n.body, a = new sr(n);
  let r = "", o = !1;
  for (; a.advance().kind !== b.EOF; ) {
    const i = a.token, c = i.kind, l = !rr(i.kind);
    o && (l || i.kind === b.SPREAD) && (r += " ");
    const m = s.slice(i.start, i.end);
    c === b.BLOCK_STRING ? r += nr(i.value, {
      minimize: !0
    }) : r += m, o = l;
  }
  return r;
}
const ln = zt(
  void 0
), ao = ({
  children: e,
  client: n
}) => /* @__PURE__ */ t.jsx(ln.Provider, { value: n, children: e }), cn = (e, n, ...[s]) => ({
  queryFn: () => e.fetch(n, s),
  queryKey: [Nr(n.toString()), s]
}), wr = (e, ...[n]) => {
  const s = is(ln);
  if (s === void 0)
    throw new Error("useGraphQL must be used within a GraphQLProvider");
  return cn(s, e, ...n === void 0 ? [] : [n]);
};
class ee extends String {
  __apiType;
  value;
  __meta__;
  constructor(n, s) {
    super(n), this.value = n, this.__meta__ = s;
  }
  toString() {
    return this.value;
  }
}
const Sr = new ee(
  `
    fragment OperationsFragment on OperationItem {
  slug
  summary
  method
  description
  operationId
  contentTypes
  path
  deprecated
  extensions
  servers {
    url
    description
  }
  parameters {
    name
    in
    description
    required
    schema
    style
    explode
    examples {
      name
      description
      externalValue
      value
      summary
    }
  }
  requestBody {
    content {
      mediaType
      encoding {
        name
      }
      examples {
        name
        description
        externalValue
        value
        summary
      }
      schema
    }
    description
    required
  }
  responses {
    statusCode
    links
    description
    content {
      examples {
        name
        description
        externalValue
        value
        summary
      }
      mediaType
      encoding {
        name
      }
      schema
    }
  }
}
    `,
  { fragmentName: "OperationsFragment" }
), Tr = new ee(`
    query ServersQuery($input: JSON!, $type: SchemaType!) {
  schema(input: $input, type: $type) {
    url
    servers {
      url
    }
  }
}
    `), Ar = new ee(`
    query SchemaWarmup($input: JSON!, $type: SchemaType!) {
  schema(input: $input, type: $type) {
    openapi
  }
}
    `), Or = new ee(`
    query OperationsForTag($input: JSON!, $type: SchemaType!, $tag: String, $untagged: Boolean) {
  schema(input: $input, type: $type) {
    servers {
      url
    }
    description
    summary
    title
    url
    version
    tag(slug: $tag, untagged: $untagged) {
      name
      description
      operations {
        slug
        ...OperationsFragment
      }
      extensions
      next {
        name
        slug
        extensions
      }
      prev {
        name
        slug
        extensions
      }
    }
  }
}
    fragment OperationsFragment on OperationItem {
  slug
  summary
  method
  description
  operationId
  contentTypes
  path
  deprecated
  extensions
  servers {
    url
    description
  }
  parameters {
    name
    in
    description
    required
    schema
    style
    explode
    examples {
      name
      description
      externalValue
      value
      summary
    }
  }
  requestBody {
    content {
      mediaType
      encoding {
        name
      }
      examples {
        name
        description
        externalValue
        value
        summary
      }
      schema
    }
    description
    required
  }
  responses {
    statusCode
    links
    description
    content {
      examples {
        name
        description
        externalValue
        value
        summary
      }
      mediaType
      encoding {
        name
      }
      schema
    }
  }
}`), Pr = new ee(`
    query GetSchemas($input: JSON!, $type: SchemaType!) {
  schema(input: $input, type: $type) {
    title
    description
    summary
    components {
      schemas {
        name
        schema
        extensions
      }
    }
  }
}
    `), $r = new ee(`
    query getServerQuery($input: JSON!, $type: SchemaType!) {
  schema(input: $input, type: $type) {
    url
    servers {
      url
    }
  }
}
    `), Lr = new ee(`
    query GetNavigationOperations($input: JSON!, $type: SchemaType!) {
  schema(input: $input, type: $type) {
    extensions
    tags {
      slug
      name
      extensions
      operations {
        summary
        slug
        method
        operationId
        path
      }
    }
    components {
      schemas {
        __typename
      }
    }
  }
}
    `), kr = {
  "\n  query ServersQuery($input: JSON!, $type: SchemaType!) {\n    schema(input: $input, type: $type) {\n      url\n      servers {\n        url\n      }\n    }\n  }\n": Tr,
  "\n  fragment OperationsFragment on OperationItem {\n    slug\n    summary\n    method\n    description\n    operationId\n    contentTypes\n    path\n    deprecated\n    extensions\n    servers {\n      url\n      description\n    }\n    parameters {\n      name\n      in\n      description\n      required\n      schema\n      style\n      explode\n      examples {\n        name\n        description\n        externalValue\n        value\n        summary\n      }\n    }\n    requestBody {\n      content {\n        mediaType\n        encoding {\n          name\n        }\n        examples {\n          name\n          description\n          externalValue\n          value\n          summary\n        }\n        schema\n      }\n      description\n      required\n    }\n    responses {\n      statusCode\n      links\n      description\n      content {\n        examples {\n          name\n          description\n          externalValue\n          value\n          summary\n        }\n        mediaType\n        encoding {\n          name\n        }\n        schema\n      }\n    }\n  }\n": Sr,
  "\n  query SchemaWarmup($input: JSON!, $type: SchemaType!) {\n    schema(input: $input, type: $type) {\n      openapi\n    }\n  }\n": Ar,
  "\n  query OperationsForTag(\n    $input: JSON!\n    $type: SchemaType!\n    $tag: String\n    $untagged: Boolean\n  ) {\n    schema(input: $input, type: $type) {\n      servers {\n        url\n      }\n      description\n      summary\n      title\n      url\n      version\n      tag(slug: $tag, untagged: $untagged) {\n        name\n        description\n        operations {\n          slug\n          ...OperationsFragment\n        }\n        extensions\n        next {\n          name\n          slug\n          extensions\n        }\n        prev {\n          name\n          slug\n          extensions\n        }\n      }\n    }\n  }\n": Or,
  "\n  query GetSchemas($input: JSON!, $type: SchemaType!) {\n    schema(input: $input, type: $type) {\n      title\n      description\n      summary\n      components {\n        schemas {\n          name\n          schema\n          extensions\n        }\n      }\n    }\n  }\n": Pr,
  "\n  query getServerQuery($input: JSON!, $type: SchemaType!) {\n    schema(input: $input, type: $type) {\n      url\n      servers {\n        url\n      }\n    }\n  }\n": $r,
  "\n  query GetNavigationOperations($input: JSON!, $type: SchemaType!) {\n    schema(input: $input, type: $type) {\n      extensions\n      tags {\n        slug\n        name\n        extensions\n        operations {\n          summary\n          slug\n          method\n          operationId\n          path\n        }\n      }\n      components {\n        schemas {\n          __typename\n        }\n      }\n    }\n  }\n": Lr
};
function Er(e) {
  return kr[e] ?? {};
}
function _({
  ...e
}) {
  return /* @__PURE__ */ t.jsx(st.Root, { "data-slot": "collapsible", ...e });
}
function de({
  ...e
}) {
  return /* @__PURE__ */ t.jsx(
    st.CollapsibleTrigger,
    {
      "data-slot": "collapsible-trigger",
      ...e
    }
  );
}
function q({
  ...e
}) {
  return /* @__PURE__ */ t.jsx(
    st.CollapsibleContent,
    {
      "data-slot": "collapsible-content",
      ...e
    }
  );
}
function Rr() {
  const n = navigator.userAgent.toLowerCase();
  if (n.indexOf("win") !== -1)
    return "windows";
  if (n.indexOf("mac") !== -1)
    return "apple";
  if (n.indexOf("linux") !== -1)
    return "linux";
  if (n.indexOf("x11") !== -1)
    return "unix";
}
function Ir() {
  return Rr() === "apple";
}
const Dr = {
  other: {
    alt: "Alt",
    meta: "Ctrl",
    ctrl: "Ctrl",
    shift: "⇧",
    option: "Alt"
  },
  apple: {
    alt: "⌥",
    meta: "⌘",
    ctrl: "Ctrl",
    shift: "⇧",
    option: "⌥"
  }
}, Fr = (e, n) => {
  const s = e.toLowerCase().split("+"), a = s.includes("meta"), r = s.includes("shift"), o = s.includes("option") || s.includes("alt"), i = s.includes("ctrl"), c = e.split("+").pop(), l = E(n);
  return U(() => {
    l.current = n;
  }, [n]), U(() => {
    const m = (d) => {
      (d.code === `Key${c?.toUpperCase()}` || d.code.toLowerCase() === c?.toLowerCase()) && d.metaKey === a && d.shiftKey === r && d.altKey === o && d.ctrlKey === i && (d.preventDefault(), l.current());
    };
    return window.addEventListener("keydown", m), () => {
      window.removeEventListener("keydown", m);
    };
  }, [c, a, r, o, i]), {
    trigger: () => l.current(),
    label: s.map(
      (m) => m === "meta" || m === "ctrl" || m === "alt" || m === "shift" || m === "option" ? Dr[Ir() ? "apple" : "other"][m] : m
    )
  };
}, zr = Ze()(
  Ye(
    (e) => ({
      selectedServer: void 0,
      setSelectedServer: (n) => e({ selectedServer: n })
    }),
    { name: "zudoku-selected-server" }
  )
), Vr = (e) => {
  const { selectedServer: n, setSelectedServer: s } = zr();
  return { selectedServer: Vt(
    () => n && e.some((r) => r.url === n) ? n : e.at(0)?.url ?? "",
    [n, e]
  ), setSelectedServer: s };
}, dn = xe.forwardRef(
  ({ className: e, ...n }, s) => /* @__PURE__ */ t.jsx(
    "textarea",
    {
      className: S(
        "flex min-h-[80px] w-full rounded-md border border-input bg-background px-3 py-2 text-base ring-offset-background placeholder:text-muted-foreground focus-visible:outline-hidden focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 md:text-sm",
        e
      ),
      ref: s,
      ...n
    }
  )
);
dn.displayName = "Textarea";
const ve = (e = 0) => {
  if (e === 0)
    return "0 B";
  const n = Math.floor(Math.log(e) / Math.log(1e3));
  return `${(e / 1e3 ** n).toFixed(n ? 2 : 0)} ${n ? `${"kMGTPEZY"[n - 1]}B` : "B"}`;
}, Y = ({
  children: e,
  className: n
}) => /* @__PURE__ */ t.jsxs(
  "div",
  {
    className: S(
      "grid grid-cols-[max-content_1fr_max-content] items-center gap-2 group bg-muted w-full h-10 ps-4 pe-2 border-b",
      n
    ),
    children: [
      e,
      /* @__PURE__ */ t.jsxs(
        de,
        {
          className: S(
            "flex items-center gap-4 group bg-muted w-full p-2 hover:bg-accent hover:brightness-95 opacity-75 rounded-md",
            n
          ),
          children: [
            /* @__PURE__ */ t.jsx(
              kn,
              {
                className: "group-data-[state=closed]:hidden shrink-0",
                size: 14
              }
            ),
            /* @__PURE__ */ t.jsx(
              En,
              {
                className: "group-data-[state=open]:hidden shrink-0",
                size: 14
              }
            )
          ]
        }
      )
    ]
  }
), H = ({
  children: e,
  className: n
}) => /* @__PURE__ */ t.jsx("span", { className: S("font-semibold w-full text-start", n), children: e }), Br = ({
  examples: e,
  onSelect: n
}) => /* @__PURE__ */ t.jsxs(Jt, { children: [
  /* @__PURE__ */ t.jsx(Wt, { asChild: !0, children: /* @__PURE__ */ t.jsxs(
    B,
    {
      variant: "ghost",
      size: "sm",
      className: "hover:bg-accent hover:brightness-95 gap-2",
      children: [
        "Use Example",
        /* @__PURE__ */ t.jsx(Je, { size: 14 })
      ]
    }
  ) }),
  /* @__PURE__ */ t.jsx(Xt, { className: "max-w-72", children: e.map((s) => /* @__PURE__ */ t.jsxs("div", { children: [
    /* @__PURE__ */ t.jsx(Ts, { children: s.mediaType }),
    /* @__PURE__ */ t.jsx(As, {}),
    /* @__PURE__ */ t.jsx(Os, { children: s.examples?.map((a) => /* @__PURE__ */ t.jsx(
      Oe,
      {
        onSelect: () => n(a, s.mediaType),
        children: /* @__PURE__ */ t.jsx(
          "span",
          {
            className: "line-clamp-1",
            title: a.summary ?? a.name,
            children: a.summary ?? a.name
          }
        )
      },
      a.name
    )) })
  ] }, s.mediaType)) })
] }), Ee = ye(
  "div",
  "grid grid-cols-[min-content_2fr_3fr] items-center gap-x-5 [&>*:last-child_[data-slot=remove-button]]:invisible"
), ge = ye(
  "div",
  "group h-9 hover:bg-accent/75 ps-4 pe-2 grid col-span-full grid-cols-subgrid items-center border-b"
), I = ye(
  Ps,
  "w-full truncate border-0 p-0 m-0 shadow-none text-xs focus-visible:ring-0 font-mono"
), at = ({
  onClick: e,
  className: n
}) => /* @__PURE__ */ t.jsx(
  K,
  {
    size: "icon-xs",
    variant: "ghost",
    className: S(
      "text-muted-foreground opacity-0 group-hover:brightness-95 focus-visible:opacity-100 group-hover:opacity-100",
      n
    ),
    onClick: e,
    type: "button",
    "data-slot": "remove-button",
    children: /* @__PURE__ */ t.jsx(It, { size: 14 })
  }
);
function be({
  className: e,
  ...n
}) {
  return /* @__PURE__ */ t.jsx(
    Tt.Root,
    {
      "data-slot": "checkbox",
      className: S(
        "peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50",
        e
      ),
      ...n,
      children: /* @__PURE__ */ t.jsx(
        Tt.Indicator,
        {
          "data-slot": "checkbox-indicator",
          className: "grid place-content-center text-current transition-none",
          children: /* @__PURE__ */ t.jsx(Dt, { className: "size-3.5" })
        }
      )
    }
  );
}
const _r = ({ index: e, manager: n }) => {
  const s = E(null), a = n.getValue(e, "value");
  return /* @__PURE__ */ t.jsxs(ge, { children: [
    /* @__PURE__ */ t.jsx(
      be,
      {
        ...n.getCheckboxProps(e),
        disabled: !n.getValue(e, "name")
      }
    ),
    /* @__PURE__ */ t.jsx(
      I,
      {
        ...n.getNameInputProps(e),
        placeholder: "Key"
      }
    ),
    /* @__PURE__ */ t.jsxs("div", { className: "flex items-center gap-1 flex-1", children: [
      a instanceof File ? /* @__PURE__ */ t.jsxs("div", { className: "flex items-center gap-1 flex-1 min-w-0", children: [
        /* @__PURE__ */ t.jsxs("div", { className: "flex items-center gap-1 border-b cursor-default", children: [
          /* @__PURE__ */ t.jsx(me, { size: 12, className: "text-muted-foreground" }),
          /* @__PURE__ */ t.jsx(
            "span",
            {
              className: "text-xs truncate",
              title: `${a.name} (${ve(a.size)})`,
              children: a.name
            }
          )
        ] }),
        /* @__PURE__ */ t.jsx(
          K,
          {
            type: "button",
            variant: "ghost",
            size: "icon-xs",
            className: "opacity-50 hover:opacity-100 hover:brightness-95 transition-opacity",
            onClick: () => n.setValue(e, "value", ""),
            children: /* @__PURE__ */ t.jsx(Rn, { size: 13 })
          }
        )
      ] }) : /* @__PURE__ */ t.jsxs(t.Fragment, { children: [
        /* @__PURE__ */ t.jsx(
          I,
          {
            ...n.getValueInputProps(e),
            placeholder: "Value"
          }
        ),
        /* @__PURE__ */ t.jsx(
          "input",
          {
            ref: s,
            type: "file",
            className: "hidden",
            onChange: (r) => {
              const o = r.target.files?.[0];
              o && (n.setValue(e, "value", o), n.setValue(e, "active", !0));
            }
          }
        ),
        /* @__PURE__ */ t.jsx(
          K,
          {
            type: "button",
            variant: "ghost",
            size: "icon-xs",
            onClick: () => s.current?.click(),
            title: "Attach file",
            className: "opacity-0 focus-visible:opacity-100 group-hover:opacity-100 group-hover:brightness-95 transition-opacity",
            children: /* @__PURE__ */ t.jsx(me, { size: 14 })
          }
        )
      ] }),
      /* @__PURE__ */ t.jsx(at, { ...n.getRemoveButtonProps(e) })
    ] })
  ] });
}, ot = (e) => {
  const {
    control: n,
    name: s,
    defaultValue: a,
    isEmpty: r,
    shouldSetActive: o
  } = e, {
    setValue: i,
    watch: c,
    setFocus: l,
    register: m
  } = oe(), { fields: d, append: f, remove: x } = Ut({ control: n, name: s }), p = c(s), v = E(-1), u = E(-1), y = D(
    (g, C, T, A) => {
      (C === "value" || C === "name") && (v.current = g), i(`${s}.${g}.${C}`, T), A?.focus === "next" ? l(
        C === "name" ? `${s}.${g}.value` : `${s}.${g + 1}.name`
      ) : A?.focus === "previous" && l(
        C === "name" ? `${s}.${g - 1}.value` : `${s}.${g}.name`
      );
    },
    [s, i, l]
  ), w = D(
    (g) => r ? r(g) : !g.name && !g.value,
    [r]
  ), R = D(
    (g) => o ? o(g) : !!(g.name || g.value),
    [o]
  );
  U(() => {
    if (!p) return;
    if (u.current === -1) {
      u.current = p.length, p.length === 0 && f(a, {
        shouldFocus: !1
      });
      return;
    }
    if (u.current = p.length, p.length === 0) {
      f(a, {
        shouldFocus: !1
      });
      return;
    }
    if (p.length > 1) {
      const C = [];
      for (let T = 0; T < p.length - 1; T++) {
        const A = p[T];
        A && w(A) && !R(A) && C.push(T);
      }
      if (C.length > 0) {
        const T = C[0];
        if (T === void 0) return;
        for (let A = C.length - 1; A >= 0; A--) {
          const M = C[A];
          M !== void 0 && x(M);
        }
        if (v.current === T) {
          const A = p.length - C.length;
          T < A ? l(`${s}.${T}.name`) : T > 0 ? l(`${s}.${T - 1}.name`) : l(`${s}.0.name`);
        }
        v.current = -1;
      }
    }
    const g = p[p.length - 1];
    g && !w(g) && f(a, {
      shouldFocus: !1
    });
  }, [
    p,
    f,
    x,
    a,
    w,
    s,
    l,
    R
  ]), U(() => {
    if (!p) return;
    const g = [];
    for (let C = 0; C < p.length; C++) {
      const T = p[C];
      if (!T) continue;
      const A = R(T);
      T.active !== A && g.push(() => y(C, "active", A));
    }
    g.length !== 0 && ls(() => g.forEach((C) => C()));
  }, [p, R, y]);
  const N = D(
    (g) => {
      const C = p?.[g];
      return C ? w(C) : !0;
    },
    [p, w]
  ), j = D(
    (g, C) => {
      const T = C === "name" ? `${s}.${g}.value` : `${s}.${g + 1}.name`, A = C === "name" ? `${s}.${g - 1}.value` : `${s}.${g}.name`, M = C === "value" || g > 0;
      return (L) => {
        if (!(L.target instanceof HTMLInputElement)) return;
        const De = L.target.selectionStart === 0, ne = L.target.selectionStart === L.target.value.length, Ce = !L.target.value;
        L.key === "Enter" ? l(T) : L.key === "Backspace" && Ce && M || L.key === "ArrowLeft" && De && M ? (L.preventDefault(), l(A)) : L.key === "ArrowRight" && ne && (L.preventDefault(), l(T));
      };
    },
    [s, l]
  ), $ = D(
    (g) => ({
      ...m(`${s}.${g}.name`),
      onChange: (C) => y(g, "name", C.target.value),
      onKeyDown: j(g, "name")
    }),
    [m, s, y, j]
  ), Q = D(
    (g) => ({
      ...m(`${s}.${g}.value`),
      onChange: (C) => y(g, "value", C.target.value),
      onKeyDown: j(g, "value")
    }),
    [m, s, y, j]
  ), je = D(
    (g) => ({
      ...m(`${s}.${g}.active`),
      checked: c(`${s}.${g}.active`) ?? !1,
      disabled: N(g),
      onCheckedChange: (C) => {
        y(g, "active", C === !0);
      }
    }),
    [s, m, N, c, y]
  ), Ie = D(
    (g, C) => p?.[g]?.[C],
    [p]
  ), te = D(
    (g) => ({
      onClick: () => x(g),
      disabled: g === d.length - 1
    }),
    [x, d.length]
  );
  return {
    fields: d,
    getNameInputProps: $,
    getValueInputProps: Q,
    getCheckboxProps: je,
    getRemoveButtonProps: te,
    setValue: y,
    getValue: Ie
  };
}, qr = ({ content: e }) => {
  const { register: n, setValue: s, watch: a, control: r } = oe(), o = (e ?? []).flatMap((j) => j.examples), [i, c, l, m, d] = a([
    "headers",
    "file",
    "bodyMode",
    "body",
    "multipartFormFields"
  ]), f = E(null), [x, p] = k(!1), v = (j) => {
    s("file", j), j && s(
      "headers",
      i.filter(
        ($) => $.name.toLowerCase() !== "content-type" || !$.active
      )
    );
  }, u = (j) => {
    const $ = j.target.files?.[0] ?? null;
    v($);
  }, y = (j) => {
    j.preventDefault(), j.stopPropagation(), p(!0);
  }, w = (j) => {
    j.preventDefault(), j.stopPropagation(), p(!1);
  }, R = (j) => {
    j.preventDefault(), j.stopPropagation(), p(!1);
    const $ = j.dataTransfer.files?.[0] ?? null;
    v($);
  }, N = ot({
    control: r,
    name: "multipartFormFields",
    defaultValue: { name: "", value: "", active: !1 },
    isEmpty: (j) => j.value instanceof File ? !1 : !j.name && !j.value
  });
  return /* @__PURE__ */ t.jsxs(_, { defaultOpen: !0, children: [
    /* @__PURE__ */ t.jsxs(Y, { className: "items-center", children: [
      /* @__PURE__ */ t.jsx(In, { size: 16 }),
      /* @__PURE__ */ t.jsxs(H, { className: "flex items-center justify-between", children: [
        "Body",
        /* @__PURE__ */ t.jsxs("div", { className: "flex items-center", children: [
          /* @__PURE__ */ t.jsxs(Jt, { children: [
            /* @__PURE__ */ t.jsx(Wt, { asChild: !0, children: /* @__PURE__ */ t.jsxs(
              K,
              {
                variant: "ghost",
                size: "sm",
                className: "hover:bg-accent hover:brightness-95 gap-2",
                children: [
                  l === "text" ? /* @__PURE__ */ t.jsxs(t.Fragment, { children: [
                    /* @__PURE__ */ t.jsx(yt, { size: 14 }),
                    "Text"
                  ] }) : l === "file" ? /* @__PURE__ */ t.jsxs(t.Fragment, { children: [
                    /* @__PURE__ */ t.jsx(me, { size: 14 }),
                    "File"
                  ] }) : /* @__PURE__ */ t.jsxs(t.Fragment, { children: [
                    /* @__PURE__ */ t.jsx(vt, { size: 14 }),
                    "Multipart"
                  ] }),
                  /* @__PURE__ */ t.jsx(Dn, { size: 14 })
                ]
              }
            ) }),
            /* @__PURE__ */ t.jsxs(Xt, { className: "min-w-40", children: [
              /* @__PURE__ */ t.jsxs(
                Oe,
                {
                  onSelect: () => s("bodyMode", "text"),
                  className: "gap-2",
                  children: [
                    /* @__PURE__ */ t.jsx(yt, { size: 14 }),
                    /* @__PURE__ */ t.jsx("span", { className: "flex-1", children: "Text" }),
                    /* @__PURE__ */ t.jsx("span", { children: m.length > 0 && /* @__PURE__ */ t.jsx("div", { className: "w-1.5 h-1.5 bg-primary rounded-full" }) })
                  ]
                }
              ),
              /* @__PURE__ */ t.jsxs(
                Oe,
                {
                  onSelect: () => s("bodyMode", "file"),
                  className: "gap-2",
                  children: [
                    /* @__PURE__ */ t.jsx(me, { size: 14 }),
                    /* @__PURE__ */ t.jsx("span", { className: "flex-1", children: "File" }),
                    /* @__PURE__ */ t.jsx("span", { children: c && /* @__PURE__ */ t.jsx("div", { className: "w-1.5 h-1.5 bg-primary rounded-full" }) })
                  ]
                }
              ),
              /* @__PURE__ */ t.jsxs(
                Oe,
                {
                  onSelect: () => s("bodyMode", "multipart"),
                  className: "gap-2",
                  children: [
                    /* @__PURE__ */ t.jsx(vt, { size: 14, strokeWidth: 1.5 }),
                    /* @__PURE__ */ t.jsx("span", { className: "flex-1", children: "Multipart" }),
                    /* @__PURE__ */ t.jsx("span", { children: d?.some((j) => j.active) && /* @__PURE__ */ t.jsx("div", { className: "w-1.5 h-1.5 bg-primary rounded-full" }) })
                  ]
                }
              )
            ] })
          ] }),
          /* @__PURE__ */ t.jsx(
            "input",
            {
              ref: f,
              type: "file",
              className: "hidden",
              onChange: u
            }
          ),
          /* @__PURE__ */ t.jsx("div", { className: "w-px mx-1 h-5 bg-border" }),
          e && o.length > 0 ? /* @__PURE__ */ t.jsx(
            Br,
            {
              examples: e,
              onSelect: (j, $) => {
                s("body", JSON.stringify(j.value, null, 2)), s("headers", [
                  ...i.filter((Q) => Q.name !== "Content-Type"),
                  {
                    name: "Content-Type",
                    value: $,
                    active: !0
                  }
                ]);
              }
            }
          ) : /* @__PURE__ */ t.jsx("div", {})
        ] })
      ] })
    ] }),
    /* @__PURE__ */ t.jsxs(q, { className: "CollapsibleContent flex flex-col gap-2", children: [
      l === "text" && /* @__PURE__ */ t.jsx(
        dn,
        {
          ...n("body"),
          className: S(
            "w-full px-4 py-2.5 h-64 font-mono md:text-xs border-none rounded-none focus-visible:ring-0 transition-colors"
          ),
          placeholder: "Body content"
        }
      ),
      l === "file" && /* @__PURE__ */ t.jsxs(
        "div",
        {
          role: "region",
          "aria-label": "File upload drop zone",
          className: S(
            "flex flex-col items-center justify-center gap-4 min-h-[300px]"
          ),
          onDragOver: y,
          onDragLeave: w,
          onDrop: R,
          children: [
            /* @__PURE__ */ t.jsx(
              "button",
              {
                type: "button",
                onClick: () => f.current?.click(),
                className: S(
                  "flex items-center justify-center gap-2 rounded-full size-20 p-0 border border-dashed border-muted-foreground/50 hover:bg-accent/75 transition-colors",
                  (c || x) && "border-solid",
                  x && "bg-accent border-primary"
                ),
                children: /* @__PURE__ */ t.jsx(
                  me,
                  {
                    className: S(
                      "text-muted-foreground",
                      x && "text-primary"
                    ),
                    size: 30
                  }
                )
              }
            ),
            c ? /* @__PURE__ */ t.jsxs("div", { className: "flex items-center justify-between gap-2 px-2 py-1.5 rounded-md border", children: [
              /* @__PURE__ */ t.jsxs("span", { className: "text-sm truncate", title: c.name, children: [
                c.name,
                " ",
                /* @__PURE__ */ t.jsxs("span", { className: "text-muted-foreground", children: [
                  "(",
                  ve(c.size),
                  ")"
                ] })
              ] }),
              /* @__PURE__ */ t.jsx(
                K,
                {
                  type: "button",
                  variant: "ghost",
                  size: "icon-xxs",
                  onClick: () => v(null),
                  children: /* @__PURE__ */ t.jsx(It, { size: 14 })
                }
              )
            ] }) : /* @__PURE__ */ t.jsx("span", { className: "text-lg font-semibold text-muted-foreground", children: "Select or drop a file" })
          ]
        }
      ),
      l === "multipart" && /* @__PURE__ */ t.jsx(Ee, { children: N.fields.map((j, $) => /* @__PURE__ */ t.jsx(_r, { index: $, manager: N }, j.id)) })
    ] })
  ] });
}, Lt = (e, n, s) => {
  const a = n.replace(/(:\w+|\{\w+})/g, (o) => {
    const i = o.replace(/[:{}]/g, "");
    return s.pathParams.find((l) => l.name === i)?.value ?? o;
  }), r = new URL(
    a.replace(/^\//, ""),
    e.endsWith("/") ? e : `${e}/`
  );
  return s.queryParams.filter((o) => o.active).forEach((o) => {
    r.searchParams.set(o.name, o.value);
  }), r;
};
function Ur(e) {
  return /^(application\/octet-stream|image\/|audio\/|video\/|font\/|application\/pdf|application\/zip|application\/x-protobuf|application\/x-binary)/i.test(
    e
  );
}
function Mr(e) {
  return /^audio\//i.test(e);
}
const Gr = (e, n) => {
  const s = e.find(
    ([a]) => a.toLowerCase() === "content-disposition"
  )?.[1];
  if (s) {
    const a = s.match(
      /filename[^;=\n]*=((['"]).*?\2|[^;\n]*)/
    );
    if (a?.[1])
      return a[1].replace(/['"]/g, "");
  }
  try {
    const r = new URL(n).pathname.split("/").pop() || "download";
    return r.includes(".") ? r : "download";
  } catch {
    return "download";
  }
}, Hr = Pe.Root, un = xe.forwardRef(({ className: e, align: n = "center", sideOffset: s = 4, ...a }, r) => /* @__PURE__ */ t.jsx(Pe.Portal, { children: /* @__PURE__ */ t.jsx(
  Pe.Content,
  {
    ref: r,
    align: n,
    sideOffset: s,
    className: S(
      "z-50 w-72 rounded-md border bg-popover p-4 text-popover-foreground shadow-md outline-hidden data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2",
      e
    ),
    ...a
  }
) }));
un.displayName = Pe.Content.displayName;
const Qr = ({
  value: e,
  options: n,
  onChange: s,
  className: a,
  placeholder: r = "Value",
  onEnterPress: o,
  onKeyDown: i,
  ref: c,
  onSelect: l
}) => {
  const [m, d] = k(!1), [f, x] = k(!1), p = Ls((u) => u.filtered.count), v = E(null);
  return /* @__PURE__ */ t.jsxs(Hr, { open: m, children: [
    /* @__PURE__ */ t.jsx($s, { children: /* @__PURE__ */ t.jsx(
      Es,
      {
        autoComplete: "off",
        ref: (u) => {
          v.current = u, typeof c == "function" ? c(u) : c && (c.current = u);
        },
        value: e ? String(e) : void 0,
        placeholder: r,
        className: S("h-9 bg-transparent", a),
        onFocus: () => d(!0),
        onBlur: () => {
          f || d(!1);
        },
        onKeyDown: (u) => {
          i?.(u), !u.defaultPrevented && u.key === "Enter" && (d(!1), v.current?.blur(), o?.(u));
        },
        onValueChange: (u) => s(u)
      }
    ) }),
    /* @__PURE__ */ t.jsx(
      un,
      {
        onMouseEnter: () => x(!0),
        onMouseLeave: () => x(!1),
        onOpenAutoFocus: (u) => u.preventDefault(),
        className: S("p-0 w-(--radix-popover-trigger-width)", {
          "border-0": p === 0
        }),
        align: "start",
        side: "bottom",
        onWheel: (u) => {
          u.stopPropagation();
        },
        onTouchMove: (u) => {
          u.stopPropagation();
        },
        children: /* @__PURE__ */ t.jsx(Rs, { className: "max-h-[140px]", children: n.map((u) => /* @__PURE__ */ t.jsx(
          Is,
          {
            value: u,
            onSelect: (y) => {
              l?.(y), s(y), d(!1);
            },
            className: "cursor-pointer",
            children: u
          },
          u
        )) })
      }
    )
  ] });
}, $e = ({ shouldFilter: e, ...n }) => /* @__PURE__ */ t.jsx(ks, { className: "bg-transparent", shouldFilter: e, children: /* @__PURE__ */ t.jsx(Qr, { ...n }) }), Jr = Object.freeze([
  "Accept",
  "Accept-Encoding",
  "Accept-Language",
  "Authorization",
  "Cache-Control",
  "Connection",
  "Content-Disposition",
  "Content-Encoding",
  "Content-Language",
  "Content-Length",
  "Content-Range",
  "Content-Security-Policy",
  "Content-Type",
  "Cookie",
  "Date",
  "ETag",
  "Expires",
  "Host",
  "If-Modified-Since",
  "Location",
  "Origin",
  "Pragma",
  "Referer",
  "Set-Cookie",
  "User-Agent",
  "X-Requested-With"
]), Wr = ({
  control: e,
  schemaHeaders: n,
  lockedHeaders: s
}) => {
  const { watch: a, formState: r } = oe(), o = a("headers"), i = ot({
    control: e,
    name: "headers",
    defaultValue: { name: "", value: "", active: !1 }
  }), c = n.filter((d) => !o.some((f) => f.name === d.name)).map(({ name: d }) => d), l = i.fields.flatMap((d, f) => !s?.map((p) => p.toLowerCase()).includes(d.name.toLowerCase()) ? [] : [f]), m = s?.map((d) => ({
    name: d,
    id: `locked-${d}`,
    value: "••••••••••",
    active: !0,
    locked: !0
  })) ?? [];
  return /* @__PURE__ */ t.jsxs(_, { defaultOpen: !0, children: [
    /* @__PURE__ */ t.jsxs(Y, { children: [
      /* @__PURE__ */ t.jsx(Fn, { size: 14 }),
      /* @__PURE__ */ t.jsx(H, { children: "Headers" })
    ] }),
    /* @__PURE__ */ t.jsx(q, { className: "CollapsibleContent", children: /* @__PURE__ */ t.jsx("div", { className: "flex flex-col gap-2", children: /* @__PURE__ */ t.jsx("div", { className: "overflow-hidden", children: /* @__PURE__ */ t.jsxs(Ee, { children: [
      m.map((d) => /* @__PURE__ */ t.jsxs(Ct, { children: [
        /* @__PURE__ */ t.jsx(Nt, { asChild: !0, children: /* @__PURE__ */ t.jsxs(
          ge,
          {
            className: "opacity-50 cursor-not-allowed font-mono text-xs min-h-10",
            children: [
              /* @__PURE__ */ t.jsx(zn, { size: 16 }),
              /* @__PURE__ */ t.jsx(I, { value: d.name, disabled: !0 }),
              /* @__PURE__ */ t.jsx("div", { children: d.value })
            ]
          },
          d.id
        ) }),
        /* @__PURE__ */ t.jsx(wt, { alignOffset: 10, side: "bottom", align: "start", children: /* @__PURE__ */ t.jsx("p", { children: "This header is set by the selected authentication." }) })
      ] }, d.id)),
      i.fields.map((d, f) => {
        const x = n.find(
          (w) => w.name === o.at(f)?.name
        ), p = x?.enum && x.enum.length > 0, v = l.includes(f), u = i.getNameInputProps(f), y = i.getValueInputProps(f);
        return /* @__PURE__ */ t.jsxs(
          ge,
          {
            className: S(
              v && "text-amber-600",
              v && !r.dirtyFields.headers?.[f]?.value && "hidden"
            ),
            children: [
              /* @__PURE__ */ t.jsx(
                be,
                {
                  className: S(v && "hidden"),
                  ...i.getCheckboxProps(f)
                }
              ),
              /* @__PURE__ */ t.jsxs(Ct, { children: [
                /* @__PURE__ */ t.jsx(Nt, { asChild: !0, children: /* @__PURE__ */ t.jsx(
                  Vn,
                  {
                    className: S(
                      "text-amber-600",
                      !v && "hidden"
                    ),
                    size: 16
                  }
                ) }),
                /* @__PURE__ */ t.jsx(
                  wt,
                  {
                    alignOffset: 10,
                    side: "bottom",
                    align: "start",
                    children: /* @__PURE__ */ t.jsx("p", { children: "This header will be overwritten by the selected authentication." })
                  }
                )
              ] }),
              /* @__PURE__ */ t.jsx(I, { asChild: !0, children: /* @__PURE__ */ t.jsx(
                $e,
                {
                  ...u,
                  value: String(i.getValue(f, "name")),
                  placeholder: "Name",
                  options: [...c, ...Jr],
                  onChange: (w) => i.setValue(f, "name", w),
                  onSelect: (w) => i.setValue(f, "name", w, { focus: "next" })
                }
              ) }),
              /* @__PURE__ */ t.jsxs("div", { className: "flex items-center gap-2", children: [
                p ? /* @__PURE__ */ t.jsx(I, { asChild: !0, children: /* @__PURE__ */ t.jsx(
                  $e,
                  {
                    ...y,
                    value: String(i.getValue(f, "value")),
                    shouldFilter: !1,
                    options: x.enum ?? [],
                    onChange: (w) => i.setValue(f, "value", w),
                    onSelect: (w) => i.setValue(f, "value", w, { focus: "next" })
                  }
                ) }) : /* @__PURE__ */ t.jsx(
                  I,
                  {
                    placeholder: "Value",
                    autoComplete: "off",
                    ...y
                  }
                ),
                /* @__PURE__ */ t.jsx(
                  at,
                  {
                    ...i.getRemoveButtonProps(f)
                  }
                )
              ] })
            ]
          },
          d.id
        );
      })
    ] }) }) }) })
  ] });
}, Xr = Fs(
  "text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
), Re = xe.forwardRef(({ className: e, ...n }, s) => /* @__PURE__ */ t.jsx(
  Kt.Root,
  {
    ref: s,
    className: S(Xr(), e),
    ...n
  }
));
Re.displayName = Kt.Root.displayName;
const mn = xe.forwardRef(({ className: e, ...n }, s) => /* @__PURE__ */ t.jsx(
  he.Root,
  {
    className: S("grid gap-2", e),
    ...n,
    ref: s
  }
));
mn.displayName = he.Root.displayName;
const pn = xe.forwardRef(({ className: e, ...n }, s) => /* @__PURE__ */ t.jsx(
  he.Item,
  {
    ref: s,
    className: S(
      "aspect-square h-4 w-4 rounded-full border border-primary text-primary ring-offset-background focus:outline-hidden focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50",
      e
    ),
    ...n,
    children: /* @__PURE__ */ t.jsx(he.Indicator, { className: "flex items-center justify-center", children: /* @__PURE__ */ t.jsx(Bn, { className: "h-2.5 w-2.5 fill-current text-current" }) })
  }
));
pn.displayName = he.Item.displayName;
const hn = ({
  identities: e,
  setValue: n,
  value: s
}) => /* @__PURE__ */ t.jsx("div", { className: "w-full overflow-hidden", children: /* @__PURE__ */ t.jsx(
  mn,
  {
    onValueChange: (a) => n(a),
    value: s,
    defaultValue: pe,
    className: "gap-0",
    disabled: e?.length === 0,
    children: [{ id: pe, label: "None" }, ...e ?? []].map(
      (a) => /* @__PURE__ */ t.jsxs(
        Re,
        {
          className: "h-10 items-center border-b font-normal flex gap-4 p-4 cursor-pointer hover:bg-accent/75",
          children: [
            /* @__PURE__ */ t.jsx(pn, { value: a.id, id: a.id }),
            /* @__PURE__ */ t.jsx("span", { children: a.label })
          ]
        },
        a.id
      )
    )
  }
) }), Kr = ({
  onSubmit: e,
  identities: n,
  open: s,
  onOpenChange: a
}) => {
  const [r, o] = k(void 0), [i, c] = k(!1);
  return /* @__PURE__ */ t.jsx(et, { open: s, onOpenChange: a, children: /* @__PURE__ */ t.jsxs(tt, { children: [
    /* @__PURE__ */ t.jsx(nt, { children: "Select an auth identity" }),
    /* @__PURE__ */ t.jsx(_t, { children: "Please select an identity for this request." }),
    /* @__PURE__ */ t.jsx(Ds, { className: "max-h-80 overflow-auto", children: /* @__PURE__ */ t.jsx(
      hn,
      {
        identities: n,
        setValue: o,
        value: r
      }
    ) }),
    /* @__PURE__ */ t.jsxs(qt, { className: "flex flex-col gap-2", children: [
      /* @__PURE__ */ t.jsxs("div", { className: "flex items-center gap-2", children: [
        /* @__PURE__ */ t.jsx(
          be,
          {
            id: "remember",
            checked: i,
            onCheckedChange: (l) => c(
              l === "indeterminate" ? !1 : !!l
            )
          }
        ),
        /* @__PURE__ */ t.jsx(Re, { htmlFor: "remember", children: "Remember my choice" })
      ] }),
      /* @__PURE__ */ t.jsx(
        B,
        {
          onClick: () => e({ identity: r, rememberedIdentity: i }),
          children: "Send"
        }
      )
    ] })
  ] }) });
}, Me = (e) => Math.abs(
  Number.isNaN(Number(e)) ? e.toLowerCase().charCodeAt(0) - 96 : Number.isNaN(Number(e)) ? 0 : Number(e)
), Ge = (e) => e.length > 1 ? Number(e.split("").reduce((n, s) => `${Me(n) + Me(s)}`)) : Me(e), kt = (e, n = {}) => {
  const s = (3 * Ge(e) + 2 * Ge(e) + Ge(e)) % 360, { saturation: a = 75, lightness: r = 60 } = n;
  return `${s}deg ${a}% ${r}%`;
}, He = "data-linked-param", Zr = (e) => {
  const { resolvedTheme: n } = ps();
  return {
    text: kt(
      e,
      n === "light" ? { saturation: 95, lightness: 38 } : {}
    ),
    background: kt(
      e,
      n === "light" ? { saturation: 85, lightness: 40 } : {}
    )
  };
}, fn = ({
  name: e,
  className: n,
  slug: s,
  title: a,
  children: r,
  onClick: o
}) => {
  const i = E(null), c = s?.replace(/[{}]/g, ""), l = e.replace(/[{}]/g, ""), { text: m, background: d } = Zr(l), f = `hsl(${m} / 100%)`, x = `hsl(${d} / 10%)`, p = `hsl(${d} / 50%)`;
  return U(() => {
    if (!c || !i.current) return;
    const v = () => {
      document.querySelectorAll(`[${He}="${c}"]`).forEach((w) => {
        w instanceof HTMLElement && (w.dataset.active = "true");
      });
    }, u = () => {
      document.querySelectorAll(`[${He}="${c}"]`).forEach((w) => {
        w instanceof HTMLElement && (w.dataset.active = "false");
      });
    }, y = i.current;
    return y.addEventListener("mouseenter", v), y.addEventListener("mouseleave", u), () => {
      y.removeEventListener("mouseenter", v), y.removeEventListener("mouseleave", u);
    };
  }, [c]), // biome-ignore lint/a11y/noStaticElementInteractions: Just passing props through
  // biome-ignore lint/a11y/useKeyWithClickEvents: Just passing props through
  /* @__PURE__ */ t.jsx(
    "span",
    {
      [He]: c,
      className: S(
        // This may not contain (inline-)flex or (inline-)block otherwise it breaks the browser's full text search
        "relative transition-all duration-100 rounded-lg",
        "border border-(--border-color) p-0.5 text-(--param-color) bg-(--background-color)",
        "data-[active=true]:border-(--param-color) data-[active=true]:shadow-sm data-[active=true]:bottom-px",
        n
      ),
      title: a,
      suppressHydrationWarning: !0,
      ref: i,
      onClick: o,
      style: {
        "--param-color": f,
        "--border-color": p,
        "--background-color": x
      },
      children: r ?? e
    }
  );
}, Yr = ({
  control: e,
  url: n
}) => {
  const { fields: s } = Ut({
    control: e,
    name: "pathParams"
  }), a = [...s].sort(
    (r, o) => n.indexOf(`{${r.name}}`) - n.indexOf(`{${o.name}}`)
  );
  return /* @__PURE__ */ t.jsx(Ee, { children: a.map((r, o) => /* @__PURE__ */ t.jsxs(ge, { children: [
    /* @__PURE__ */ t.jsx(
      St,
      {
        control: e,
        name: `pathParams.${o}.name`,
        render: () => /* @__PURE__ */ t.jsx("div", { className: "flex items-center", children: /* @__PURE__ */ t.jsx(
          fn,
          {
            slug: r.name,
            name: r.name,
            className: "font-mono text-xs px-2"
          }
        ) })
      }
    ),
    /* @__PURE__ */ t.jsx("div", { className: "flex justify-between items-center col-span-2", children: /* @__PURE__ */ t.jsx(
      St,
      {
        control: e,
        name: `pathParams.${o}.value`,
        render: ({ field: i }) => /* @__PURE__ */ t.jsx(I, { ...i, required: !0, placeholder: "Value" })
      }
    ) })
  ] }, r.id)) });
}, ea = ({
  control: e,
  schemaQueryParams: n
}) => {
  const { watch: s } = oe(), a = s("queryParams"), r = ot({
    control: e,
    name: "queryParams",
    defaultValue: { name: "", value: "", active: !1 }
  }), o = n.map(
    (c) => !!c.isRequired
  ), i = n.length > 0;
  return /* @__PURE__ */ t.jsxs(_, { defaultOpen: !0, children: [
    /* @__PURE__ */ t.jsxs(Y, { children: [
      /* @__PURE__ */ t.jsx(_n, { size: 16 }),
      /* @__PURE__ */ t.jsx(H, { children: "Query Parameters" })
    ] }),
    /* @__PURE__ */ t.jsx(q, { className: "CollapsibleContent", children: /* @__PURE__ */ t.jsx(Ee, { children: r.fields.map((c, l) => {
      const m = n.find(
        (p) => p.name === a.at(l)?.name
      ), d = m?.enum && m.enum.length > 0, f = r.getNameInputProps(l), x = r.getValueInputProps(l);
      return /* @__PURE__ */ t.jsxs(ge, { children: [
        /* @__PURE__ */ t.jsx(be, { ...r.getCheckboxProps(l) }),
        o[l] ? /* @__PURE__ */ t.jsx(I, { asChild: !0, children: /* @__PURE__ */ t.jsxs(
          "label",
          {
            className: "flex items-center cursor-pointer gap-1",
            htmlFor: `queryParams.${l}.active`,
            title: o[l] ? "Required field" : void 0,
            children: [
              a[l]?.name,
              o[l] && /* @__PURE__ */ t.jsx("sup", { children: " *" })
            ]
          }
        ) }) : i ? /* @__PURE__ */ t.jsx(I, { asChild: !0, children: /* @__PURE__ */ t.jsx(
          $e,
          {
            ...f,
            value: String(r.getValue(l, "name")),
            placeholder: "Name",
            options: n.map((p) => p.name),
            onChange: (p) => r.setValue(l, "name", p),
            onSelect: (p) => r.setValue(l, "name", p, { focus: "next" })
          }
        ) }) : /* @__PURE__ */ t.jsx(I, { ...f, placeholder: "Name" }),
        /* @__PURE__ */ t.jsxs("div", { className: "flex justify-between items-center", children: [
          d ? /* @__PURE__ */ t.jsx(I, { asChild: !0, children: /* @__PURE__ */ t.jsx(
            $e,
            {
              ...x,
              value: String(r.getValue(l, "value")),
              shouldFilter: !1,
              options: m.enum ?? [],
              onChange: (p) => r.setValue(l, "value", p),
              onSelect: (p) => r.setValue(l, "value", p, { focus: "next" })
            }
          ) }) : /* @__PURE__ */ t.jsx(
            I,
            {
              placeholder: "Value",
              "aria-label": "Query parameter value",
              ...x
            }
          ),
          /* @__PURE__ */ t.jsx(
            at,
            {
              ...r.getRemoveButtonProps(l)
            }
          )
        ] })
      ] }, c.id);
    }) }) })
  ] });
}, ta = ({
  open: e,
  setOpen: n,
  onSignUp: s,
  onLogin: a,
  onSkip: r
}) => {
  const [o, i] = k(!1), c = () => {
    r?.(o), n(!1);
  };
  return /* @__PURE__ */ t.jsx(et, { open: e, onOpenChange: n, children: /* @__PURE__ */ t.jsxs(tt, { children: [
    /* @__PURE__ */ t.jsx(nt, { children: "Welcome to the Playground!" }),
    /* @__PURE__ */ t.jsx(_t, { children: "The Playground is a tool for developers to test and explore our APIs. To use the Playground, you need to login." }),
    /* @__PURE__ */ t.jsxs(Re, { className: "flex items-center gap-2 font-normal", children: [
      /* @__PURE__ */ t.jsx(
        be,
        {
          checked: o,
          onCheckedChange: (l) => i(l === !0)
        }
      ),
      "Don't show this again"
    ] }),
    /* @__PURE__ */ t.jsxs(qt, { className: "flex gap-2 sm:justify-between", children: [
      /* @__PURE__ */ t.jsx(B, { type: "button", variant: "outline", onClick: c, children: "Skip" }),
      /* @__PURE__ */ t.jsxs("div", { className: "flex gap-2", children: [
        s && /* @__PURE__ */ t.jsx(B, { type: "button", variant: "outline", onClick: s, children: "Sign Up" }),
        a && /* @__PURE__ */ t.jsx(B, { type: "button", variant: "default", onClick: a, children: "Login" })
      ] })
    ] })
  ] }) });
}, na = Ze()(
  Ye(
    (e, n) => ({
      rememberedIdentity: null,
      setRememberedIdentity: (s) => e({ rememberedIdentity: s }),
      getRememberedIdentity: (s) => s.find(
        (a) => a === n().rememberedIdentity
      )
    }),
    {
      name: "identity-storage",
      storage: Bt(() => sessionStorage)
    }
  )
), sa = ({
  path: e,
  renderParam: n
}) => {
  let s = 0;
  return e.split("/").map((a, r, o) => {
    const i = Array.from(a.matchAll(/{([^}]+)}/g)), c = [];
    let l = 0;
    return i.forEach((m) => {
      const [d, f] = m;
      if (!f) return;
      const x = m.index;
      x > l && c.push(
        /* @__PURE__ */ t.jsx(le, { children: a.slice(l, x) }, `text-${l}-${x}`)
      ), c.push(
        /* @__PURE__ */ t.jsx(le, { children: n({ name: f, originalValue: d, index: s++ }) }, `param-${f}`)
      ), l = x + d.length;
    }), l < a.length && c.push(
      /* @__PURE__ */ t.jsx(le, { children: a.slice(l) }, `text-${l}-${a.length}`)
    ), // biome-ignore lint/suspicious/noArrayIndexKey: index should be stable
    /* @__PURE__ */ t.jsxs(le, { children: [
      c,
      r < o.length - 1 && "/",
      /* @__PURE__ */ t.jsx("wbr", {})
    ] }, `${a}-${r}`);
  });
}, ra = ({ url: e }) => {
  const { watch: n, setFocus: s } = oe(), [a] = n(["pathParams"]);
  return /* @__PURE__ */ t.jsx(
    sa,
    {
      path: e,
      renderParam: ({ name: r, originalValue: o, index: i }) => {
        const c = a.find(
          (l) => l.name === r
        )?.value;
        return /* @__PURE__ */ t.jsx(
          fn,
          {
            name: r,
            backgroundOpacity: "0",
            slug: r,
            onClick: () => s(`pathParams.${i}.value`),
            children: c || o
          }
        );
      }
    }
  );
}, aa = () => {
  const { watch: e } = oe(), s = e("queryParams").filter((a) => a.active && a.name).map((a, r, o) => /* @__PURE__ */ t.jsxs(le, { children: [
    a.name,
    "=",
    encodeURIComponent(a.value).replaceAll("%20", "+"),
    r < o.length - 1 && "&",
    /* @__PURE__ */ t.jsx("wbr", {})
  ] }, `${r}-${a.name}`));
  return /* @__PURE__ */ t.jsxs(t.Fragment, { children: [
    s.length > 0 ? "?" : "",
    s
  ] });
}, oa = ({ status: e }) => {
  if (!e) return /* @__PURE__ */ t.jsx("div", { className: "w-2 h-2 rounded-full bg-gray-500" });
  switch (Number(e.toString().slice(0, 1))) {
    case 2:
      return /* @__PURE__ */ t.jsx("div", { className: "w-2 h-2 rounded-full bg-green-500" });
    case 4:
      return /* @__PURE__ */ t.jsx("div", { className: "w-2 h-2 rounded-full bg-yellow-400" });
    case 5:
      return /* @__PURE__ */ t.jsx("div", { className: "w-2 h-2 rounded-full bg-red-500" });
    default:
      return /* @__PURE__ */ t.jsx("div", { className: "w-2 h-2 rounded-full bg-gray-500" });
  }
}, ia = () => {
  const e = E(null), n = E(null), s = E(Date.now()), a = E(null), r = D(() => {
    if (n.current) {
      const o = Date.now() - s.current;
      n.current.nodeValue = o.toFixed();
    }
    a.current = requestAnimationFrame(r);
  }, []);
  return cs(() => {
    const o = e.current;
    if (!o) return;
    const i = document.createTextNode("0");
    return o.appendChild(i), n.current = i, s.current = Date.now(), a.current = requestAnimationFrame(r), () => {
      a.current && cancelAnimationFrame(a.current), n.current && o.contains(i) && o.removeChild(i);
    };
  }, [r]), /* @__PURE__ */ t.jsx("span", { className: "inline-block text-end w-[3ch]", ref: e });
}, la = ({
  status: e,
  time: n,
  size: s,
  isFinished: a,
  progress: r
}) => {
  const o = {
    200: "OK",
    201: "Created",
    202: "Accepted",
    204: "No Content",
    400: "Bad Request",
    401: "Unauthorized",
    403: "Forbidden",
    404: "Not Found",
    405: "Method Not Allowed",
    500: "Internal Server Error"
  };
  return /* @__PURE__ */ t.jsxs("div", { className: "relative shrink-0 flex h-10 text-xs gap-4 px-4 items-center justify-between font-mono border-b", children: [
    /* @__PURE__ */ t.jsxs("div", { className: "flex items-center gap-2", children: [
      /* @__PURE__ */ t.jsx(oa, { status: e }),
      " ",
      e ?? "Sending Request...",
      e ? ` ${o[e]}` : ""
    ] }),
    /* @__PURE__ */ t.jsxs("div", { className: "flex gap-2", children: [
      /* @__PURE__ */ t.jsxs("div", { children: [
        /* @__PURE__ */ t.jsx("span", { className: "text-muted-foreground", children: "Size" }),
        " ",
        /* @__PURE__ */ t.jsx("span", { className: "inline-block text-end min-w-[5ch]", children: s ? ve(s) : "- B" })
      ] }),
      /* @__PURE__ */ t.jsxs("div", { children: [
        /* @__PURE__ */ t.jsx("span", { className: "text-muted-foreground", children: "Time" }),
        " ",
        n !== void 0 ? `${n.toFixed(0)}` : /* @__PURE__ */ t.jsx(ia, {}),
        "ms"
      ] })
    ] }),
    /* @__PURE__ */ t.jsx(
      "div",
      {
        className: "h-full bg-neutral-500/10 absolute left-0 bottom-0 z-10 transition-all duration-200 ease-in-out",
        style: {
          opacity: a ? 0 : 1,
          width: a ? 0 : `${r * 100}%`
        }
      }
    )
  ] });
}, ca = ({
  blob: e,
  fileName: n,
  size: s,
  onDownload: a
}) => {
  const [r, o] = k(null);
  return U(() => {
    const i = URL.createObjectURL(e);
    return o(i), () => {
      URL.revokeObjectURL(i);
    };
  }, [e]), r ? /* @__PURE__ */ t.jsx("div", { className: "p-4 text-center", children: /* @__PURE__ */ t.jsxs("div", { className: "flex flex-col items-center gap-4", children: [
    /* @__PURE__ */ t.jsx("audio", { controls: !0, src: r, className: "w-full max-w-md", children: "Your browser does not support the audio element." }),
    /* @__PURE__ */ t.jsxs(B, { onClick: a, className: "flex items-center gap-2", children: [
      /* @__PURE__ */ t.jsx(Ft, { className: "h-4 w-4" }),
      "Download ",
      n,
      " (",
      ve(s),
      ")"
    ] })
  ] }) }) : /* @__PURE__ */ t.jsx("div", { className: "p-4 text-center", children: /* @__PURE__ */ t.jsx("div", { className: "text-sm text-muted-foreground", children: "Loading audio..." }) });
};
function it(e) {
  if (e === null) return "null";
  if (Array.isArray(e)) {
    if (e.length === 0) return "any[]";
    const n = e[0];
    return n === void 0 ? "any[]" : `${it(n)}[]`;
  }
  return typeof e == "object" ? da(e) : typeof e;
}
function da(e, n = "") {
  const s = ["{"];
  for (const [a, r] of Object.entries(e)) {
    const o = it(r);
    s.push(`  ${a}: ${o};`);
  }
  return s.push("}"), s.join(`
`);
}
function ua(e) {
  return { lines: [`type GeneratedType = ${it(e)};`] };
}
const ma = (e) => Object.entries({
  "application/json": "json",
  "text/json": "json",
  "text/html": "html",
  "text/css": "css",
  "text/javascript": "javascript",
  "application/xml": "xml",
  "application/xhtml+xml": "xhtml"
}).find(
  ([s]) => e.includes(s)
)?.[1], gn = (e) => e.find(([n]) => n.toLowerCase() === "content-type")?.[1] || "", pa = (e) => {
  const n = gn(e);
  return ma(n);
}, ha = (e) => {
  try {
    return JSON.stringify(JSON.parse(e), null, 2);
  } catch {
    return null;
  }
}, fa = (e) => {
  const n = [
    "Content-Type",
    "Content-Length",
    "Authorization",
    "X-RateLimit-Remaining",
    "X-RateLimit-Limit",
    "Cache-Control",
    "ETag"
  ].map((s) => s.toLowerCase());
  return [...e].sort(([s], [a]) => {
    const r = n.indexOf(s.toLowerCase()), o = n.indexOf(a.toLowerCase());
    return r === o ? 0 : r === -1 ? 1 : o === -1 ? -1 : r - o;
  });
}, z = 3, Se = ye(
  "div",
  "grid-cols-subgrid grid border-b col-span-full px-4 py-1.5 font-mono text-xs"
), ue = ye("div", "py-1 break-words"), Te = ({ value: e, header: n }) => {
  const a = ["authorization", "key", "secret", "token"].includes(n.toLowerCase()), [r, o] = k(!a);
  return /* @__PURE__ */ t.jsx(
    ue,
    {
      className: S(
        "max-h-28 overflow-auto",
        a && "cursor-pointer flex group"
      ),
      onClick: () => {
        a && o((i) => !i);
      },
      children: a ? /* @__PURE__ */ t.jsxs(t.Fragment, { children: [
        /* @__PURE__ */ t.jsx(Ns, { secret: e, previewChars: 0, revealed: r }),
        r ? /* @__PURE__ */ t.jsx(Gn, { size: 14, className: S("hidden group-hover:block") }) : /* @__PURE__ */ t.jsx(Hn, { size: 14, className: S("hidden group-hover:block") })
      ] }) : e
    }
  );
}, ga = ({
  body: e = "",
  headers: n,
  request: s,
  size: a,
  isBinary: r = !1,
  fileName: o,
  blob: i
}) => {
  const c = pa(n), l = ha(e), m = l || e, [d, f] = k(
    l ? "formatted" : "raw"
  ), x = qs({
    queryKey: ["types", m],
    queryFn: async () => ua(JSON.parse(m)),
    enabled: d === "types" && !r
  }), p = () => {
    if (i && o) {
      const u = URL.createObjectURL(i), y = document.createElement("a");
      y.href = u, y.download = o, document.body.appendChild(y), y.click(), document.body.removeChild(y), URL.revokeObjectURL(u);
    }
  }, v = fa([...n]);
  return /* @__PURE__ */ t.jsxs(t.Fragment, { children: [
    /* @__PURE__ */ t.jsxs(_, { defaultOpen: !0, children: [
      /* @__PURE__ */ t.jsxs(Y, { children: [
        /* @__PURE__ */ t.jsx(qn, { size: 14 }),
        /* @__PURE__ */ t.jsx(H, { children: "Request Headers" })
      ] }),
      /* @__PURE__ */ t.jsx(q, { children: /* @__PURE__ */ t.jsxs("div", { className: "grid grid-cols-[2fr_3fr] gap-x-6 text-sm", children: [
        s.headers.slice(0, z).map(([u, y]) => /* @__PURE__ */ t.jsxs(Se, { children: [
          /* @__PURE__ */ t.jsx(ue, { children: u }),
          /* @__PURE__ */ t.jsx(Te, { value: y, header: u })
        ] }, u)),
        s.headers.length > z && /* @__PURE__ */ t.jsxs(_, { className: "col-span-full grid-cols-subgrid grid group", children: [
          /* @__PURE__ */ t.jsxs(de, { className: "data-[state=open]:hidden justify-center col-span-2 text-xs text-muted-foreground hover:text-primary border-b h-8 flex items-center gap-2", children: [
            "Show ",
            s.headers.length - z,
            " more headers",
            /* @__PURE__ */ t.jsx(Je, { size: 12, className: "text-muted-foreground" })
          ] }),
          /* @__PURE__ */ t.jsxs(q, { className: "col-span-full grid grid-cols-subgrid", children: [
            s.headers.slice(z).map(([u, y]) => /* @__PURE__ */ t.jsxs(Se, { children: [
              /* @__PURE__ */ t.jsx(ue, { children: u }),
              /* @__PURE__ */ t.jsx(Te, { value: y, header: u })
            ] }, u)),
            /* @__PURE__ */ t.jsxs(de, { className: "justify-center col-span-2 text-xs text-muted-foreground hover:text-primary border-b h-8 flex items-center gap-2", children: [
              "Hide ",
              s.headers.length - z,
              " headers",
              /* @__PURE__ */ t.jsx(
                bt,
                {
                  size: 12,
                  className: "text-muted-foreground"
                }
              )
            ] })
          ] })
        ] })
      ] }) })
    ] }),
    /* @__PURE__ */ t.jsxs(_, { defaultOpen: !0, children: [
      /* @__PURE__ */ t.jsxs(Y, { children: [
        /* @__PURE__ */ t.jsx(Un, { size: 14 }),
        /* @__PURE__ */ t.jsx(H, { children: "Response Headers" })
      ] }),
      /* @__PURE__ */ t.jsx(q, { children: /* @__PURE__ */ t.jsxs("div", { className: "grid grid-cols-[2fr_3fr] gap-x-6 text-sm", children: [
        v.slice(0, z).map(([u, y]) => /* @__PURE__ */ t.jsxs(Se, { children: [
          /* @__PURE__ */ t.jsx(ue, { children: u }),
          /* @__PURE__ */ t.jsx(Te, { value: y, header: u })
        ] }, u)),
        v.length > z && /* @__PURE__ */ t.jsxs(_, { className: "col-span-full grid-cols-subgrid grid group", children: [
          /* @__PURE__ */ t.jsxs(de, { className: "data-[state=open]:hidden justify-center col-span-2 text-xs text-muted-foreground hover:text-primary border-b h-8 flex items-center gap-2", children: [
            "Show ",
            v.length - z,
            " more headers",
            /* @__PURE__ */ t.jsx(Je, { size: 12, className: "text-muted-foreground" })
          ] }),
          /* @__PURE__ */ t.jsxs(q, { className: "col-span-full grid grid-cols-subgrid", children: [
            v.slice(z).map(([u, y]) => /* @__PURE__ */ t.jsxs(Se, { children: [
              /* @__PURE__ */ t.jsx(ue, { children: u }),
              /* @__PURE__ */ t.jsx(Te, { value: y, header: u })
            ] }, u)),
            /* @__PURE__ */ t.jsxs(de, { className: "justify-center col-span-2 text-xs text-muted-foreground hover:text-primary border-b h-8 flex items-center gap-2", children: [
              "Hide ",
              v.length - z,
              " headers",
              /* @__PURE__ */ t.jsx(
                bt,
                {
                  size: 12,
                  className: "text-muted-foreground"
                }
              )
            ] })
          ] })
        ] })
      ] }) })
    ] }),
    /* @__PURE__ */ t.jsxs("div", { className: "flex gap-2 justify-between items-center border-b px-2 flex-0", children: [
      /* @__PURE__ */ t.jsxs(H, { className: "flex items-center gap-2", children: [
        /* @__PURE__ */ t.jsx(Mn, { size: 14 }),
        "Response body"
      ] }),
      l && !r && /* @__PURE__ */ t.jsxs(
        Mt,
        {
          value: d,
          onValueChange: (u) => f(u),
          children: [
            /* @__PURE__ */ t.jsx(Gt, { className: "max-w-32 border-0 bg-transparent", children: /* @__PURE__ */ t.jsx(Ht, { placeholder: "View" }) }),
            /* @__PURE__ */ t.jsxs(Qt, { children: [
              /* @__PURE__ */ t.jsx(Ae, { value: "formatted", children: "Formatted" }),
              /* @__PURE__ */ t.jsx(Ae, { value: "raw", children: "Raw" }),
              /* @__PURE__ */ t.jsx(Ae, { value: "types", children: "Types" })
            ] })
          ]
        }
      )
    ] }),
    /* @__PURE__ */ t.jsx("div", { className: "flex-1", children: r ? i && Mr(gn(n)) ? /* @__PURE__ */ t.jsx(
      ca,
      {
        blob: i,
        fileName: o ?? "audio",
        size: a,
        onDownload: p
      }
    ) : /* @__PURE__ */ t.jsx("div", { className: "p-4 text-center", children: /* @__PURE__ */ t.jsxs("div", { className: "flex flex-col items-center gap-4", children: [
      /* @__PURE__ */ t.jsx("div", { className: "text-lg font-semibold", children: "Binary Content" }),
      /* @__PURE__ */ t.jsx("div", { className: "text-sm text-muted-foreground", children: "This response contains binary data that cannot be displayed as text." }),
      /* @__PURE__ */ t.jsxs(
        B,
        {
          onClick: p,
          className: "flex items-center gap-2",
          disabled: !i,
          children: [
            /* @__PURE__ */ t.jsx(Ft, { className: "h-4 w-4" }),
            "Download ",
            o || "file",
            " (",
            ve(a),
            ")"
          ]
        }
      )
    ] }) }) : /* @__PURE__ */ t.jsx(
      Cs,
      {
        className: "text-xs flex-1",
        embedded: !0,
        fullHeight: !0,
        language: d === "types" ? "typescript" : d === "raw" ? l ? "plain" : c : "json",
        code: (d === "raw" ? e : d === "types" ? x.data?.lines.join(`
`) : m) ?? ""
      }
    ) })
  ] });
}, xa = ({
  queryMutation: e,
  showLongRunningWarning: n,
  onCancel: s,
  tip: a,
  isFinished: r,
  progress: o
}) => /* @__PURE__ */ t.jsxs("div", { className: "flex flex-col overflow-y-auto h-[80vh] bg-muted/50", children: [
  (e.isPending || e.data) && /* @__PURE__ */ t.jsx(
    la,
    {
      status: e.data?.status,
      time: e.data?.time,
      size: e.data?.size,
      isFinished: r,
      progress: o
    }
  ),
  e.error ? /* @__PURE__ */ t.jsx("div", { className: "max-w-2/3 mx-auto mt-20", children: /* @__PURE__ */ t.jsxs(Vs, { children: [
    /* @__PURE__ */ t.jsx(Qn, { size: 24, strokeWidth: 1.5, className: "me-5" }),
    /* @__PURE__ */ t.jsx(Bs, { children: "Request failed" }),
    /* @__PURE__ */ t.jsx(_s, { children: e.error.message || String(e.error) || "Unexpected error" })
  ] }) }) : e.data ? /* @__PURE__ */ t.jsx(
    ga,
    {
      request: e.data.request,
      size: e.data.size,
      headers: e.data.headers,
      body: e.data.body,
      isBinary: e.data.isBinary,
      fileName: e.data.fileName,
      blob: e.data.blob
    }
  ) : e.isPending ? /* @__PURE__ */ t.jsx("div", { className: "grid place-items-center h-full", children: /* @__PURE__ */ t.jsxs("div", { className: "flex flex-col gap-2 items-center mt-20", children: [
    /* @__PURE__ */ t.jsx(zs, {}),
    /* @__PURE__ */ t.jsxs(
      "div",
      {
        className: S(
          "opacity-0 pointer-events-none transition-opacity h-20 text-sm text-muted-foreground duration-300 flex flex-col gap-2 items-center",
          n && "opacity-100 pointer-events-auto"
        ),
        children: [
          "Looks like the request is taking longer than expected.",
          /* @__PURE__ */ t.jsx(
            K,
            {
              type: "button",
              onClick: s,
              size: "sm",
              className: "w-fit",
              variant: "outline",
              children: "Cancel"
            }
          )
        ]
      }
    )
  ] }) }) : /* @__PURE__ */ t.jsx("div", { className: "h-full grid place-items-center", children: /* @__PURE__ */ t.jsxs("div", { className: "flex flex-col gap-4 items-center", children: [
    /* @__PURE__ */ t.jsx(
      Jn,
      {
        size: 64,
        className: "text-muted-foreground",
        strokeWidth: 1.2
      }
    ),
    /* @__PURE__ */ t.jsx("span", { className: "text-[16px] font-semibold text-muted-foreground", children: "Send your first request" }),
    a
  ] }) })
] }), xn = Ze()(
  Ye(
    (e) => ({
      skipLogin: !1,
      setSkipLogin: (n) => e({ skipLogin: n })
    }),
    {
      name: "remember-skip-login",
      storage: Bt(() => sessionStorage)
    }
  )
);
fs(xn);
const pe = "__none", ya = ({
  server: e,
  servers: n = [],
  url: s,
  method: a,
  headers: r = [],
  queryParams: o = [],
  pathParams: i = [],
  defaultBody: c = "",
  examples: l,
  requiresLogin: m = !1,
  onLogin: d,
  onSignUp: f
}) => {
  const { selectedServer: x, setSelectedServer: p } = Vr(
    n.map((h) => ({ url: h }))
  ), [v, u] = k(!1), y = as(), { setRememberedIdentity: w, getRememberedIdentity: R } = na(), [, N] = ds(), { skipLogin: j, setSkipLogin: $ } = xn(), [Q, je] = k(!1), [Ie, te] = k(!1), g = E(void 0), C = Ss(w), T = E(null), { label: A } = Fr("meta+enter", () => {
    T.current?.requestSubmit();
  }), M = s.match(/\{([^}]+)\}/g)?.map((h) => h.slice(1, -1)) ?? [], L = [...i].sort(
    (h, re) => M.indexOf(h.name) - M.indexOf(re.name)
  ), { register: De, control: ne, handleSubmit: Ce, watch: ct, setValue: Fe, ...ze } = bs({
    defaultValues: {
      body: c,
      bodyMode: "text",
      file: null,
      multipartFormFields: [],
      queryParams: o.length > 0 ? o.map((h) => ({
        name: h.name,
        value: h.defaultValue ?? "",
        active: h.defaultActive ?? !1,
        enum: h.enum ?? []
      })) : [{ name: "", value: "", active: !1, enum: [] }],
      pathParams: L.map((h) => ({
        name: h.name,
        value: h.defaultValue ?? ""
      })),
      headers: r.length > 0 ? r.map((h) => ({
        name: h.name,
        value: h.defaultValue ?? "",
        active: h.defaultActive ?? !1
      })) : [{ name: "", value: "", active: !1 }],
      identity: R([
        pe,
        ...y.data?.map((h) => h.id) ?? []
      ])
    }
  }), se = ct("identity"), Cn = Vt(
    () => y.data?.find((h) => h.id === se)?.authorizationFields,
    [y.data, se]
  );
  U(() => {
    se && C.current(se);
  }, [C, se]);
  const J = vs({
    gcTime: 0,
    mutationFn: async (h) => {
      const re = performance.now(), Ve = new window.Headers(
        h.headers.filter((O) => O.name && O.active).map((O) => [O.name, O.value])
      );
      let Ne;
      switch (h.bodyMode) {
        case "file":
          Ne = h.file || void 0, Ve.delete("Content-Type");
          break;
        case "multipart": {
          const O = new FormData();
          h.multipartFormFields?.filter((X) => X.name && X.active).forEach((X) => O.append(X.name, X.value)), Ne = O, Ve.delete("Content-Type");
          break;
        }
        default:
          Ne = h.body || void 0;
          break;
      }
      const W = new Request(
        Lt(e ?? x, s, h),
        {
          method: a,
          headers: Ve,
          body: ["GET", "HEAD"].includes(a.toUpperCase()) ? null : Ne
        }
      );
      h.identity !== pe && await y.data?.find((O) => O.id === h.identity)?.authorizeRequest(W);
      const Be = setTimeout(
        () => te(!0),
        3210
      );
      g.current = new AbortController(), g.current.signal.addEventListener("abort", () => {
        clearTimeout(Be);
      });
      try {
        const O = await fetch(W, {
          cache: "no-store",
          signal: g.current.signal
        });
        clearTimeout(Be), te(!1);
        const X = performance.now() - re, Ln = new URL(W.url), mt = Array.from(O.headers.entries()), pt = O.headers.get("content-type") || "", ht = Ur(pt);
        let we = "", ft, gt;
        ht ? (ft = await O.blob(), gt = Gr(mt, W.url), we = `Binary content (${pt})`) : we = await O.text();
        const xt = O.headers.get("content-length");
        let ae = "";
        switch (h.bodyMode) {
          case "text":
            ae = h.body;
            break;
          case "file":
            ae = `[File: ${h.file?.name ?? "Unknown"}]`;
            break;
          case "multipart":
            ae = `[Multipart Form Data]
`, ae += h.multipartFormFields?.filter((G) => G.name && G.active).map(
              (G) => G.value instanceof File ? `${G.name}: [File: ${G.value.name}]` : `${G.name}: ${G.value}`
            ).join(`
`);
            break;
          default:
            ae = h.body;
            break;
        }
        return {
          status: O.status,
          headers: mt,
          size: xt ? Number(xt) : we.length,
          body: we,
          time: X,
          isBinary: ht,
          fileName: gt,
          blob: ft,
          request: {
            method: W.method.toUpperCase(),
            url: W.url,
            headers: [
              ["Host", Ln.host],
              ["User-Agent", "Zudoku Playground"],
              ...Array.from(W.headers.entries())
            ],
            body: ae
          }
        };
      } catch (O) {
        throw clearTimeout(Be), te(!1), O instanceof TypeError ? new Error(
          "The request failed, possibly due to network issues or CORS policy."
        ) : O;
      }
    }
  }), dt = J.isPending, [Nn, wn] = k(!1);
  U(() => {
    const h = setTimeout(() => wn(dt), 100);
    return () => clearTimeout(h);
  }, [dt]);
  const { isFinished: Sn, progress: Tn } = xs({ isAnimating: Nn });
  U(() => () => {
    g.current?.abort();
  }, []);
  const An = /* @__PURE__ */ t.jsx("div", { className: "inline-block opacity-50 hover:opacity-100 transition", children: e ? /* @__PURE__ */ t.jsx("span", { children: e.replace(/^https?:\/\//, "").replace(/\/$/, "") }) : n.length > 1 && /* @__PURE__ */ t.jsxs(
    Mt,
    {
      onValueChange: (h) => {
        N(() => p(h));
      },
      value: x,
      defaultValue: x,
      children: [
        /* @__PURE__ */ t.jsx(Gt, { className: "p-0 h-fit shadow-none border-none flex-row-reverse bg-transparent text-xs gap-0.5 translate-y-[4px]", children: /* @__PURE__ */ t.jsx(Ht, {}) }),
        /* @__PURE__ */ t.jsx(Qt, { children: n.map((h) => /* @__PURE__ */ t.jsx(Ae, { value: h, children: h.replace(/^https?:\/\//, "").replace(/\/$/, "") }, h)) })
      ]
    }
  ) }), On = m && !j && !Q, Pn = ["POST", "PUT", "PATCH", "DELETE"].includes(
    a.toUpperCase()
  ), [ut, $n] = ws();
  return /* @__PURE__ */ t.jsx(
    js,
    {
      register: De,
      control: ne,
      handleSubmit: Ce,
      watch: ct,
      setValue: Fe,
      ...ze,
      children: /* @__PURE__ */ t.jsx(ys, { delayDuration: 150, children: /* @__PURE__ */ t.jsxs(
        "form",
        {
          ref: T,
          onKeyDown: (h) => {
            h.key === "Enter" && h.target instanceof HTMLInputElement && h.preventDefault();
          },
          onSubmit: Ce((h) => {
            y.data?.length === 0 || h.identity ? J.mutate(h) : u(!0);
          }),
          className: "relative",
          children: [
            /* @__PURE__ */ t.jsx(
              Kr,
              {
                identities: y.data ?? [],
                open: v,
                onOpenChange: u,
                onSubmit: ({ rememberedIdentity: h, identity: re }) => {
                  h && Fe("identity", re ?? pe), u(!1), J.mutate({ ...ze.getValues(), identity: re });
                }
              }
            ),
            /* @__PURE__ */ t.jsx(
              ta,
              {
                open: On,
                setOpen: (h) => {
                  h || je(!0);
                },
                onSkip: (h) => {
                  je(!0), h && $(!0);
                },
                onSignUp: f,
                onLogin: d
              }
            ),
            /* @__PURE__ */ t.jsxs("div", { className: "grid grid-cols-[1fr_1px_1fr] text-sm", children: [
              /* @__PURE__ */ t.jsxs("div", { className: "col-span-3 p-4 border-b flex gap-2 items-stretch", children: [
                /* @__PURE__ */ t.jsxs("div", { className: "flex flex-1 items-center w-full border rounded-md relative overflow-hidden", children: [
                  /* @__PURE__ */ t.jsx("div", { className: "border-r p-2 bg-muted rounded-l-md self-stretch font-semibold font-mono flex items-center", children: a.toUpperCase() }),
                  /* @__PURE__ */ t.jsx("div", { className: "items-center px-2 font-mono text-xs break-all leading-6 relative h-full w-full", children: /* @__PURE__ */ t.jsxs("div", { className: "h-full py-1.5", children: [
                    An,
                    /* @__PURE__ */ t.jsx(ra, { url: s }),
                    /* @__PURE__ */ t.jsx(aa, {})
                  ] }) }),
                  /* @__PURE__ */ t.jsx("div", { className: "px-1", children: /* @__PURE__ */ t.jsx(
                    B,
                    {
                      type: "button",
                      onClick: () => {
                        $n(
                          Lt(
                            e ?? x,
                            s,
                            ze.getValues()
                          ).toString()
                        );
                      },
                      variant: "ghost",
                      size: "icon-xs",
                      className: S(
                        "hover:opacity-100 transition",
                        ut ? "text-emerald-600 opacity-100" : "opacity-50"
                      ),
                      children: ut ? /* @__PURE__ */ t.jsx(Dt, { className: "text-green-500", size: 14 }) : /* @__PURE__ */ t.jsx(Wn, { size: 14 })
                    }
                  ) })
                ] }),
                /* @__PURE__ */ t.jsx(
                  B,
                  {
                    type: "submit",
                    variant: J.isPending ? "destructive" : "default",
                    onClick: (h) => {
                      J.isPending && (g.current?.abort(
                        "Request cancelled by user"
                      ), h.preventDefault());
                    },
                    className: "w-18",
                    children: J.isPending ? "Cancel" : "Send"
                  }
                )
              ] }),
              /* @__PURE__ */ t.jsxs("div", { className: "relative overflow-y-auto h-[80vh]", children: [
                y.data?.length !== 0 && /* @__PURE__ */ t.jsxs(_, { defaultOpen: !0, children: [
                  /* @__PURE__ */ t.jsxs(Y, { children: [
                    /* @__PURE__ */ t.jsx(Xn, { size: 16 }),
                    /* @__PURE__ */ t.jsx(H, { children: "Authentication" })
                  ] }),
                  /* @__PURE__ */ t.jsx(q, { className: "CollapsibleContent", children: /* @__PURE__ */ t.jsx(
                    hn,
                    {
                      value: se,
                      identities: y.data ?? [],
                      setValue: (h) => Fe("identity", h)
                    }
                  ) })
                ] }),
                L.length > 0 && /* @__PURE__ */ t.jsxs(_, { defaultOpen: !0, children: [
                  /* @__PURE__ */ t.jsxs(Y, { children: [
                    /* @__PURE__ */ t.jsx(Kn, { size: 16 }),
                    /* @__PURE__ */ t.jsx(H, { children: "Path Parameters" })
                  ] }),
                  /* @__PURE__ */ t.jsx(q, { className: "CollapsibleContent", children: /* @__PURE__ */ t.jsx(Yr, { url: s, control: ne }) })
                ] }),
                /* @__PURE__ */ t.jsx(ea, { control: ne, schemaQueryParams: o }),
                /* @__PURE__ */ t.jsx(
                  Wr,
                  {
                    control: ne,
                    schemaHeaders: r,
                    lockedHeaders: Cn?.headers
                  }
                ),
                Pn && /* @__PURE__ */ t.jsx(qr, { content: l })
              ] }),
              /* @__PURE__ */ t.jsx("div", { className: "w-full bg-muted-foreground/20" }),
              /* @__PURE__ */ t.jsx(
                xa,
                {
                  queryMutation: J,
                  showLongRunningWarning: Ie,
                  isFinished: Sn,
                  progress: Tn,
                  tip: /* @__PURE__ */ t.jsx("div", { className: "text-xs w-full", children: /* @__PURE__ */ t.jsxs("span", { className: "text-muted-foreground", children: [
                    "Press",
                    " ",
                    /* @__PURE__ */ t.jsx("kbd", { className: "text-foreground border rounded m-0.5 px-1 py-0.5 capitalize", children: A.join(" + ") }),
                    " ",
                    "to send a request"
                  ] }) }),
                  onCancel: () => {
                    g.current?.abort(
                      "Request cancelled by the user"
                    ), te(!1);
                  }
                }
              )
            ] })
          ]
        }
      ) })
    }
  );
}, va = (e) => {
  const [n, s] = k(!1), { isAuthEnabled: a, login: r, signup: o, isPending: i, isAuthenticated: c } = hs();
  return /* @__PURE__ */ t.jsxs(et, { onOpenChange: (l) => s(l), children: [
    /* @__PURE__ */ t.jsx(gs, { asChild: !0, children: e.children ?? /* @__PURE__ */ t.jsx(B, { variant: "ghost", size: "icon-xs", className: "group", children: /* @__PURE__ */ t.jsx(
      Zn,
      {
        className: "fill-muted-foreground group-hover:fill-foreground transition",
        size: 16,
        strokeWidth: 1.5
      }
    ) }) }),
    /* @__PURE__ */ t.jsxs(
      tt,
      {
        className: "max-w-screen-xl! w-full overflow-hidden p-0",
        "aria-describedby": void 0,
        showCloseButton: !0,
        children: [
          /* @__PURE__ */ t.jsx(ms, { children: /* @__PURE__ */ t.jsx(nt, { children: "Playground" }) }),
          n && /* @__PURE__ */ t.jsx(
            ya,
            {
              requiresLogin: a && !c && !i,
              onLogin: () => r(),
              onSignUp: () => o(),
              ...e
            }
          )
        ]
      }
    )
  ] });
}, ba = {
  get: "green",
  post: "blue",
  put: "yellow",
  delete: "red",
  patch: "purple",
  options: "gray",
  head: "gray"
}, Et = ({
  label: e,
  path: n,
  operations: s,
  collapsible: a,
  collapsed: r
}) => ({
  type: "category",
  label: e,
  link: {
    type: "doc",
    path: n,
    file: n,
    label: e
  },
  collapsible: a,
  collapsed: r,
  items: s.map((o) => ({
    type: "link",
    label: o.summary ?? o.path,
    to: `${n}#${o.slug}`,
    badge: {
      label: o.method,
      // biome-ignore lint/style/noNonNullAssertion: is guaranteed to be defined
      color: ba[o.method.toLowerCase()],
      invert: !0
    }
  }))
}), yn = zt(
  void 0
), oo = yn.Provider, ja = () => {
  const e = us(yn);
  if (!e)
    throw new Error("useOasConfig must be used within a OasConfigProvider");
  return e.config;
}, Qe = (e) => ({
  path: e.routePath,
  async lazy() {
    const { OasProvider: n } = await import("./OasProvider-D_N0fTah.js");
    return {
      element: /* @__PURE__ */ t.jsx(
        n,
        {
          basePath: e.basePath,
          version: e.version,
          client: e.client,
          config: e.config
        }
      )
    };
  },
  children: e.routes
}), vn = ({
  path: e,
  tag: n,
  untagged: s
}) => ({
  path: e,
  async lazy() {
    const { OperationList: a } = await import("./OperationList-JcMzwiuc.js");
    return { element: /* @__PURE__ */ t.jsx(a, { tag: n, untagged: s }) };
  }
}), Ca = ({
  render: e,
  path: n
}) => {
  const { type: s, input: a } = ja(), { tag: r } = ts(), o = ns(), i = wr(jn, { type: s, input: a }), {
    data: { schema: c }
  } = Us(i), l = c.tags.at(0);
  return !r && l?.slug ? /* @__PURE__ */ t.jsx(
    ss,
    {
      to: {
        pathname: rs(n, { tag: l.slug }),
        search: o.search
      }
    }
  ) : r && c.tags.some((m) => m.slug === r) ? e(r) : null;
}, Rt = ({ path: e }) => ({
  path: e,
  async lazy() {
    const { OperationList: n } = await import("./OperationList-JcMzwiuc.js");
    return {
      element: /* @__PURE__ */ t.jsx(
        Ca,
        {
          path: e,
          render: (s) => /* @__PURE__ */ t.jsx(n, { tag: s })
        }
      )
    };
  }
}), Ke = (e) => [
  // Category without tagged operations
  vn({
    path: V(e, lt),
    untagged: !0
  }),
  // Schema list route
  {
    path: V(e, "~schemas"),
    lazy: async () => {
      const { SchemaList: n } = await import("./SchemaList-3gK0DopV.js");
      return { element: /* @__PURE__ */ t.jsx(n, {}) };
    }
  }
], Na = (e, n) => {
  const s = V(e, n.at(0) ?? lt);
  return [
    // Redirect to first tag on the index route
    { index: !0, loader: () => es(s) },
    // Create routes for each tag
    ...n.map(
      (a) => vn({
        path: V(e, a),
        tag: a
      })
    ),
    ...Ke(e)
  ];
}, bn = (e) => e.type === "raw" || !Array.isArray(e.input) ? { versions: [], labels: {}, downloadUrls: {} } : {
  versions: e.input.map((n) => n.path),
  labels: Object.fromEntries(
    e.input.map((n) => [n.path, n.label ?? n.path])
  ),
  downloadUrls: Object.fromEntries(
    e.input.map((n) => [n.path, n.downloadUrl])
  )
}, wa = ({
  basePath: e,
  config: n,
  client: s
}) => {
  const a = n.tagPages, { versions: r } = bn(n);
  return a ? (r.length > 1 ? [void 0, ...r] : [void 0]).map((i) => {
    const c = V(e, i);
    return Qe({
      basePath: e,
      version: i,
      routePath: c,
      routes: Na(c, a),
      client: s,
      config: n
    });
  }) : r.length > 0 ? (r.length > 1 ? [void 0, ...r] : [void 0]).map((c) => {
    const l = V(e, c);
    return Qe({
      basePath: e,
      version: c,
      routePath: l,
      routes: [
        Rt({ path: `${l}/:tag?` }),
        ...Ke(l)
      ],
      client: s,
      config: n
    });
  }) : [
    Qe({
      basePath: e,
      routePath: e,
      routes: [
        Rt({ path: `${e}/:tag?` }),
        ...Ke(e)
      ],
      client: s,
      config: n
    })
  ];
}, jn = Er(`
  query GetNavigationOperations($input: JSON!, $type: SchemaType!) {
    schema(input: $input, type: $type) {
      extensions
      tags {
        slug
        name
        extensions
        operations {
          summary
          slug
          method
          operationId
          path
        }
      }
      components {
        schemas {
          __typename
        }
      }
    }
  }
`), lt = "~endpoints", io = (e) => {
  const n = V(e.path), s = new Gs(e);
  return {
    getHead: () => {
      if (e.type === "url" && !e.skipPreload)
        return (Array.isArray(e.input) ? e.input.map((r) => r.input) : [e.input]).map((r) => /* @__PURE__ */ t.jsx(
          "link",
          {
            href: r,
            rel: "preload",
            as: "fetch",
            crossOrigin: "anonymous"
          },
          r
        ));
      if (e.server)
        return /* @__PURE__ */ t.jsx("link", { rel: "preconnect", href: e.server });
    },
    getMdxComponents: () => ({
      OpenPlaygroundButton: ({
        server: a,
        method: r = "get",
        url: o = "/",
        children: i,
        ...c
      }) => {
        if (!a)
          throw new Error("Server is required");
        return /* @__PURE__ */ t.jsx(
          va,
          {
            url: o,
            method: r,
            server: a,
            ...c,
            children: /* @__PURE__ */ t.jsx(K, { className: "gap-2 items-center", variant: "outline", children: i ?? /* @__PURE__ */ t.jsxs(t.Fragment, { children: [
              "Open in Playground",
              /* @__PURE__ */ t.jsx(Yn, { size: 16 })
            ] }) })
          }
        );
      }
    }),
    getNavigation: async (a, r) => {
      if (!jt({ path: n, end: !1 }, a))
        return [];
      const o = jt(
        { path: `${n}/:version?/:tag`, end: !0 },
        a
      );
      try {
        const i = o?.params.version, { versions: c } = bn(e), l = i ?? c.at(0), { type: m } = e, d = Array.isArray(e.input) ? e.input.find((N) => N.path === l)?.input ?? e.input[0]?.input : e.input, f = cn(s, jn, {
          type: m,
          input: d
        }), x = await r.queryClient.ensureQueryData(f), p = new Map(
          x.schema.tags.filter((N) => N.name && N.operations.length > 0).map((N) => {
            if (!N.name)
              throw new Error(`Tag ${N.slug} has no name`);
            const j = V(n, i, N.slug), $ = N.extensions?.["x-zudoku-collapsed"] ?? !e.options?.expandAllTags, Q = N.extensions?.["x-zudoku-collapsible"] ?? !0;
            return [
              N.name,
              Et({
                label: N.extensions?.["x-displayName"] ?? N.name,
                path: j,
                operations: N.operations,
                collapsed: $,
                collapsible: Q
              })
            ];
          })
        ), v = x.schema.extensions?.["x-tagGroups"] ?? [], u = new Set(
          v.flatMap(
            (N) => N.tags.filter((j) => p.has(j))
          )
        ), w = [
          ...v.flatMap(
            (N) => {
              const j = N.tags.map(($) => p.get($)).filter(Boolean);
              return j.length === 0 ? [] : [
                {
                  type: "category",
                  label: N.name,
                  items: j,
                  collapsible: !0,
                  collapsed: !e.options?.expandAllTags
                }
              ];
            }
          ),
          ...Array.from(p.entries()).filter(([N]) => !u.has(N)).map(([, N]) => N)
        ], R = x.schema.tags.find(
          (N) => !N.name
        )?.operations;
        return R && R.length > 0 && w.push(
          Et({
            label: w.length === 0 ? "Endpoints" : "Other endpoints",
            path: V(n, i, lt),
            operations: R,
            collapsed: !e.options?.expandAllTags
          })
        ), x.schema.components?.schemas?.length && w.push({
          type: "link",
          label: "Schemas",
          to: V(n, i, "~schemas")
        }), w;
      } catch {
        return [];
      }
    },
    getRoutes: () => wa({ basePath: n, config: e, client: s })
  };
};
export {
  _ as C,
  ao as G,
  sr as L,
  oo as O,
  va as P,
  Ks as Q,
  on as S,
  b as T,
  lt as U,
  wr as a,
  de as b,
  q as c,
  Er as d,
  Vr as e,
  fn as f,
  bn as g,
  sa as h,
  Cr as i,
  Pt as j,
  so as k,
  rr as l,
  rt as m,
  an as n,
  Hs as o,
  jr as p,
  qe as q,
  Qs as r,
  F as s,
  ro as t,
  ja as u,
  nr as v,
  Ys as w,
  tn as x,
  jn as y,
  io as z
};
