// zudoku/lib/IndexingDialog-CCA1H6gd.js
import { j as e } from "./jsx-runtime-BzflLqGi.js";
import { DialogTrigger as c } from "@radix-ui/react-dialog";
import { useState as u, useCallback as x, useEffect as m } from "react";
import { B as d } from "./Button-qpSb3K7a.js";
import { D as g, a as p, b as h, c as f, d as j, e as D } from "./Dialog-hlvmmQ_c.js";
const C = ({
  total: i,
  current: s,
  barLength: a = 25,
  emptyChar: r = "░",
  filledChar: o = "█"
}) => {
  const t = Math.round(s / i * 100), l = Math.round(t / 100 * a), n = a - l;
  return /* @__PURE__ */ e.jsxs(e.Fragment, { children: [
    o.repeat(l),
    r.repeat(n),
    " ",
    t,
    "% (",
    s,
    "/",
    i,
    ")"
  ] });
}, F = ({ children: i }) => {
  const [s, a] = u({
    status: "idle"
  }), r = x(() => {
    a({ status: "indexing", total: 0, current: 0, path: "" });
    const t = new EventSource("/__z/pagefind-reindex");
    return t.onmessage = (l) => {
      const n = JSON.parse(l.data);
      n.type === "progress" ? a({
        status: "indexing",
        total: n.total,
        current: n.current,
        path: n.path
      }) : n.type === "complete" && (t.close(), n.success ? a({ status: "complete", indexed: n.indexed }) : a({
        status: "error",
        message: n.error ?? "Indexing failed"
      }));
    }, t.onerror = () => {
      t.close(), a({
        status: "error",
        message: "Connection lost during indexing"
      });
    }, () => t.close();
  }, []);
  m(() => {
    if (s.status === "idle")
      return r();
  }, [s.status, r]);
  const o = () => {
    s.status === "complete" && window.location.reload();
  };
  return /* @__PURE__ */ e.jsxs(g, { children: [
    /* @__PURE__ */ e.jsx(c, { asChild: !0, children: i }),
    /* @__PURE__ */ e.jsxs(
      p,
      {
        className: "max-w-sm! top-1/3",
        showCloseButton: !1,
        onInteractOutside: (t) => t.preventDefault(),
        children: [
          /* @__PURE__ */ e.jsxs(h, { children: [
            /* @__PURE__ */ e.jsxs(f, { children: [
              s.status === "indexing" && "Building Search Index",
              s.status === "complete" && "Indexing Complete",
              s.status === "error" && "Indexing Failed",
              s.status === "idle" && "Build Search Index"
            ] }),
            /* @__PURE__ */ e.jsxs(j, { children: [
              s.status === "indexing" && /* @__PURE__ */ e.jsxs(e.Fragment, { children: [
                s.total > 0 && /* @__PURE__ */ e.jsx("div", { className: "font-mono text-sm mb-2", children: /* @__PURE__ */ e.jsx(C, { ...s }) }),
                s.path && /* @__PURE__ */ e.jsx("span", { className: "block text-xs truncate", children: s.path })
              ] }),
              s.status === "complete" && /* @__PURE__ */ e.jsxs(e.Fragment, { children: [
                "Successfully indexed ",
                s.indexed,
                " pages."
              ] }),
              s.status === "error" && /* @__PURE__ */ e.jsx("span", { className: "text-destructive", children: s.message })
            ] })
          ] }),
          /* @__PURE__ */ e.jsx(D, { children: /* @__PURE__ */ e.jsxs("div", { className: "flex justify-end gap-2", children: [
            s.status === "complete" && /* @__PURE__ */ e.jsx(d, { size: "sm", onClick: o, children: "Close and reload" }),
            s.status === "error" && /* @__PURE__ */ e.jsxs(e.Fragment, { children: [
              /* @__PURE__ */ e.jsx(d, { variant: "outline", onClick: o, children: "Cancel" }),
              /* @__PURE__ */ e.jsx(d, { onClick: r, children: "Retry" })
            ] })
          ] }) })
        ]
      }
    )
  ] });
};
export {
  F as default
};
