// zudoku/lib/kotlin-v2plddBQ.js
import { j as s } from "./jsx-runtime-BzflLqGi.js";
const o = (e) => /* @__PURE__ */ s.jsxs(
  "svg",
  {
    xmlns: "http://www.w3.org/2000/svg",
    xmlSpace: "preserve",
    viewBox: "0 0 500 500",
    width: "1em",
    height: "1em",
    ...e,
    children: [
      /* @__PURE__ */ s.jsx("title", { children: "Kotlin" }),
      /* @__PURE__ */ s.jsxs(
        "linearGradient",
        {
          id: "a",
          x1: 500.003,
          x2: -0.097,
          y1: 579.106,
          y2: 1079.206,
          gradientTransform: "translate(.097 -578.99) scale(.9998)",
          gradientUnits: "userSpaceOnUse",
          children: [
            /* @__PURE__ */ s.jsx("stop", { offset: 3e-3, stopColor: "#e44857" }),
            /* @__PURE__ */ s.jsx("stop", { offset: 0.469, stopColor: "#c711e1" }),
            /* @__PURE__ */ s.jsx("stop", { offset: 1, stopColor: "#7f52ff" })
          ]
        }
      ),
      /* @__PURE__ */ s.jsx("path", { d: "M500 500H0V0h500L250 250z", fill: "url(#a)" })
    ]
  }
);
export {
  o as default
};
