// zudoku/lib/MdxPage-Ek2UQGJg.js
import { j as e } from "./jsx-runtime-BzflLqGi.js";
import { s as H } from "./index-BqzS1N22.js";
import { H as R } from "./index.esm-Ca5zvoff.js";
import { CheckIcon as U, CopyIcon as _, ChevronDownIcon as $, Link2Icon as A, ExternalLinkIcon as q, EditIcon as B } from "lucide-react";
import { useState as F, useEffect as G } from "react";
import { a as V } from "./chunk-EPOLDU6W-CLHrVxOG.js";
import { B as j } from "./Button-qpSb3K7a.js";
import { D as W, a as X, b as J, c as l } from "./DropdownMenu-DN0jNrjj.js";
import { C as K } from "./CategoryHeading-DhmodDcq.js";
import { T as O, H as f } from "./Zudoku-BPvTuMy8.js";
import { T as Q } from "./Toc-De-NLOWO.js";
import { b as Y, j as Z, c as ee } from "./ZudokuContext-Mfno-z8f.js";
import { C as se, a as oe, P as ie } from "./ClaudeLogo-BcAJKUho.js";
import { u as ne } from "./useExposedProps-WhS0f5mn.js";
const ae = {
  h2: ({ children: i, id: n }) => /* @__PURE__ */ e.jsx(f, { level: 2, id: n, registerNavigationAnchor: !0, children: i }),
  h3: ({ children: i, id: n }) => /* @__PURE__ */ e.jsx(f, { level: 3, id: n, registerNavigationAnchor: !0, children: i })
}, ve = ({
  mdxComponent: i,
  basePath: n,
  frontmatter: o = {},
  defaultOptions: a,
  __filepath: c,
  tableOfContents: m,
  excerpt: b
}) => {
  const k = Y()?.categoryLabel, y = V(), [T, u] = F(!1), t = o.title, w = o.description ?? b, p = o.category ?? k, M = o.toc === !1 || a?.toc === !1, E = t ?? m.find((s) => s.depth === 1)?.value, I = o.disable_pager ?? o.disablePager ?? a?.disablePager ?? !1, v = o.showLastModified ?? a?.showLastModified ?? !0, r = o.lastModifiedTime ? new Date(o.lastModifiedTime) : null, d = o.suggestEdit !== !1 && (o.suggestEdit ?? a?.suggestEdit), x = d ? d.url.replaceAll("{filePath}", c) : null, L = d ? d.text || "Edit this page" : null, P = o.copyPage !== !1 && a?.copyPage !== !1, C = Z(n, `${y.pathname}.md`), z = async () => {
    const s = await fetch(C);
    if (!s.ok) throw new Error("Failed to fetch markdown");
    const S = await s.text();
    navigator.clipboard.writeText(S), u(!0), setTimeout(() => u(!1), 2e3);
  }, N = m.find((s) => s.depth === 1)?.children ?? // if `title` is provided by frontmatter it does not appear in the table of contents
  m.filter((s) => s.depth === 2), D = !M && N.length > 0, { prev: g, next: h } = ee();
  return G(() => {
    if (process.env.NODE_ENV === "development")
      return window.__getReactRefreshIgnoredExports = ({ id: s }) => {
        if (s.endsWith(c))
          return ["frontmatter", "tableOfContents"];
      }, () => {
        window.__getReactRefreshIgnoredExports = void 0;
      };
  }, [c]), /* @__PURE__ */ e.jsxs(
    "div",
    {
      className: "grid grid-cols-1 xl:grid-cols-(--sidecar-grid-cols) gap-8 justify-between",
      "data-pagefind-filter": "section:markdown",
      "data-pagefind-meta": "section:markdown",
      children: [
        /* @__PURE__ */ e.jsxs(R, { children: [
          /* @__PURE__ */ e.jsx("title", { children: E }),
          w && /* @__PURE__ */ e.jsx("meta", { name: "description", content: w })
        ] }),
        /* @__PURE__ */ e.jsxs(O, { className: "max-w-full xl:w-full xl:max-w-3xl flex-1 shrink pt-(--padding-content-top)", children: [
          (p || t) && /* @__PURE__ */ e.jsxs("header", { className: "flex items-start justify-between gap-4", children: [
            /* @__PURE__ */ e.jsxs("div", { className: "flex-1", children: [
              p && /* @__PURE__ */ e.jsx(K, { children: p }),
              t && /* @__PURE__ */ e.jsx(f, { level: 1, id: H(t), children: t })
            ] }),
            P && /* @__PURE__ */ e.jsxs("div", { className: "flex items-center border rounded-md", children: [
              /* @__PURE__ */ e.jsxs(
                j,
                {
                  variant: "ghost",
                  size: "sm",
                  onClick: z,
                  className: "rounded-r-none border-r gap-2 h-7",
                  children: [
                    T ? /* @__PURE__ */ e.jsx(U, { size: 14, className: "text-emerald-600" }) : /* @__PURE__ */ e.jsx(_, { size: 14 }),
                    /* @__PURE__ */ e.jsx("span", { children: "Copy page" })
                  ]
                }
              ),
              /* @__PURE__ */ e.jsxs(W, { children: [
                /* @__PURE__ */ e.jsx(X, { asChild: !0, children: /* @__PURE__ */ e.jsx(
                  j,
                  {
                    variant: "ghost",
                    size: "icon-xs",
                    className: "rounded-l-none",
                    children: /* @__PURE__ */ e.jsx($, { size: 14 })
                  }
                ) }),
                /* @__PURE__ */ e.jsxs(J, { align: "end", children: [
                  /* @__PURE__ */ e.jsxs(
                    l,
                    {
                      className: "gap-2",
                      onClick: () => void navigator.clipboard.writeText(window.location.href),
                      children: [
                        /* @__PURE__ */ e.jsx(A, { className: "size-4" }),
                        "Copy link to page"
                      ]
                    }
                  ),
                  /* @__PURE__ */ e.jsx(l, { className: "gap-2", asChild: !0, children: /* @__PURE__ */ e.jsxs(
                    "a",
                    {
                      href: C,
                      target: "_blank",
                      rel: "noopener noreferrer",
                      children: [
                        /* @__PURE__ */ e.jsx(q, { className: "size-4" }),
                        "Open Markdown page"
                      ]
                    }
                  ) }),
                  /* @__PURE__ */ e.jsxs(
                    l,
                    {
                      className: "gap-2",
                      onClick: () => {
                        const s = encodeURIComponent(
                          `Help me understand this documentation page: ${window.location.href}`
                        );
                        window.open(
                          `https://claude.ai/new?q=${s}`,
                          "_blank"
                        );
                      },
                      children: [
                        /* @__PURE__ */ e.jsx(se, { className: "size-4" }),
                        "Open in Claude"
                      ]
                    }
                  ),
                  /* @__PURE__ */ e.jsxs(
                    l,
                    {
                      className: "gap-2",
                      onClick: () => {
                        const s = encodeURIComponent(
                          `Help me understand this documentation page: ${window.location.href}`
                        );
                        window.open(
                          `https://chatgpt.com/?q=${s}`,
                          "_blank"
                        );
                      },
                      children: [
                        /* @__PURE__ */ e.jsx(oe, { className: "size-4" }),
                        "Open in ChatGPT"
                      ]
                    }
                  )
                ] })
              ] })
            ] })
          ] }),
          /* @__PURE__ */ e.jsx(
            i,
            {
              components: { ...ne(), ...ae }
            }
          ),
          /* @__PURE__ */ e.jsx("div", { className: "h-16" }),
          v && r || x ? /* @__PURE__ */ e.jsxs("div", { className: "flex justify-between text-xs text-muted-foreground ", children: [
            /* @__PURE__ */ e.jsx("div", {}),
            /* @__PURE__ */ e.jsxs("div", { className: "flex items-center gap-2", children: [
              /* @__PURE__ */ e.jsx("div", { children: x && /* @__PURE__ */ e.jsx(j, { asChild: !0, variant: "ghost", size: "sm", children: /* @__PURE__ */ e.jsxs(
                "a",
                {
                  href: x,
                  target: "_blank",
                  rel: "noopener noreferrer",
                  className: "flex items-center gap-1",
                  children: [
                    /* @__PURE__ */ e.jsx(B, { size: 12 }),
                    L
                  ]
                }
              ) }) }),
              /* @__PURE__ */ e.jsx("div", { children: v && r && /* @__PURE__ */ e.jsxs(
                "div",
                {
                  title: r.toLocaleString(void 0, {
                    dateStyle: "full",
                    timeStyle: "medium"
                  }),
                  children: [
                    "Last modified on",
                    " ",
                    /* @__PURE__ */ e.jsx("time", { dateTime: r.toISOString(), children: r.toLocaleDateString("en-US", {
                      dateStyle: "long"
                    }) })
                  ]
                }
              ) })
            ] })
          ] }) : null,
          !I && /* @__PURE__ */ e.jsxs(e.Fragment, { children: [
            /* @__PURE__ */ e.jsx("div", { className: "h-px bg-border mt-2 mb-6" }),
            /* @__PURE__ */ e.jsx(
              ie,
              {
                prev: g ? { to: g.id, label: g.label ?? "" } : void 0,
                next: h ? { to: h.id, label: h.label ?? "" } : void 0,
                className: "mb-10"
              }
            )
          ] })
        ] }),
        /* @__PURE__ */ e.jsx("div", { className: "hidden xl:block", "data-pagefind-ignore": "all", children: D && /* @__PURE__ */ e.jsx(Q, { entries: N }) })
      ]
    }
  );
};
export {
  ve as MdxPage
};
