// zudoku/lib/Mermaid-CJQJeZ9Q.js
import { j as r } from "./jsx-runtime-BzflLqGi.js";
import { u as f } from "./useQuery-DSKGiCQr.js";
import { useId as g } from "react";
import { c as x } from "./index-DI5SPFK9.js";
import { c as o } from "./cn-5-Gd1Dss.js";
import { z as p } from "./index-DAWHN3cH.js";
import "./ZudokuReactContext-DGJAP1sN.js";
import { S as v } from "./Spinner-CI6bRyZw.js";
const b = x(
  "grid gap-0.5 rounded-lg border text-left text-sm has-data-[slot=alert-action]:relative has-data-[slot=alert-action]:pr-18 has-[>svg]:grid-cols-[auto_1fr] has-[>svg]:gap-x-2 *:[svg]:row-span-2 *:[svg]:translate-y-0.5 *:[svg]:text-current *:[svg:not([class*='size-'])]:size-4 w-full relative group/alert",
  {
    variants: {
      variant: {
        default: "bg-card text-card-foreground",
        info: "bg-blue-50 text-blue-700 border-blue-200 dark:bg-blue-900/45 dark:text-blue-200 dark:border-blue-900/75",
        destructive: "text-destructive bg-card bg-destructive/5 border-destructive/20 *:data-[slot=alert-description]:text-destructive *:[svg]:text-current",
        warning: "text-warning-foreground bg-card bg-warning/5 border-warning/50 *:data-[slot=alert-description]:text-warning-foreground *:[svg]:text-current"
      },
      fit: {
        default: "px-2.5 py-2",
        loose: "p-4"
      }
    },
    defaultVariants: {
      variant: "default",
      fit: "default"
    }
  }
);
function h({
  className: e,
  fit: t,
  variant: a,
  ...s
}) {
  return /* @__PURE__ */ r.jsx(
    "div",
    {
      "data-slot": "alert",
      role: "alert",
      className: o(b({ variant: a, fit: t }), e),
      ...s
    }
  );
}
function w({ className: e, ...t }) {
  return /* @__PURE__ */ r.jsx(
    "div",
    {
      "data-slot": "alert-title",
      className: o(
        "font-medium group-has-[>svg]/alert:col-start-2 [&_a]:hover:text-foreground [&_a]:underline [&_a]:underline-offset-3",
        e
      ),
      ...t
    }
  );
}
function y({
  className: e,
  ...t
}) {
  return /* @__PURE__ */ r.jsx(
    "div",
    {
      "data-slot": "alert-description",
      className: o(
        "text-muted-foreground text-sm text-balance md:text-pretty [&_p:not(:last-child)]:mb-4 [&_a]:hover:text-foreground [&_a]:underline [&_a]:underline-offset-3",
        e
      ),
      ...t
    }
  );
}
let n = null;
const j = () => (n || (n = import("mermaid").then((e) => e.default).catch((e) => {
  throw new Error(
    "Mermaid is not installed. Please install it with: npm install mermaid",
    { cause: e }
  );
})), n), P = ({ chart: e, config: t, ...a }) => {
  const s = g(), i = p(), {
    data: u,
    error: d,
    isPending: c
  } = f({
    queryKey: ["mermaid", e, t, i.resolvedTheme],
    queryFn: async () => {
      const l = await j();
      l.initialize({
        theme: i.resolvedTheme === "dark" ? "dark" : "base",
        ...t
      });
      const { svg: m } = await l.render(s, e);
      return m;
    },
    enabled: typeof window < "u",
    retry: !1
  });
  return d ? /* @__PURE__ */ r.jsxs(h, { className: "flex flex-col gap-2", variant: "destructive", children: [
    /* @__PURE__ */ r.jsx(w, { children: "Mermaid Error" }),
    /* @__PURE__ */ r.jsx(y, { className: "overflow-auto wrap-break-word whitespace-pre-wrap font-mono text-xs", children: d.message })
  ] }) : c ? /* @__PURE__ */ r.jsx(v, {}) : /* @__PURE__ */ r.jsx("div", { ...a, dangerouslySetInnerHTML: { __html: u } });
};
export {
  h as A,
  P as M,
  w as a,
  y as b
};
