// zudoku/lib/OasProvider-D_N0fTah.js
import { j as s } from "./jsx-runtime-BzflLqGi.js";
import { useMemo as d } from "react";
import { O as f } from "./chunk-EPOLDU6W-CLHrVxOG.js";
import { j } from "./ZudokuContext-Mfno-z8f.js";
import { g as c, O as x, G as O } from "./index-Ckdk300F.js";
const y = ({
  basePath: n,
  version: e,
  config: o,
  client: a
}) => {
  const p = d(() => {
    const {
      versions: i,
      labels: u,
      downloadUrls: l
    } = c(o), t = e ?? i.at(0), m = Object.fromEntries(
      i.map((r) => [
        r,
        {
          path: j(n, r),
          label: u[r] ?? r,
          downloadUrl: l[r]
        }
      ])
    );
    return {
      config: {
        ...o,
        version: t,
        versions: m,
        input: (() => {
          if (!Array.isArray(o.input))
            return o.input;
          const r = t ? o.input.find((v) => v.path === t) : o.input[0];
          if (!r)
            throw new Error(`No input found for version: ${t}`);
          return r.input;
        })()
      }
    };
  }, [o, n, e]);
  return /* @__PURE__ */ s.jsx(x, { value: p, children: /* @__PURE__ */ s.jsx(O, { client: a, children: /* @__PURE__ */ s.jsx(f, {}) }) });
};
export {
  y as OasProvider
};
