// zudoku/lib/RouteGuard-DnAFajw9.js
import { j as e } from "./jsx-runtime-BzflLqGi.js";
import { H as D } from "./index.esm-Ca5zvoff.js";
import { createContext as b, use as A, useMemo as B, useCallback as E, useEffect as P } from "react";
import { u as T, a as w, m as y, c as O, O as k } from "./chunk-EPOLDU6W-CLHrVxOG.js";
import { B as f } from "./Button-qpSb3K7a.js";
import { D as H, a as L, b as S, c as F, d as N, e as Z } from "./Dialog-hlvmmQ_c.js";
import { u as z } from "./hook-BqOwTUgE.js";
import { a as I } from "./ZudokuContext-Mfno-z8f.js";
import { Z as _ } from "./invariant-B_t_F2s_.js";
const G = b(!1), M = "protected", $ = ({
  open: t,
  onCancel: o,
  onLogin: h,
  onRegister: i
}) => /* @__PURE__ */ e.jsx(H, { open: t, onOpenChange: (a) => !a && o(), children: /* @__PURE__ */ e.jsxs(L, { children: [
  /* @__PURE__ */ e.jsx(S, { children: /* @__PURE__ */ e.jsx(F, { children: "Login to continue" }) }),
  /* @__PURE__ */ e.jsx(N, { children: "Please login to access this page." }),
  /* @__PURE__ */ e.jsxs(Z, { children: [
    /* @__PURE__ */ e.jsx(f, { variant: "outline", onClick: o, children: "Cancel" }),
    /* @__PURE__ */ e.jsx("div", { className: "w-full" }),
    /* @__PURE__ */ e.jsx(f, { variant: "secondary", onClick: i, children: "Register" }),
    /* @__PURE__ */ e.jsx(f, { onClick: h, children: "Login" })
  ] })
] }) }), q = ({ isProtectedRoute: t }) => /* @__PURE__ */ e.jsxs(e.Fragment, { children: [
  t && /* @__PURE__ */ e.jsx(D, { children: /* @__PURE__ */ e.jsx(
    "meta",
    {
      name: "pagefind",
      "data-pagefind-filter": `section:${M}`,
      content: "true"
    }
  ) }),
  /* @__PURE__ */ e.jsx(k, {})
] }), te = () => {
  const t = z(), o = I(), h = T(), i = w(), a = A(G), c = B(
    () => ({ auth: t, context: o }),
    [t, o]
  ), { protectedRoutes: p } = o, r = E(
    (n) => {
      if (p) {
        for (const [l, v] of Object.entries(p))
          if (y({ path: l, end: !0 }, n))
            return v;
      }
    },
    [p]
  ), g = r(i.pathname), m = g !== void 0, C = g?.(c) ?? !0, u = m && !C, s = O(({ nextLocation: n }) => {
    if (a) return !1;
    const l = r(n.pathname);
    return l !== void 0 && !l(c);
  }), x = s.state === "blocked", d = x ? s.location.pathname : void 0;
  if (P(() => {
    if (!t.isAuthenticated || !d) return;
    const n = r(d);
    (!n || n(c)) && s.proceed?.();
  }, [
    t.isAuthenticated,
    d,
    s,
    c,
    r
  ]), a)
    return /* @__PURE__ */ e.jsx(q, { isProtectedRoute: m });
  if (m && !t.isAuthEnabled)
    throw new _("Authentication is not enabled", {
      title: "Authentication is not enabled",
      developerHint: "To use protectedRoutes you need authentication to be enabled"
    });
  if (u && t.isPending && typeof window < "u")
    return null;
  const R = u || x, j = d ?? i.pathname;
  return /* @__PURE__ */ e.jsxs(e.Fragment, { children: [
    !u && /* @__PURE__ */ e.jsx(k, {}),
    /* @__PURE__ */ e.jsx(
      $,
      {
        open: R,
        onCancel: u ? () => h(-1) : () => s.reset?.(),
        onLogin: () => void t.login({ redirectTo: j }),
        onRegister: () => void t.signup({ redirectTo: j })
      }
    )
  ] });
};
export {
  G as B,
  te as R,
  M as S
};
