// zudoku/lib/SchemaView-Cm6nRtpo.js
import { j as t } from "./jsx-runtime-BzflLqGi.js";
import { ChevronUpIcon as _, ChevronDownIcon as J, ChevronsLeftRightIcon as U, MinusIcon as W, PlusIcon as G, RefreshCcwDotIcon as K, InfoIcon as X } from "lucide-react";
import { useState as f, isValidElement as Q, Fragment as P } from "react";
import { I as S, a as g, b as $, c as Y, F as N, d as h, e as F, f as Z, g as H, h as ee, i as te } from "./Frame-DKlOmSkU.js";
import { b as u, I as R, M as V } from "./Zudoku-BPvTuMy8.js";
import { B as se } from "./Button-qpSb3K7a.js";
import { c as j } from "./cn-5-Gd1Dss.js";
import * as E from "@radix-ui/react-collapsible";
import { Button as re } from "./ui/Button.js";
import { Badge as ne } from "./ui/Badge.js";
import { Frame as ie, FramePanel as oe } from "./ui/Frame.js";
import { C as le, S as b } from "./circular-BvNS-yVz.js";
const Fe = ({
  name: e,
  children: s
}) => /* @__PURE__ */ t.jsx("span", { "data-pagefind-meta": e, className: "sr-only", children: s }), ae = (e, s) => e.reduce(
  (r, i) => {
    const o = s(i);
    return r[o] || (r[o] = []), r[o].push(i), r;
  },
  {}
), M = ({
  schema: e,
  hideDescription: s = !1
}) => /* @__PURE__ */ t.jsx("div", { className: "flex flex-col gap-1", children: /* @__PURE__ */ t.jsxs("div", { children: [
  /* @__PURE__ */ t.jsx("span", { className: "text-muted-foreground", children: "Const value: " }),
  /* @__PURE__ */ t.jsx(u, { className: "border rounded px-1 font-mono", children: e.const }),
  !s && e.description && /* @__PURE__ */ t.jsx("div", { className: "text-muted-foreground", children: e.description })
] }) }), v = ({
  values: e,
  className: s,
  maxVisibleValues: r = 8
}) => {
  const [i, o] = f(!1);
  if (!e.length) return null;
  const n = e.length > r, l = n && !i ? e.slice(0, r) : e;
  return /* @__PURE__ */ t.jsxs("div", { className: j("flex flex-wrap gap-1.5", s), children: [
    /* @__PURE__ */ t.jsx("span", { className: "text-muted-foreground", children: "Enum values:" }),
    l.map((d) => /* @__PURE__ */ t.jsx("div", { children: /* @__PURE__ */ t.jsx(u, { className: "border rounded-sm px-1 font-mono", children: d }) }, d)),
    n && /* @__PURE__ */ t.jsx(
      se,
      {
        variant: "ghost",
        size: "sm",
        className: "h-fit px-0",
        onClick: () => o(!i),
        children: i ? /* @__PURE__ */ t.jsxs("div", { className: "flex items-center gap-1", children: [
          /* @__PURE__ */ t.jsx(_, { size: 12 }),
          /* @__PURE__ */ t.jsx("span", { className: "text-muted-foreground", children: "show less" })
        ] }) : /* @__PURE__ */ t.jsxs("div", { className: "flex items-center gap-1", children: [
          /* @__PURE__ */ t.jsx(J, { size: 12 }),
          /* @__PURE__ */ t.jsxs("span", { className: "text-muted-foreground", children: [
            "show ",
            e.length - r,
            " more"
          ] })
        ] })
      }
    )
  ] });
}, de = ({ pattern: e }) => {
  const [s, r] = f(!1), i = e.length > 20, o = i ? `${e.slice(0, 20)}…` : e;
  return /* @__PURE__ */ t.jsxs(
    R,
    {
      className: j("text-xs", i && "cursor-pointer"),
      onClick: () => r(!s),
      selectOnClick: !1,
      children: [
        s ? e : o,
        i && /* @__PURE__ */ t.jsx("button", { type: "button", className: "p-1 translate-y-[2px]", children: !s && /* @__PURE__ */ t.jsx(U, { size: 12 }) })
      ]
    }
  );
}, ce = (e) => {
  if (!e) return [];
  const s = e.type === "array" && typeof e.items == "object" ? e.items : void 0;
  return [
    s?.type ? Array.isArray(s.type) ? `(${s.type.join(" | ")})[]` : `${s.type}[]` : Array.isArray(e.type) ? e.type.join(" | ") : e.type,
    e.enum && "enum",
    e.const && "const",
    e.format,
    s?.contentMediaType,
    e.minimum !== void 0 && `min: ${e.minimum}`,
    e.maximum !== void 0 && `max: ${e.maximum}`,
    e.minLength !== void 0 && `minLength: ${e.minLength}`,
    e.maxLength !== void 0 && `maxLength: ${e.maxLength}`,
    e.minItems !== void 0 && `minItems: ${e.minItems}`,
    e.maxItems !== void 0 && `maxItems: ${e.maxItems}`,
    e.minProperties !== void 0 && `minProps: ${e.minProperties}`,
    e.maxProperties !== void 0 && `maxProps: ${e.maxProperties}`,
    e.uniqueItems && "unique",
    e.readOnly && "readOnly",
    e.writeOnly && "writeOnly",
    e.deprecated && "deprecated",
    e.pattern && /* @__PURE__ */ t.jsxs(t.Fragment, { children: [
      "pattern: ",
      /* @__PURE__ */ t.jsx(de, { pattern: e.pattern })
    ] })
  ];
}, I = ({
  schema: e,
  extraItems: s = [],
  className: r
}) => {
  const i = [...ce(e), ...s].flatMap(
    (o) => typeof o == "string" || Q(o) ? o : []
  );
  return /* @__PURE__ */ t.jsx("span", { className: r, children: i.map((o, n) => (
    // biome-ignore lint/suspicious/noArrayIndexKey: index should be stable
    /* @__PURE__ */ t.jsxs("span", { className: "text-muted-foreground", children: [
      o,
      n < i.length - 1 && /* @__PURE__ */ t.jsx("span", { className: "text-muted-foreground/50", children: " · " })
    ] }, n)
  )) });
}, O = ({
  schema: e
}) => {
  const s = e.examples?.at(0), r = e.default;
  return s === void 0 && r === void 0 ? null : /* @__PURE__ */ t.jsxs("div", { className: "flex flex-col gap-1", children: [
    s !== void 0 && /* @__PURE__ */ t.jsxs("div", { children: [
      /* @__PURE__ */ t.jsx("span", { className: "text-muted-foreground", children: "Example: " }),
      /* @__PURE__ */ t.jsx(u, { className: "border rounded-sm px-1 font-mono", children: typeof s == "object" || typeof s == "boolean" ? JSON.stringify(s) : s })
    ] }),
    r !== void 0 && /* @__PURE__ */ t.jsxs("div", { children: [
      /* @__PURE__ */ t.jsx("span", { className: "text-muted-foreground", children: "Default: " }),
      /* @__PURE__ */ t.jsx(u, { className: "border rounded-sm px-1 font-mono", children: typeof r == "object" || typeof r == "boolean" ? JSON.stringify(r) : r })
    ] })
  ] });
}, B = (e) => typeof e == "string" && ["string", "number", "boolean", "integer", "null"].includes(e) || Array.isArray(e) && e.every(B), m = (e) => e.type === "array" || // schema.type might be an array of types, so we need to check if "array" is one of them
Array.isArray(e.type) && e.type.includes("array"), k = (e) => e && (e.type === "object" && Object.keys(e.properties ?? {}).length > 0 || e.type === "array" && typeof e.items == "object" && (!e.items.type || e.items.type === "object")), z = (e) => typeof e == "string" && (e.startsWith(le) || e.startsWith(b)), w = (e) => m(e) && "items" in e && z(e.items), pe = (e) => {
  if (typeof e == "string")
    return e.startsWith(b) ? e.slice(b.length).split("/").pop() : e.split(":")[1];
}, q = ({ circularProp: e }) => /* @__PURE__ */ t.jsxs(
  R,
  {
    className: "inline-flex items-center gap-1.5 text-xs translate-y-0.5",
    selectOnClick: !1,
    children: [
      /* @__PURE__ */ t.jsx(K, { size: 13 }),
      /* @__PURE__ */ t.jsx("span", { children: e ? `${e} (circular)` : "circular" })
    ]
  }
), xe = ({
  name: e,
  schema: s,
  group: r,
  defaultOpen: i = !1,
  showCollapseButton: o = !0
}) => {
  const [n, l] = f(i);
  if (z(s))
    return /* @__PURE__ */ t.jsx(S, { children: /* @__PURE__ */ t.jsxs(g, { className: "gap-y-2", children: [
      /* @__PURE__ */ t.jsxs("div", { children: [
        /* @__PURE__ */ t.jsx($, { className: "inline me-2", children: /* @__PURE__ */ t.jsx("code", { children: e }) }),
        "​",
        /* @__PURE__ */ t.jsx(
          I,
          {
            className: "inline",
            schema: s,
            extraItems: [
              r !== "optional" && /* @__PURE__ */ t.jsx("span", { className: "text-primary", children: "required" }),
              /* @__PURE__ */ t.jsx(q, {}, "circular-ref")
            ]
          }
        )
      ] }),
      /* @__PURE__ */ t.jsx(O, { schema: s })
    ] }) });
  const d = !!((s.allOf || s.anyOf || s.oneOf || k(s) || m(s) && "items" in s && k(s.items) || s.additionalProperties) && !w(s)), c = !!(s.description || "items" in s && s.items?.enum || s.const || s.enum || s.example !== void 0 || s.default !== void 0);
  return /* @__PURE__ */ t.jsxs(S, { children: [
    /* @__PURE__ */ t.jsxs(g, { className: "gap-y-2", children: [
      /* @__PURE__ */ t.jsxs("div", { children: [
        /* @__PURE__ */ t.jsx($, { className: "inline me-2", children: d ? /* @__PURE__ */ t.jsx(
          "button",
          {
            onClick: () => l(!n),
            type: "button",
            className: "hover:underline",
            children: /* @__PURE__ */ t.jsx("code", { children: e })
          }
        ) : /* @__PURE__ */ t.jsx("code", { children: e }) }),
        "​",
        /* @__PURE__ */ t.jsx(
          I,
          {
            className: "inline",
            schema: s,
            extraItems: [
              r !== "optional" && /* @__PURE__ */ t.jsx("span", { className: "text-primary", children: "required" }),
              w(s) && /* @__PURE__ */ t.jsx(
                q,
                {
                  circularProp: pe(s.items)
                }
              )
            ]
          }
        )
      ] }),
      c && /* @__PURE__ */ t.jsxs("div", { className: "flex flex-col gap-1.5", children: [
        s.description && /* @__PURE__ */ t.jsx(V, { className: "prose-sm", content: s.description }),
        "items" in s && s.items?.enum && /* @__PURE__ */ t.jsx(v, { values: s.items.enum }),
        s.const && /* @__PURE__ */ t.jsx(M, { schema: s, hideDescription: !0 }),
        s.enum && /* @__PURE__ */ t.jsx(v, { values: s.enum }),
        /* @__PURE__ */ t.jsx(O, { schema: s })
      ] })
    ] }),
    d && o && /* @__PURE__ */ t.jsx(Y, { className: "self-start", children: /* @__PURE__ */ t.jsx(
      re,
      {
        variant: "ghost",
        size: "icon",
        className: "rounded-full",
        onClick: () => l(!n),
        "aria-label": "Toggle properties",
        children: n ? /* @__PURE__ */ t.jsx(W, { size: 16 }) : /* @__PURE__ */ t.jsx(G, { size: 16 })
      }
    ) }),
    d && /* @__PURE__ */ t.jsx(
      E.Root,
      {
        defaultOpen: i,
        open: n,
        onOpenChange: l,
        className: j("w-full", !n && "contents"),
        children: /* @__PURE__ */ t.jsx(E.Content, { asChild: !0, children: /* @__PURE__ */ t.jsx(g, { children: s.anyOf || s.oneOf || s.type === "object" ? /* @__PURE__ */ t.jsx(x, { schema: s }) : m(s) && "items" in s ? /* @__PURE__ */ t.jsx(x, { schema: s.items }) : null }) })
      }
    )
  ] });
}, D = (e) => {
  const s = e.oneOf ?? e.anyOf ?? [];
  return e.properties && Object.keys(e.properties).length > 0 ? s.map((r) => !r.properties && !r.type && !r.oneOf && !r.anyOf ? {
    ...r,
    type: "object",
    properties: e.properties,
    required: r.required ?? e.required
  } : r) : s;
}, ue = (e) => {
  if (Array.isArray(e.oneOf)) return "exactly-one";
  const s = e.discriminator?.propertyName;
  if (!s) return "at-least-one";
  const r = D(e), i = /* @__PURE__ */ new Set();
  for (const o of r) {
    const n = o.properties?.[s], l = n?.const ?? (Array.isArray(n?.enum) && n.enum.length === 1 ? String(n.enum[0]) : void 0);
    if (l == null || i.has(String(l))) return "at-least-one";
    i.add(String(l));
  }
  return "exactly-one";
}, A = (e, s) => s.title?.trim() || `Variant ${e + 1}`, me = (e, s) => {
  const r = [];
  e.type && r.push(
    `type = ${Array.isArray(e.type) ? e.type.join("|") : e.type}`
  );
  const i = s?.discriminator?.propertyName;
  if (i) {
    const n = e.properties?.[i], l = n?.const ?? (Array.isArray(n?.enum) && n.enum.length === 1 ? n.enum[0] : void 0);
    l !== void 0 && r.push(`${i}=${JSON.stringify(l)}`);
  }
  const o = (e.required ?? []).filter((n) => n !== i);
  if (o.length) {
    const n = o.slice(0, 3).join(", "), l = o.length > 3 ? ` +${o.length - 3} more` : "";
    r.push(`requires: ${n}${l}`);
  }
  return r;
}, fe = ({
  variants: e,
  schema: s,
  selectedVariant: r,
  onSelectVariant: i
}) => {
  const o = e.map((n, l) => ({
    label: A(l, n),
    guards: me(n, s)
  }));
  return /* @__PURE__ */ t.jsxs("div", { className: "flex flex-col gap-2 text-sm", children: [
    /* @__PURE__ */ t.jsx("h4", { className: "font-medium", children: "Decision Table" }),
    /* @__PURE__ */ t.jsx("div", { className: "border rounded-md overflow-hidden", children: /* @__PURE__ */ t.jsxs("table", { className: "w-full", children: [
      /* @__PURE__ */ t.jsx("thead", { children: /* @__PURE__ */ t.jsxs("tr", { className: "border-b bg-muted/50", children: [
        /* @__PURE__ */ t.jsx("th", { className: "text-left p-2 font-medium", children: "Variant" }),
        /* @__PURE__ */ t.jsx("th", { className: "text-left p-2 font-medium", children: "Matching Criteria" })
      ] }) }),
      /* @__PURE__ */ t.jsx("tbody", { className: "divide-y", children: o.map((n) => /* @__PURE__ */ t.jsxs("tr", { className: "hover:bg-muted/30", children: [
        /* @__PURE__ */ t.jsx("td", { className: "p-2 font-medium", children: /* @__PURE__ */ t.jsx(
          "button",
          {
            type: "button",
            className: j(
              "hover:underline",
              r === n.label && "text-primary"
            ),
            onClick: () => i(n.label),
            children: n.label
          }
        ) }),
        /* @__PURE__ */ t.jsx("td", { className: "p-2 text-muted-foreground text-xs", children: n.guards.length > 0 ? n.guards.join(" · ") : "No specific criteria" })
      ] }, n.label)) })
    ] }) })
  ] });
}, je = ({
  schema: e,
  cardHeader: s
}) => {
  const r = Array.isArray(e.oneOf) ? "oneOf" : Array.isArray(e.anyOf) ? "anyOf" : void 0, i = r ? D(e) : [], [o, n] = f(
    () => i[0] ? A(0, i[0]) : ""
  );
  if (!r) return null;
  const d = ue(e) === "exactly-one" ? /* @__PURE__ */ t.jsxs(t.Fragment, { children: [
    "Exactly one variant ",
    /* @__PURE__ */ t.jsx("b", { children: "must match" }),
    "."
  ] }) : /* @__PURE__ */ t.jsxs(t.Fragment, { children: [
    "At least one variant ",
    /* @__PURE__ */ t.jsx("b", { children: "must match" }),
    ". Multiple variants",
    " ",
    /* @__PURE__ */ t.jsx("i", { children: "may match" }),
    " simultaneously."
  ] }), c = i.findIndex(
    (p, y) => A(y, p) === o
  ), a = c >= 0 ? i[c] : null;
  return /* @__PURE__ */ t.jsxs(ie, { children: [
    s,
    /* @__PURE__ */ t.jsxs(oe, { className: "text-sm flex flex-col gap-4", children: [
      /* @__PURE__ */ t.jsxs("div", { className: "flex items-center gap-2", children: [
        /* @__PURE__ */ t.jsx(ne, { variant: "outline", children: r }),
        /* @__PURE__ */ t.jsx("div", { className: "flex-1 p-2", children: /* @__PURE__ */ t.jsx("span", { className: "text-sm", children: d }) })
      ] }),
      /* @__PURE__ */ t.jsx(
        fe,
        {
          variants: i,
          schema: e,
          selectedVariant: o,
          onSelectVariant: n
        }
      ),
      /* @__PURE__ */ t.jsxs("strong", { children: [
        "Properties for ",
        o,
        ":"
      ] }),
      a && /* @__PURE__ */ t.jsx(x, { schema: a })
    ] })
  ] });
}, ye = (e) => e && /* @__PURE__ */ t.jsx(
  V,
  {
    className: "text-sm leading-normal line-clamp-4",
    content: e
  }
), ge = (e, s, r) => {
  const i = /* @__PURE__ */ t.jsxs(t.Fragment, { children: [
    /* @__PURE__ */ t.jsx("span", { className: "text-sm text-muted-foreground", children: /* @__PURE__ */ t.jsx(I, { schema: e }) }),
    e.enum && /* @__PURE__ */ t.jsx(v, { values: e.enum }),
    ye(e.description),
    /* @__PURE__ */ t.jsx(O, { schema: e })
  ] });
  return r ? /* @__PURE__ */ t.jsx("div", { className: "space-y-2 p-4", children: i }) : /* @__PURE__ */ t.jsxs(N, { children: [
    s,
    /* @__PURE__ */ t.jsx(h, { className: "space-y-2", children: i })
  ] });
}, x = ({
  schema: e,
  defaultOpen: s = !1,
  cardHeader: r,
  embedded: i
}) => {
  if (!e || Object.keys(e).length === 0)
    return /* @__PURE__ */ t.jsxs(N, { children: [
      r,
      /* @__PURE__ */ t.jsx(h, { children: /* @__PURE__ */ t.jsx("div", { className: "text-sm text-muted-foreground italic", children: "No data returned" }) })
    ] });
  if (e.const)
    return /* @__PURE__ */ t.jsx(M, { schema: e });
  if (Array.isArray(e.oneOf) || Array.isArray(e.anyOf))
    return /* @__PURE__ */ t.jsx(je, { schema: e, cardHeader: r });
  if (B(e.type))
    return ge(e, r, i);
  if (m(e) && typeof e.items == "object") {
    const n = {
      type: "object",
      properties: { "": e }
    };
    return /* @__PURE__ */ t.jsx(x, { schema: n, cardHeader: r, defaultOpen: !0 });
  }
  const o = typeof e.additionalProperties == "object" && /* @__PURE__ */ t.jsx(x, { schema: e.additionalProperties, embedded: !0 });
  if (e.type === "object") {
    const n = ae(
      Object.entries(e.properties ?? {}),
      ([a, p]) => p.deprecated ? "deprecated" : e.required?.includes(a) ? "required" : "optional"
    ), c = ["required", "optional", "deprecated"].flatMap((a) => {
      const p = n[a];
      return p ? { group: a, properties: p } : [];
    }).map(({ group: a, properties: p }, y) => /* @__PURE__ */ t.jsxs(P, { children: [
      y > 0 && /* @__PURE__ */ t.jsx(F, {}),
      /* @__PURE__ */ t.jsx(Z, { className: "overflow-clip", children: p.map(([C, L], T) => /* @__PURE__ */ t.jsxs(P, { children: [
        T > 0 && /* @__PURE__ */ t.jsx(F, {}),
        /* @__PURE__ */ t.jsx(
          xe,
          {
            name: C,
            schema: L,
            group: a,
            defaultOpen: s
          }
        )
      ] }, C)) })
    ] }, a));
    return i ? c : /* @__PURE__ */ t.jsxs(N, { children: [
      r,
      e.description && /* @__PURE__ */ t.jsx(H, { children: /* @__PURE__ */ t.jsx(ee, { children: e.description }) }),
      (c.length > 0 || o) && /* @__PURE__ */ t.jsxs(h, { className: "p-0!", children: [
        c,
        o
      ] }),
      e.additionalProperties === !0 && /* @__PURE__ */ t.jsx(te, { children: /* @__PURE__ */ t.jsxs(
        "a",
        {
          className: "text-sm flex items-center gap-1 hover:underline",
          href: "https://swagger.io/docs/specification/v3_0/data-models/dictionaries/",
          rel: "noopener noreferrer",
          target: "_blank",
          children: [
            "Additional properties are allowed",
            /* @__PURE__ */ t.jsx(X, { size: 14 })
          ]
        }
      ) })
    ] });
  }
};
export {
  v as E,
  Fe as P,
  x as S,
  I as a,
  O as b,
  z as c,
  ae as g,
  m as i
};
