// zudoku/lib/SchemaList-3gK0DopV.js
import { j as e } from "./jsx-runtime-BzflLqGi.js";
import { s as o } from "./index-BqzS1N22.js";
import { u as j } from "./useSuspenseQuery-DbPM3Wv6.js";
import { H as c } from "./index.esm-Ca5zvoff.js";
import { ChevronsUpDownIcon as y, ChevronsDownUpIcon as N, ChevronRightIcon as b } from "lucide-react";
import { B as v } from "./Button-qpSb3K7a.js";
import { u as C, a as w, C as m, b as l, c as d, d as S } from "./index-Ckdk300F.js";
import { C as A } from "./CategoryHeading-DhmodDcq.js";
import { H as p, M as H } from "./Zudoku-BPvTuMy8.js";
import { T as I } from "./Toc-De-NLOWO.js";
import { P as z, S as M } from "./SchemaView-Cm6nRtpo.js";
const P = S(
  /* GraphQL */
  `
  query GetSchemas($input: JSON!, $type: SchemaType!) {
    schema(input: $input, type: $type) {
      title
      description
      summary
      components {
        schemas {
          name
          schema
          extensions
        }
      }
    }
  }
`
);
function R() {
  const { input: h, type: x, versions: u, version: t, options: r } = C(), f = w(P, {
    input: h,
    type: x
  }), { data: a } = j(f), n = a.schema.components?.schemas ?? [], g = Object.entries(u).length > 1, i = r?.showVersionSelect === "always" || g && r?.showVersionSelect !== "hide";
  return n.length ? /* @__PURE__ */ e.jsxs(
    "div",
    {
      className: "grid grid-cols-(--sidecar-grid-cols) gap-8 justify-between",
      "data-pagefind-filter": "section:openapi",
      "data-pagefind-meta": "section:openapi",
      children: [
        /* @__PURE__ */ e.jsx(z, { name: "category", children: a.schema.title }),
        /* @__PURE__ */ e.jsxs(c, { children: [
          /* @__PURE__ */ e.jsxs("title", { children: [
            "Schemas ",
            i ? t : ""
          ] }),
          /* @__PURE__ */ e.jsx("meta", { name: "description", content: "List of schemas used by the API." })
        ] }),
        /* @__PURE__ */ e.jsxs("div", { className: "pt-(--padding-content-top) pb-(--padding-content-bottom)", children: [
          /* @__PURE__ */ e.jsxs(m, { className: "w-full", children: [
            /* @__PURE__ */ e.jsxs("div", { className: "flex flex-col gap-y-4 sm:flex-row justify-around items-start sm:items-end", children: [
              /* @__PURE__ */ e.jsxs("div", { className: "flex-1", children: [
                /* @__PURE__ */ e.jsx(A, { children: a.schema.title }),
                /* @__PURE__ */ e.jsxs(
                  p,
                  {
                    level: 1,
                    id: "schemas",
                    registerNavigationAnchor: !0,
                    className: "mb-0",
                    children: [
                      "Schemas",
                      i && /* @__PURE__ */ e.jsxs("span", { className: "text-xl text-muted-foreground ms-1.5", children: [
                        "(",
                        t,
                        ")"
                      ] })
                    ]
                  }
                )
              ] }),
              a.schema.description && /* @__PURE__ */ e.jsxs(l, { className: "flex items-center gap-1 text-sm font-medium text-muted-foreground group", children: [
                /* @__PURE__ */ e.jsx("span", { children: "API information" }),
                /* @__PURE__ */ e.jsx(
                  y,
                  {
                    className: "group-data-[state=open]:hidden translate-y-px",
                    size: 14
                  }
                ),
                /* @__PURE__ */ e.jsx(
                  N,
                  {
                    className: "group-data-[state=closed]:hidden translate-y-px",
                    size: 13
                  }
                )
              ] })
            ] }),
            a.schema.description && /* @__PURE__ */ e.jsx(d, { className: "CollapsibleContent", children: /* @__PURE__ */ e.jsx("div", { className: "mt-4 max-w-full border rounded-sm bg-muted/25", children: /* @__PURE__ */ e.jsx(
              H,
              {
                className: "max-w-full prose-img:max-w-prose border-border p-3 lg:p-5",
                content: a.schema.description
              }
            ) }) })
          ] }),
          /* @__PURE__ */ e.jsx("hr", { className: "my-8" }),
          /* @__PURE__ */ e.jsx("div", { className: "flex flex-col gap-y-5", children: n.map((s) => /* @__PURE__ */ e.jsxs(m, { className: "group", defaultOpen: !0, children: [
            /* @__PURE__ */ e.jsxs(
              p,
              {
                registerNavigationAnchor: !0,
                level: 2,
                className: "flex items-center gap-1 justify-between w-fit",
                id: o(s.name),
                children: [
                  s.name,
                  " ",
                  /* @__PURE__ */ e.jsx(l, { asChild: !0, children: /* @__PURE__ */ e.jsx(v, { variant: "ghost", size: "icon", className: "size-6", children: /* @__PURE__ */ e.jsx(
                    b,
                    {
                      size: 16,
                      className: "group-data-[state=open]:rotate-90 transition cursor-pointer"
                    }
                  ) }) })
                ]
              }
            ),
            /* @__PURE__ */ e.jsx(d, { className: "mt-4 CollapsibleContent", children: /* @__PURE__ */ e.jsx(M, { schema: s.schema }) })
          ] }, s.name)) })
        ] }),
        /* @__PURE__ */ e.jsx(
          I,
          {
            entries: n.map((s) => ({
              id: o(s.name),
              value: s.name,
              depth: 1
            }))
          }
        )
      ]
    }
  ) : /* @__PURE__ */ e.jsxs("div", { children: [
    /* @__PURE__ */ e.jsxs(c, { children: [
      /* @__PURE__ */ e.jsxs("title", { children: [
        "Schemas ",
        i ? t : ""
      ] }),
      /* @__PURE__ */ e.jsx("meta", { name: "description", content: "List of schemas used by the API." })
    ] }),
    "No schemas found"
  ] });
}
export {
  R as SchemaList
};
