// zudoku/lib/Secret-9xKh4cip.js
import { j as e } from "./jsx-runtime-BzflLqGi.js";
import { memo as y, useState as z } from "react";
import { a as N } from "./ZudokuContext-Mfno-z8f.js";
import { h as S } from "./SyntaxHighlight-B1AjydLl.js";
import { i as k } from "./invariant-B_t_F2s_.js";
import { CodeBlock as C } from "./ui/CodeBlock.js";
import { EmbeddedCodeBlock as I } from "./ui/EmbeddedCodeBlock.js";
import * as i from "@radix-ui/react-select";
import { ChevronDownIcon as u, CheckIcon as p, ChevronUpIcon as _, EyeOffIcon as B, EyeIcon as T, CopyIcon as R } from "lucide-react";
import { cn as l } from "./ui/util.js";
import { c } from "./cn-5-Gd1Dss.js";
import { u as E } from "./useCopyToClipboard-B_085nfO.js";
import { Button as m } from "./ui/Button.js";
const H = y(
  ({ code: t, children: o, embedded: a, ...s }) => {
    const { syntaxHighlighting: r } = N().options;
    k(r?.highlighter, "Highlighter not found");
    const n = S(
      r.highlighter,
      t ?? o,
      s.language,
      r.themes
    );
    return a ? /* @__PURE__ */ e.jsx(I, { ...s, children: n }) : /* @__PURE__ */ e.jsx(C, { ...s, children: n });
  }
);
H.displayName = "SyntaxHighlight";
function Q({
  ...t
}) {
  return /* @__PURE__ */ e.jsx(i.Root, { "data-slot": "select", ...t });
}
function W({
  ...t
}) {
  return /* @__PURE__ */ e.jsx(i.Group, { "data-slot": "select-group", ...t });
}
function X({
  ...t
}) {
  return /* @__PURE__ */ e.jsx(i.Value, { "data-slot": "select-value", ...t });
}
function Y({
  className: t,
  size: o = "default",
  children: a,
  ...s
}) {
  return /* @__PURE__ */ e.jsxs(
    i.Trigger,
    {
      "data-slot": "select-trigger",
      "data-size": o,
      className: l(
        "border-input data-placeholder:text-muted-foreground [&_svg:not([class*='text-'])]:text-muted-foreground focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 dark:hover:bg-input/50 flex w-fit items-center justify-between gap-2 rounded-md border bg-transparent px-3 py-2 text-sm whitespace-nowrap shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 data-[size=default]:h-9 data-[size=sm]:h-8 *:data-[slot=select-value]:line-clamp-1 *:data-[slot=select-value]:flex *:data-[slot=select-value]:items-center *:data-[slot=select-value]:gap-2 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4",
        t
      ),
      ...s,
      children: [
        a,
        /* @__PURE__ */ e.jsx(i.Icon, { asChild: !0, children: /* @__PURE__ */ e.jsx(u, { className: "size-4 opacity-50" }) })
      ]
    }
  );
}
function ee({
  className: t,
  children: o,
  position: a = "popper",
  align: s = "center",
  ...r
}) {
  return /* @__PURE__ */ e.jsx(i.Portal, { children: /* @__PURE__ */ e.jsxs(
    i.Content,
    {
      "data-slot": "select-content",
      className: l(
        "bg-popover text-popover-foreground data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[side=bottom]:slide-in-from-top-2 data-[side=left]:slide-in-from-right-2 data-[side=right]:slide-in-from-left-2 data-[side=top]:slide-in-from-bottom-2 relative z-50 max-h-(--radix-select-content-available-height) min-w-32 origin-(--radix-select-content-transform-origin) overflow-x-hidden overflow-y-auto rounded-md border shadow-md",
        a === "popper" && "data-[side=bottom]:translate-y-1 data-[side=left]:-translate-x-1 data-[side=right]:translate-x-1 data-[side=top]:-translate-y-1",
        t
      ),
      position: a,
      align: s,
      ...r,
      children: [
        /* @__PURE__ */ e.jsx(P, {}),
        /* @__PURE__ */ e.jsx(
          i.Viewport,
          {
            className: l(
              "p-1",
              a === "popper" && "h-(--radix-select-trigger-height) w-full min-w-(--radix-select-trigger-width) scroll-my-1"
            ),
            children: o
          }
        ),
        /* @__PURE__ */ e.jsx(D, {})
      ]
    }
  ) });
}
function te({
  className: t,
  children: o,
  ...a
}) {
  return /* @__PURE__ */ e.jsxs(
    i.Item,
    {
      "data-slot": "select-item",
      className: l(
        "focus:bg-accent focus:text-accent-foreground [&_svg:not([class*='text-'])]:text-muted-foreground relative flex w-full cursor-default items-center gap-2 rounded-sm py-1.5 pr-8 pl-2 text-sm outline-hidden select-none data-disabled:pointer-events-none data-disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-0 [&_svg:not([class*='size-'])]:size-4 *:[span]:last:flex *:[span]:last:items-center *:[span]:last:gap-2",
        t
      ),
      ...a,
      children: [
        /* @__PURE__ */ e.jsx("span", { className: "absolute right-2 flex size-3.5 items-center justify-center", children: /* @__PURE__ */ e.jsx(i.ItemIndicator, { children: /* @__PURE__ */ e.jsx(p, { className: "size-4" }) }) }),
        /* @__PURE__ */ e.jsx(i.ItemText, { children: o })
      ]
    }
  );
}
function P({
  className: t,
  ...o
}) {
  return /* @__PURE__ */ e.jsx(
    i.ScrollUpButton,
    {
      "data-slot": "select-scroll-up-button",
      className: l(
        "flex cursor-default items-center justify-center py-1",
        t
      ),
      ...o,
      children: /* @__PURE__ */ e.jsx(_, { className: "size-4" })
    }
  );
}
function D({
  className: t,
  ...o
}) {
  return /* @__PURE__ */ e.jsx(
    i.ScrollDownButton,
    {
      "data-slot": "select-scroll-down-button",
      className: l(
        "flex cursor-default items-center justify-center py-1",
        t
      ),
      ...o,
      children: /* @__PURE__ */ e.jsx(u, { className: "size-4" })
    }
  );
}
const U = ({
  secret: t,
  revealed: o = !1,
  previewChars: a = 5,
  className: s
}) => {
  if (o)
    return /* @__PURE__ */ e.jsx("span", { className: s, children: t });
  const r = "•••• ".repeat(t.slice(0, -a).length / 5) + "•".repeat(t.slice(0, -a).length % 5), n = a > 0 ? t.slice(-a) : "";
  return /* @__PURE__ */ e.jsx(
    "span",
    {
      className: c(
        "block overflow-hidden text-ellipsis whitespace-nowrap",
        s
      ),
      dir: "rtl",
      children: /* @__PURE__ */ e.jsxs("span", { dir: "ltr", className: "inline", children: [
        /* @__PURE__ */ e.jsx("span", { className: "opacity-50", children: r }),
        n && /* @__PURE__ */ e.jsx("span", { children: n })
      ] })
    }
  );
}, se = ({
  secret: t,
  preview: o = 5,
  className: a,
  status: s,
  onCopy: r,
  revealed: n,
  onReveal: x,
  noReveal: h
}) => {
  const g = Math.abs(o), [f, v] = z(!1), [j, b] = E(), d = n ?? f;
  return /* @__PURE__ */ e.jsxs(
    "div",
    {
      className: c(
        "flex gap-2 items-center text-sm border rounded-md px-1",
        a
      ),
      children: [
        /* @__PURE__ */ e.jsxs(
          "div",
          {
            className: c(
              "font-mono w-full h-9 items-center flex px-2 text-xs gap-2 min-w-0",
              d ? "overflow-x-auto" : "overflow-hidden"
            ),
            children: [
              s && /* @__PURE__ */ e.jsx(
                "div",
                {
                  className: c(
                    "rounded-full shrink-0 w-2 h-2 mr-2",
                    s === "active" && "bg-emerald-400",
                    s === "expired" && "bg-neutral-200",
                    s === "expiring" && "bg-yellow-400",
                    s === "revoked" && "bg-red-400",
                    s === "none" && "opacity-0"
                  )
                }
              ),
              /* @__PURE__ */ e.jsx(
                U,
                {
                  secret: t,
                  revealed: d,
                  previewChars: g
                }
              )
            ]
          }
        ),
        h !== !0 && /* @__PURE__ */ e.jsx(
          m,
          {
            variant: "ghost",
            onClick: () => {
              v((w) => !w), x?.(!d);
            },
            size: "icon-xxs",
            children: d ? /* @__PURE__ */ e.jsx(B, { className: "size-3.5" }) : /* @__PURE__ */ e.jsx(T, { className: "size-3.5" })
          }
        ),
        /* @__PURE__ */ e.jsx(
          m,
          {
            variant: "ghost",
            onClick: () => {
              b(t), r?.(t);
            },
            size: "icon-xxs",
            children: j ? /* @__PURE__ */ e.jsx(p, { className: "size-3.5" }) : /* @__PURE__ */ e.jsx(R, { className: "size-3.5" })
          }
        )
      ]
    }
  );
};
export {
  H as S,
  Q as a,
  Y as b,
  X as c,
  ee as d,
  W as e,
  te as f,
  se as g,
  U as h
};
