// zudoku/lib/zudoku.auth-supabase.js
import { j as a } from "./jsx-runtime-BzflLqGi.js";
import { createClient as g } from "@supabase/supabase-js";
import { C as m, A as h } from "./errors-Bbk86Y_Y.js";
import { S as f } from "./SignUp-2kBWzyaB.js";
import { u } from "./ZudokuReactContext-DGJAP1sN.js";
import { Auth as v } from "@supabase/auth-ui-react";
import { ThemeSupa as S } from "@supabase/auth-ui-shared";
import { b as T } from "./chunk-EPOLDU6W-CLHrVxOG.js";
import { H as A } from "./Zudoku-BPvTuMy8.js";
const c = ({
  client: i,
  config: e,
  view: t = "sign_in"
}) => {
  const [s] = T(), r = s.get("redirectTo"), n = e.provider ? [e.provider] : e.providers, o = e.basePath ?? "/", d = r ?? e.redirectToAfterSignUp ?? o, l = r ?? e.redirectToAfterSignIn ?? o, p = r ?? e.redirectToAfterSignOut ?? o;
  return /* @__PURE__ */ a.jsx("div", { className: "flex items-center justify-center", children: /* @__PURE__ */ a.jsxs("div", { className: "max-w-md w-full mt-10", children: [
    /* @__PURE__ */ a.jsx(A, { level: 1, children: t === "sign_in" ? "Sign in" : "Sign up" }),
    /* @__PURE__ */ a.jsx(
      v,
      {
        view: t,
        redirectToAfterSignIn: l,
        redirectToAfterSignUp: d,
        redirectToAfterSignOut: p,
        supabaseClient: i,
        onlyThirdPartyProviders: e.onlyThirdPartyProviders,
        appearance: {
          theme: S,
          variables: {
            default: {
              colors: {
                dividerBackground: "var(--border)",
                brand: "var(--primary)",
                brandAccent: "hsla(from var(--primary) h s l / 0.8)",
                brandButtonText: "var(--primary-foreground)",
                defaultButtonBorder: "var(--border)",
                inputBorder: "var(--border)",
                inputText: "var(--foreground)",
                inputBorderHover: "var(--accent)",
                defaultButtonBackground: "var(--secondary)",
                defaultButtonBackgroundHover: "var(--accent)"
              },
              radii: {
                borderRadiusButton: "var(--radius)",
                buttonBorderRadius: "var(--radius)",
                inputBorderRadius: "var(--radius)"
              }
            }
          }
        },
        providers: n,
        redirectTo: e.redirectToAfterSignIn ?? "/"
      }
    )
  ] }) });
};
class b extends m {
  client;
  config;
  constructor(e) {
    const { supabaseUrl: t, supabaseKey: s } = e;
    super(), this.client = g(t, s, {
      auth: {
        autoRefreshToken: !0,
        persistSession: !0
      }
    }), this.config = e, this.client.auth.onAuthStateChange(async (r, n) => {
      n && (r === "SIGNED_IN" || r === "TOKEN_REFRESHED") ? await this.updateUserState(n) : r === "SIGNED_OUT" && u.getState().setLoggedOut();
    });
  }
  async updateUserState(e) {
    const { user: t } = e, s = {
      sub: t.id,
      email: t.email,
      name: t.user_metadata.full_name || t.user_metadata.name,
      emailVerified: t.email_confirmed_at != null,
      pictureUrl: t.user_metadata.avatar_url
    };
    u.getState().setLoggedIn({
      profile: s,
      providerData: { session: e }
    });
  }
  async getAccessToken() {
    const { data: e, error: t } = await this.client.auth.getSession();
    if (t || !e.session)
      throw new h("User is not authenticated");
    return e.session.access_token;
  }
  async signRequest(e) {
    const t = await this.getAccessToken();
    return e.headers.set("Authorization", `Bearer ${t}`), e;
  }
  signUp = async ({ navigate: e }, { redirectTo: t }) => {
    e(
      t ? `/signup?redirectTo=${encodeURIComponent(t)}` : "/signup"
    );
  };
  signIn = async ({ navigate: e }, { redirectTo: t }) => {
    e(
      t ? `/signin?redirectTo=${encodeURIComponent(t)}` : "/signin"
    );
  };
  getRoutes = () => [
    {
      path: "/signin",
      element: /* @__PURE__ */ a.jsx(
        c,
        {
          view: "sign_in",
          client: this.client,
          config: this.config
        }
      )
    },
    {
      path: "/signup",
      element: /* @__PURE__ */ a.jsx(
        c,
        {
          view: "sign_up",
          client: this.client,
          config: this.config
        }
      )
    },
    {
      path: "/signout",
      element: /* @__PURE__ */ a.jsx(f, {})
    }
  ];
  signOut = async () => {
    const { error: e } = await this.client.auth.signOut({ scope: "local" });
    e && console.error("Error signing out", e), u.getState().setLoggedOut();
  };
  onPageLoad = async () => {
    const { data: e, error: t } = await this.client.auth.getSession();
    !t && e.session && await this.updateUserState(e.session);
  };
}
const P = (i) => new b(i);
export {
  P as default
};
