// zudoku/lib/zudoku.plugin-search-pagefind.js
import { j as e } from "./jsx-runtime-BzflLqGi.js";
import { C as R } from "./ClientOnly-E7hGysn1.js";
import { VisuallyHidden as F } from "@radix-ui/react-visually-hidden";
import { u as C, k as T } from "./useQuery-DSKGiCQr.js";
import { BracketsIcon as $, FileTextIcon as q } from "lucide-react";
import { useRef as _, useLayoutEffect as I, lazy as P, useState as y, useEffect as A } from "react";
import { B as D } from "./Button-qpSb3K7a.js";
import { C as B, a as S, b as N, c as O, d as V, e as z } from "./Command-N6VujV30.js";
import { c as K } from "./Dialog-hlvmmQ_c.js";
import { cn as w } from "./ui/util.js";
import { u as G } from "./ZudokuReactContext-DGJAP1sN.js";
import { j as E, a as L } from "./ZudokuContext-Mfno-z8f.js";
import { S as U } from "./RouteGuard-DnAFajw9.js";
import { u as H, L as k } from "./chunk-EPOLDU6W-CLHrVxOG.js";
function h({ className: r, ...a }) {
  return /* @__PURE__ */ e.jsx(
    "kbd",
    {
      "data-slot": "kbd",
      className: w(
        "bg-muted text-muted-foreground pointer-events-none inline-flex h-5 w-fit min-w-5 items-center justify-center gap-1 rounded-sm px-1 font-sans text-xs font-medium select-none",
        "[&_svg:not([class*='size-'])]:size-3",
        "[[data-slot=tooltip-content]_&]:bg-background/20 [[data-slot=tooltip-content]_&]:text-background dark:[[data-slot=tooltip-content]_&]:bg-background/10",
        r
      ),
      ...a
    }
  );
}
function g({ className: r, ...a }) {
  return /* @__PURE__ */ e.jsx(
    "kbd",
    {
      "data-slot": "kbd-group",
      className: w("inline-flex items-center gap-1", r),
      ...a
    }
  );
}
const Y = async ({
  search: r,
  options: a,
  auth: s,
  context: n
}) => {
  const c = a.maxResults ?? 10, o = a.transformResults ?? (() => !0), i = [], m = M({
    search: r,
    transformFn: o,
    auth: s,
    context: n
  });
  for await (const t of m)
    if (i.push(t), i.length >= c) break;
  return i;
};
async function* M({
  search: r,
  transformFn: a,
  auth: s,
  context: n
}) {
  let o = 0;
  for (; o < r.results.length; ) {
    const i = r.results.slice(
      o,
      o + 5
    );
    o += i.length;
    const m = await Promise.all(i.map((t) => t.data()));
    for (const t of m) {
      const l = a({ result: t, auth: s, context: n });
      l === !1 || (l === !0 || l == null ? yield t : yield l);
    }
  }
}
const Q = (r, a) => {
  const s = r.weighted_locations.reduce(
    (c, o) => c + o.balanced_score,
    0
  );
  return a.weighted_locations.reduce(
    (c, o) => c + o.balanced_score,
    0
  ) - s;
}, v = "cursor-pointer border border-transparent data-[selected=true]:border-border", W = ({
  basePath: r,
  searchResults: a,
  searchTerm: s,
  onClose: n,
  maxSubResults: c = 4
}) => {
  const o = H(), i = _(null);
  I(() => {
    s && requestAnimationFrame(() => {
      i.current?.scrollTo({ top: 0 });
    });
  }, [s]);
  const m = (t) => r && t.startsWith(r) ? E(t.slice(r.length)) : t;
  return /* @__PURE__ */ e.jsxs(B, { className: "max-h-[450px]", ref: i, children: [
    s && a.length > 0 && /* @__PURE__ */ e.jsx(
      S,
      {
        className: "text-sm text-muted-foreground",
        heading: `${a.length} results for "${s}"`
      }
    ),
    s && a.map((t) => /* @__PURE__ */ e.jsxs(
      S,
      {
        children: [
          /* @__PURE__ */ e.jsx(
            N,
            {
              asChild: !0,
              value: `${t.meta.title}-${t.url}`,
              className: v,
              onSelect: () => {
                o(m(t.url)), n();
              },
              children: /* @__PURE__ */ e.jsxs(k, { to: m(t.url), children: [
                t.meta.section === "openapi" ? /* @__PURE__ */ e.jsx($, {}) : /* @__PURE__ */ e.jsx(q, {}),
                t.meta.category && /* @__PURE__ */ e.jsxs("span", { className: "text-muted-foreground", children: [
                  t.meta.category,
                  " ›",
                  " "
                ] }),
                t.meta.title
              ] })
            }
          ),
          t.sub_results.sort(Q).slice(0, c).map((l) => /* @__PURE__ */ e.jsx(
            N,
            {
              asChild: !0,
              value: `sub-${t.meta.title}-${l.url}`,
              className: v,
              onSelect: () => {
                o(m(l.url)), n();
              },
              children: /* @__PURE__ */ e.jsx(k, { to: m(l.url), onClick: n, children: /* @__PURE__ */ e.jsxs("div", { className: "flex flex-col items-start gap-2 ms-2.5 ps-5 border-l border-muted-foreground/50", children: [
                /* @__PURE__ */ e.jsx("span", { className: "font-bold", children: l.title }),
                /* @__PURE__ */ e.jsx(
                  "span",
                  {
                    className: "line-clamp-2 text-[13px] [&_mark]:bg-primary [&_mark]:text-primary-foreground",
                    dangerouslySetInnerHTML: { __html: l.excerpt }
                  }
                )
              ] }) })
            },
            `sub-${t.meta.title}-${l.url}`
          ))
        ]
      },
      [t.meta.title ?? t.excerpt, t.url].join("-")
    ))
  ] });
};
P(() => import("./IndexingDialog-CCA1H6gd.js"));
const p = {
  // Slightly lower than default because API docs tend to have repetitive terms (parameter names, HTTP methods, etc.)
  termFrequency: 0.8,
  // Lower than default because API documentation pages tend to be longer due to comprehensive endpoint documentation
  pageLength: 0.6,
  // Slightly higher than default because in technical documentation, exact matches should be prioritized
  termSimilarity: 1.2,
  // Slightly lower than default because API docs might have legitimate repetition of terms
  termSaturation: 1.2
}, Z = (r) => import(
  /* @vite-ignore */
  E(r, "/pagefind/pagefind.js")
), J = (r) => {
  const {
    options: { basePath: a }
  } = L(), { data: s, ...n } = C({
    queryKey: ["pagefind", r.ranking],
    retry: !1,
    queryFn: async () => {
      const c = await Z(a);
      return await c.init(), await c.options({
        ranking: {
          termFrequency: r.ranking?.termFrequency ?? p.termFrequency,
          pageLength: r.ranking?.pageLength ?? p.pageLength,
          termSimilarity: r.ranking?.termSimilarity ?? p.termSimilarity,
          termSaturation: r.ranking?.termSaturation ?? p.termSaturation
        }
      }), c;
    },
    enabled: typeof window < "u"
  });
  return n.isError && n.error.message !== "NOT_BUILT_YET" && console.error(n.error), { ...n, pagefind: s };
}, X = ({
  isOpen: r,
  onClose: a,
  options: s
}) => {
  const { pagefind: n, error: c, isError: o } = J(s), [i, m] = y(""), [t, l] = y(""), x = G(), j = L(), b = _(null), { data: u } = C({
    queryKey: ["pagefind-search", i, x.isAuthenticated],
    queryFn: async () => {
      const d = x.isAuthenticated ? void 0 : { not: { section: U } }, f = await n?.search(i, { filters: d });
      return f ? Y({ search: f, options: s, auth: x, context: j }) : [];
    },
    placeholderData: T,
    enabled: !!n && !!i
  });
  return A(() => {
    if (!u?.length) return;
    const d = u.at(0);
    if (!d) return;
    const f = `${d.meta.title}-${d.url}`;
    l(f);
  }, [u]), /* @__PURE__ */ e.jsxs(
    O,
    {
      command: {
        shouldFilter: !1,
        loop: !0,
        value: t,
        onValueChange: l
      },
      content: { className: "max-w-[750px]" },
      open: r,
      onOpenChange: a,
      children: [
        /* @__PURE__ */ e.jsx(F, { children: /* @__PURE__ */ e.jsx(K, { children: "Search" }) }),
        /* @__PURE__ */ e.jsx(
          V,
          {
            ref: b,
            placeholder: "Search...",
            value: i,
            onValueChange: m,
            disabled: o
          }
        ),
        /* @__PURE__ */ e.jsx(z, { children: i ? /* @__PURE__ */ e.jsxs("div", { className: "flex flex-col items-center", children: [
          "No results found.",
          /* @__PURE__ */ e.jsx(
            D,
            {
              variant: "link",
              onClick: () => {
                m(""), b.current?.focus();
              },
              children: "Clear search"
            }
          )
        ] }) : "Start typing to search" }),
        o && c.message !== "NOT_BUILT_YET" ? /* @__PURE__ */ e.jsx("div", { className: "p-4 text-sm", children: "An error occurred while loading search." }) : /* @__PURE__ */ e.jsxs(e.Fragment, { children: [
          /* @__PURE__ */ e.jsx(
            W,
            {
              basePath: j.options.basePath,
              searchResults: u ?? [],
              searchTerm: i,
              onClose: a,
              maxSubResults: s.maxSubResults
            }
          ),
          /* @__PURE__ */ e.jsxs("div", { className: "flex justify-between p-2 items-center", children: [
            /* @__PURE__ */ e.jsxs("div", { className: "flex items-center text-xs text-muted-foreground", children: [
              /* @__PURE__ */ e.jsxs(g, { className: "ms-0 me-1", children: [
                /* @__PURE__ */ e.jsx(h, { children: "↑" }),
                /* @__PURE__ */ e.jsx(h, { children: "↓" })
              ] }),
              "Navigate",
              /* @__PURE__ */ e.jsx(g, { className: "ms-4 me-1", children: /* @__PURE__ */ e.jsx(h, { children: "↵" }) }),
              "Select",
              /* @__PURE__ */ e.jsx(g, { className: "ms-4 me-1", children: /* @__PURE__ */ e.jsx(h, { children: "Esc" }) }),
              "Close dialog"
            ] }),
            !1
          ] })
        ] })
      ]
    }
  );
}, he = (r) => ({
  renderSearch: ({ isOpen: a, onClose: s }) => /* @__PURE__ */ e.jsx(R, { children: /* @__PURE__ */ e.jsx(X, { isOpen: a, onClose: s, options: r }) })
});
export {
  he as pagefindSearchPlugin
};
