// zudoku/lib/zudoku.auth-auth0.js
import { j as a } from "./ZudokuContext-Mfno-z8f.js";
import { u as r } from "./ZudokuReactContext-DGJAP1sN.js";
import { OpenIDAuthenticationProvider as p } from "./zudoku.auth-openid.js";
class d extends p {
  options;
  constructor(t) {
    super({
      ...t,
      type: "openid",
      issuer: `https://${t.domain}/`,
      clientId: t.clientId,
      audience: t.audience,
      scopes: t.scopes
    }), this.options = t.options;
  }
  onAuthorizationUrl = async (t, { isSignUp: e }) => {
    this.options?.prompt !== void 0 ? this.options.prompt !== "" && t.searchParams.set("prompt", this.options.prompt) : this.options?.alwaysPromptLogin !== !1 && t.searchParams.set("prompt", "login"), e && t.searchParams.set("screen_hint", "signup");
  };
  signOut = async (t) => {
    const e = await this.getAuthServer(), s = r.getState().providerData?.idToken;
    r.setState({
      isAuthenticated: !1,
      isPending: !1,
      profile: void 0,
      providerData: void 0
    });
    const n = new URL(window.location.origin);
    if (n.pathname = a(
      "/",
      this.redirectToAfterSignOut
    ), e.end_session_endpoint) {
      const o = new URL(e.end_session_endpoint);
      s && o.searchParams.set("id_token_hint", s), o.searchParams.set(
        "post_logout_redirect_uri",
        n.toString()
      ), window.location.href = o.toString();
    }
  };
}
const l = (i) => new d(i);
export {
  l as default
};
