// zudoku/lib/zudoku.plugin-api-catalog.js
import { j as t } from "./jsx-runtime-BzflLqGi.js";
import { s as f } from "./index-BqzS1N22.js";
import { d as b, L as j, m as x } from "./chunk-EPOLDU6W-CLHrVxOG.js";
import { j as p } from "./ZudokuContext-Mfno-z8f.js";
import { u as v } from "./useSuspenseQuery-DbPM3Wv6.js";
import { H as y } from "./index.esm-Ca5zvoff.js";
import "./ui/Button.js";
import "lucide-react";
import "react";
import { H as N, M as k } from "./Zudoku-BPvTuMy8.js";
import { u as S } from "./ZudokuReactContext-DGJAP1sN.js";
import "./index-DAWHN3cH.js";
const w = ({
  items: l,
  filterCatalogItems: n = (i) => i,
  label: c = "API Library",
  categoryLabel: s
}) => {
  const i = S(), a = b({ path: "/catalog/:category" })?.params.category, o = v({
    queryFn: () => n(l, { auth: i }),
    queryKey: ["catalogItems", i]
  }), m = a ? { "data-pagefind-ignore": "all" } : {};
  return /* @__PURE__ */ t.jsxs("section", { className: "pt-(--padding-content-top)", ...m, children: [
    /* @__PURE__ */ t.jsx(y, { children: /* @__PURE__ */ t.jsxs("title", { children: [
      s ? `${s} - ` : "",
      c
    ] }) }),
    /* @__PURE__ */ t.jsxs("div", { className: "grid gap-4", children: [
      /* @__PURE__ */ t.jsxs(N, { level: 2, children: [
        c,
        s && ` - ${s}`
      ] }),
      /* @__PURE__ */ t.jsx("div", { className: "grid grid-cols-2 gap-4", children: o.data.filter(
        (e) => !a || e.categories.find(
          (r) => r.tags.find((g) => h(r.label, g) === a)
        )
      ).map((e) => /* @__PURE__ */ t.jsx(
        j,
        {
          to: p(e.path),
          className: "no-underline hover:!text-foreground",
          children: /* @__PURE__ */ t.jsxs("div", { className: "border h-full rounded-lg p-4 flex flex-col gap-2 cursor-pointer hover:bg-border/20 font-normal", children: [
            /* @__PURE__ */ t.jsx("span", { className: "font-semibold", children: e.label }),
            /* @__PURE__ */ t.jsx(
              k,
              {
                className: "text-sm whitespace-pre-wrap mb-6 line-clamp-2",
                content: e.description,
                components: {
                  // Because we're wrapping the description in a Link already,
                  // we need to strip out other links to not get a hydration error, like:
                  // > In HTML, <a> cannot be a descendant of <a>.
                  // > This will cause a hydration error.
                  a: (r) => /* @__PURE__ */ t.jsx("span", { ...r })
                }
              }
            )
          ] })
        },
        e.path
      )) })
    ] })
  ] });
}, h = (l, n) => f(`${l}-${n}`), F = ({
  path: l,
  items: n,
  label: c,
  categories: s = [],
  filterCatalogItems: i
}) => {
  const d = Object.fromEntries(
    s.flatMap(
      (a) => [void 0, ...a.tags].map((o) => [
        p(l, o ? h(a.label, o) : void 0),
        o
      ])
    )
  );
  return {
    getNavigation: async (a) => {
      if (!Object.keys(d).some(
        (e) => x(e, a)
      ))
        return [];
      const m = s.map((e) => ({
        type: "category",
        label: e.label,
        collapsible: !1,
        items: e.tags.map((r) => ({
          type: "link",
          to: p(l, h(e.label, r)),
          label: r,
          badge: {
            label: String(
              n.filter(
                (g) => g.categories.find((u) => u.tags.includes(r))
              ).length
            ),
            color: "outline"
          }
        }))
      }));
      return m.unshift({
        type: "link",
        to: p(l),
        label: "Overview",
        badge: { label: String(n.length), color: "outline" }
      }), m;
    },
    getRoutes: () => Object.entries(d).map(([a, o]) => ({
      path: a,
      element: /* @__PURE__ */ t.jsx(
        w,
        {
          label: c,
          categoryLabel: o,
          items: n,
          filterCatalogItems: i,
          categories: s
        }
      )
    }))
  };
};
export {
  F as apiCatalogPlugin,
  h as getKey
};
