// zudoku/lib/zudoku.plugins.js
const t = (n) => n, i = (n) => n, e = (n) => "events" in n && typeof n.events == "object", o = (n) => "getProfileMenuItems" in n && typeof n.getProfileMenuItems == "function", s = (n) => "getRoutes" in n && typeof n.getRoutes == "function", f = (n) => "signUp" in n && typeof n.signUp == "function", c = (n) => "renderSearch" in n && typeof n.renderSearch == "function", u = (n) => "initialize" in n && typeof n.initialize == "function", g = (n) => "getHead" in n && typeof n.getHead == "function", r = (n) => "getMdxComponents" in n && typeof n.getMdxComponents == "function", a = (n) => "getIdentities" in n && typeof n.getIdentities == "function", l = (n) => "transformConfig" in n && typeof n.transformConfig == "function";
export {
  t as createApiIdentityPlugin,
  i as createProfileMenuPlugin,
  g as hasHead,
  a as isApiIdentityPlugin,
  f as isAuthenticationPlugin,
  e as isEventConsumerPlugin,
  r as isMdxProviderPlugin,
  s as isNavigationPlugin,
  o as isProfileMenuPlugin,
  c as isSearchPlugin,
  l as isTransformConfigPlugin,
  u as needsInitialization
};
