// zudoku/dist/lib/plugins/openapi/playground/Headers.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { CircleAlertIcon, LockIcon, TableOfContentsIcon } from "lucide-react";
import { useFormContext } from "react-hook-form";
import { Checkbox } from "zudoku/ui/Checkbox.js";
import { Collapsible, CollapsibleContent } from "zudoku/ui/Collapsible.js";
import { Tooltip, TooltipContent, TooltipTrigger } from "zudoku/ui/Tooltip.js";
import { Autocomplete } from "../../../components/Autocomplete.js";
import { cn } from "../../../util/cn.js";
import { CollapsibleHeader, CollapsibleHeaderTrigger, } from "./CollapsibleHeader.js";
import ParamsGrid, { ParamsGridInput, ParamsGridItem, ParamsGridRemoveButton, } from "./ParamsGrid.js";
import { useKeyValueFieldManager } from "./request-panel/useKeyValueFieldManager.js";
// biome-ignore format: Easier to read
const headerOptions = Object.freeze([
    "Accept", "Accept-Encoding", "Accept-Language", "Authorization", "Cache-Control", "Connection",
    "Content-Disposition", "Content-Encoding", "Content-Language", "Content-Length", "Content-Range",
    "Content-Security-Policy", "Content-Type", "Cookie", "Date", "ETag", "Expires", "Host",
    "If-Modified-Since", "Location", "Origin", "Pragma", "Referer", "Set-Cookie", "User-Agent",
    "X-Requested-With",
]);
export const Headers = ({ control, schemaHeaders, lockedHeaders, }) => {
    const { watch, formState } = useFormContext();
    const watchedHeaders = watch("headers");
    const manager = useKeyValueFieldManager({
        control,
        name: "headers",
        defaultValue: { name: "", value: "", active: false },
    });
    const missingHeaders = schemaHeaders
        .filter((h) => !watchedHeaders.some((f) => f.name === h.name))
        .map(({ name }) => name);
    const hiddenHeadersIndex = manager.fields.flatMap((f, index) => {
        const keep = !lockedHeaders
            ?.map((h) => h.toLowerCase())
            .includes(f.name.toLowerCase());
        return keep ? [] : [index];
    });
    const lockedHeaderFields = lockedHeaders?.map((h) => ({
        name: h,
        id: `locked-${h}`,
        value: "••••••••••",
        active: true,
        locked: true,
    })) ?? [];
    return (_jsxs(Collapsible, { defaultOpen: true, children: [_jsxs(CollapsibleHeaderTrigger, { children: [_jsx(TableOfContentsIcon, { size: 14 }), _jsx(CollapsibleHeader, { children: "Headers" })] }), _jsx(CollapsibleContent, { className: "CollapsibleContent", children: _jsx("div", { className: "flex flex-col gap-2", children: _jsx("div", { className: "overflow-hidden", children: _jsxs(ParamsGrid, { children: [lockedHeaderFields.map((field) => (_jsxs(Tooltip, { children: [_jsx(TooltipTrigger, { asChild: true, children: _jsxs(ParamsGridItem, { className: "opacity-50 cursor-not-allowed font-mono text-xs min-h-10", children: [_jsx(LockIcon, { size: 16 }), _jsx(ParamsGridInput, { value: field.name, disabled: true }), _jsx("div", { children: field.value })] }, field.id) }), _jsx(TooltipContent, { alignOffset: 10, side: "bottom", align: "start", children: _jsx("p", { children: "This header is set by the selected authentication." }) })] }, field.id))), manager.fields.map((field, i) => {
                                    const currentSchemaHeader = schemaHeaders.find((h) => h.name === watchedHeaders.at(i)?.name);
                                    const hasEnum = currentSchemaHeader?.enum &&
                                        currentSchemaHeader.enum.length > 0;
                                    const isHidden = hiddenHeadersIndex.includes(i);
                                    const nameInputProps = manager.getNameInputProps(i);
                                    const valueInputProps = manager.getValueInputProps(i);
                                    return (_jsxs(ParamsGridItem, { className: cn(isHidden && "text-amber-600", isHidden &&
                                            !formState.dirtyFields.headers?.[i]?.value &&
                                            "hidden"), children: [_jsx(Checkbox, { className: cn(isHidden && "hidden"), ...manager.getCheckboxProps(i) }), _jsxs(Tooltip, { children: [_jsx(TooltipTrigger, { asChild: true, children: _jsx(CircleAlertIcon, { className: cn("text-amber-600", !isHidden && "hidden"), size: 16 }) }), _jsx(TooltipContent, { alignOffset: 10, side: "bottom", align: "start", children: _jsx("p", { children: "This header will be overwritten by the selected authentication." }) })] }), _jsx(ParamsGridInput, { asChild: true, children: _jsx(Autocomplete, { ...nameInputProps, value: String(manager.getValue(i, "name")), placeholder: "Name", options: [...missingHeaders, ...headerOptions], onChange: (v) => manager.setValue(i, "name", v), onSelect: (v) => manager.setValue(i, "name", v, { focus: "next" }) }) }), _jsxs("div", { className: "flex items-center gap-2", children: [!hasEnum ? (_jsx(ParamsGridInput, { placeholder: "Value", autoComplete: "off", ...valueInputProps })) : (_jsx(ParamsGridInput, { asChild: true, children: _jsx(Autocomplete, { ...valueInputProps, value: String(manager.getValue(i, "value")), shouldFilter: false, options: currentSchemaHeader.enum ?? [], onChange: (v) => manager.setValue(i, "value", v), onSelect: (v) => manager.setValue(i, "value", v, { focus: "next" }) }) })), _jsx(ParamsGridRemoveButton, { ...manager.getRemoveButtonProps(i) })] })] }, field.id));
                                })] }) }) }) })] }));
};
