// zudoku/dist/lib/plugins/openapi/OasProvider.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useMemo } from "react";
import { Outlet } from "react-router";
import { joinUrl } from "../../util/joinUrl.js";
import { GraphQLProvider } from "./client/GraphQLContext.js";
import { OasConfigProvider } from "./context.js";
import { getVersionMetadata } from "./util/getRoutes.js";
export const OasProvider = ({ basePath, version, config, client, }) => {
    const value = useMemo(() => {
        const { versions: availableVersions, labels, downloadUrls, } = getVersionMetadata(config);
        const currentVersion = version ?? availableVersions.at(0);
        const versionLinks = Object.fromEntries(availableVersions.map((id) => [
            id,
            {
                path: joinUrl(basePath, id),
                label: labels[id] ?? id,
                downloadUrl: downloadUrls[id],
            },
        ]));
        const resolveInput = () => {
            if (!Array.isArray(config.input)) {
                return config.input;
            }
            const versionConfig = currentVersion
                ? config.input.find((v) => v.path === currentVersion)
                : config.input[0];
            if (!versionConfig) {
                throw new Error(`No input found for version: ${currentVersion}`);
            }
            return versionConfig.input;
        };
        return {
            config: {
                ...config,
                version: currentVersion,
                versions: versionLinks,
                input: resolveInput(),
            },
        };
    }, [config, basePath, version]);
    return (_jsx(OasConfigProvider, { value: value, children: _jsx(GraphQLProvider, { client: client, children: _jsx(Outlet, {}) }) }));
};
